// devframe@0.5.4/dist/client/index.mjs
import { t as hash } from "../hash-bwOD8RgU.mjs";
import { colors } from "../utils/colors.mjs";
import { createEventEmitter } from "../utils/events.mjs";
import { humanId } from "../utils/human-id.mjs";
import { t as createSharedState } from "../shared-state-u0y123fi.mjs";
import { i as structuredCloneStringify, n as structuredCloneParse, t as structuredCloneDeserialize } from "../structured-clone-PdCZwt7F.mjs";
import { createStreamReader, createStreamSink } from "../utils/streaming-channel.mjs";
import { promiseWithResolver } from "../utils/promise.mjs";
import { defineDiagnostics } from "nostics";
import { ansiFormatter } from "nostics/formatters/ansi";
import { createBirpc } from "birpc";
//#region src/constants.ts
const DEVFRAME_CONNECTION_META_FILENAME = "__connection.json";
const DEVFRAME_RPC_DUMP_MANIFEST_FILENAME = "__rpc-dump/index.json";
//#endregion
//#region src/utils/diagnostics-reporter.ts
const formatAnsi = ansiFormatter(colors);
function devframeReporter(d, { method = "warn" } = {}) {
	console[method](formatAnsi(d));
}
//#endregion
//#region src/rpc/diagnostics.ts
const diagnostics = defineDiagnostics({
	docsBase: "https://devfra.me/errors",
	reporters: [devframeReporter],
	codes: {
		DF0019: {
			why: (p) => `RPC function "${p.name}" has \`agent\` set but \`jsonSerializable\` is not \`true\` — MCP requires JSON-serializable data.`,
			fix: "Set `jsonSerializable: true` if the payload is JSON-safe, or remove `agent` to keep it RPC-only."
		},
		DF0020: {
			why: (p) => `RPC function "${p.name}" declares \`jsonSerializable: true\` but the value at "${p.path}" is a ${p.type}.`,
			fix: "Either drop `jsonSerializable: true` (falls back to structured-clone) or change the value to a JSON-safe shape."
		},
		DF0021: {
			why: (p) => `RPC function "${p.name}" is already registered`,
			fix: "Use the `force` parameter to overwrite an existing registration."
		},
		DF0022: { why: (p) => `RPC function "${p.name}" is not registered. Use register() to add new functions.` },
		DF0023: { why: (p) => `RPC function "${p.name}" is not registered` },
		DF0024: { why: (p) => `Either handler or setup function must be provided for RPC function "${p.name}"` },
		DF0025: { why: (p) => `Function "${p.name}" not found in dump store` },
		DF0026: { why: (p) => `No dump match for "${p.name}" with args: ${p.args}` },
		DF0027: { why: (p) => `Function "${p.name}" with type "${p.type}" cannot have dump configuration. Only "static" and "query" types support dumps.` },
		DF0028: {
			why: (p) => `Function "${p.name}" with type "${p.type}" cannot use \`snapshot: true\`. Only "query" functions support this sugar; "static" functions have equivalent default behavior already.`,
			fix: "Remove `snapshot: true`, or change the function type to `query`."
		}
	}
});
//#endregion
//#region src/rpc/dump/error.ts
/**
* Inverse of {@link serializeDumpError}: rebuild a thrown `Error` from
* the plain object stored in a dump record. Preserves `cause`, restores
* the original `name`, and re-attaches any custom own properties.
*/
function reviveDumpError(stored) {
	const cause = stored.cause instanceof Error ? stored.cause : isPlainErrorShape(stored.cause) ? reviveDumpError(stored.cause) : stored.cause;
	const error = cause !== void 0 ? new Error(stored.message, { cause }) : new Error(stored.message);
	error.name = stored.name;
	for (const key of Object.keys(stored)) {
		if (key === "name" || key === "message" || key === "cause") continue;
		error[key] = stored[key];
	}
	return error;
}
function isPlainErrorShape(value) {
	return typeof value === "object" && value !== null && typeof value.message === "string" && typeof value.name === "string";
}
//#endregion
//#region src/rpc/handler.ts
async function getRpcResolvedSetupResult(definition, context) {
	if (!definition.setup) return {};
	if (typeof context === "object" && context !== null) {
		definition.__cache ??= /* @__PURE__ */ new WeakMap();
		let promise = definition.__cache.get(context);
		if (!promise) {
			promise = Promise.resolve(definition.setup(context));
			definition.__cache.set(context, promise);
		}
		return await promise;
	}
	definition.__promise ??= Promise.resolve(definition.setup(context));
	return await definition.__promise;
}
async function getRpcHandler(definition, context) {
	if (definition.handler) return definition.handler;
	const result = await getRpcResolvedSetupResult(definition, context);
	if (!result.handler) throw diagnostics.DF0024({ name: definition.name });
	return result.handler;
}
//#endregion
//#region src/rpc/cache.ts
/**
* @experimental API is expected to change.
*/
var RpcCacheManager = class {
	cacheMap = /* @__PURE__ */ new Map();
	options;
	keySerializer;
	constructor(options) {
		this.options = options;
		this.keySerializer = options.keySerializer || ((args) => hash(args));
	}
	updateOptions(options) {
		this.options = {
			...this.options,
			...options
		};
	}
	cached(m, a) {
		const methodCache = this.cacheMap.get(m);
		if (methodCache) return methodCache.get(this.keySerializer(a));
	}
	apply(req, res) {
		const methodCache = this.cacheMap.get(req.m) || /* @__PURE__ */ new Map();
		methodCache.set(this.keySerializer(req.a), res);
		this.cacheMap.set(req.m, methodCache);
	}
	validate(m) {
		return this.options.functions.includes(m);
	}
	clear(fn) {
		if (fn) this.cacheMap.delete(fn);
		else this.cacheMap.clear();
	}
};
//#endregion
//#region src/rpc/collector.ts
var RpcFunctionsCollectorBase = class {
	context;
	definitions = /* @__PURE__ */ new Map();
	functions;
	_onChanged = [];
	constructor(context) {
		this.context = context;
		const definitions = this.definitions;
		const self = this;
		this.functions = new Proxy({}, {
			get(_, prop) {
				const definition = definitions.get(prop);
				if (!definition) return void 0;
				return getRpcHandler(definition, self.context);
			},
			has(_, prop) {
				return definitions.has(prop);
			},
			getOwnPropertyDescriptor(_, prop) {
				return {
					value: definitions.get(prop)?.handler,
					configurable: true,
					enumerable: true
				};
			},
			ownKeys() {
				return Array.from(definitions.keys());
			}
		});
	}
	register(fn, force = false) {
		if (this.definitions.has(fn.name) && !force) throw diagnostics.DF0021({ name: fn.name });
		assertAgentJsonSerializable(fn);
		this.definitions.set(fn.name, fn);
		this._onChanged.forEach((cb) => cb(fn.name));
	}
	update(fn, force = false) {
		if (!this.definitions.has(fn.name) && !force) throw diagnostics.DF0022({ name: fn.name });
		assertAgentJsonSerializable(fn);
		this.definitions.set(fn.name, fn);
		this._onChanged.forEach((cb) => cb(fn.name));
	}
	onChanged(fn) {
		this._onChanged.push(fn);
		return () => {
			const index = this._onChanged.indexOf(fn);
			if (index !== -1) this._onChanged.splice(index, 1);
		};
	}
	async getHandler(name) {
		return await getRpcHandler(this.definitions.get(name), this.context);
	}
	getSchema(name) {
		const definition = this.definitions.get(name);
		if (!definition) throw diagnostics.DF0023({ name: String(name) });
		return {
			args: definition.args,
			returns: definition.returns
		};
	}
	has(name) {
		return this.definitions.has(name);
	}
	get(name) {
		return this.definitions.get(name);
	}
	list() {
		return Array.from(this.definitions.keys());
	}
};
function assertAgentJsonSerializable(fn) {
	if (fn.agent && fn.jsonSerializable !== true) throw diagnostics.DF0019({ name: fn.name });
}
//#endregion
//#region src/rpc/serialization.ts
/**
* `JSON.stringify` with a single-pass strict replacer.
*
* Throws `DF0020` synchronously when the value contains a type JSON
* cannot round-trip losslessly: `Map`, `Set`, `Date`, `BigInt`, class
* instances, or `undefined` inside an array (silently becomes `null`).
*
* Native pass-throughs (no extra work needed):
*   - circular references — `JSON.stringify` raises `TypeError`.
*   - `BigInt` at top level — caught here for a friendlier error path.
*
* Lenient cases (allowed without throwing):
*   - `undefined` as an object property — legitimate optional field;
*     JSON.stringify just omits it.
*   - `undefined` at the root — legitimate "action returned nothing".
*   - `Symbol` / `Function` values — semantically "drop me" in JSON.
*
* `fnName` is used only for the diagnostic message — pass the RPC
* function name when calling from a wire serializer / dump writer so
* the error points at the offending function.
*/
function strictJsonStringify(value, fnName = "") {
	return JSON.stringify(value, function strictReplacer(key, val) {
		const holder = this;
		const original = holder != null ? holder[key] : val;
		if (original === void 0) {
			if (Array.isArray(holder)) throw nonJsonAt(fnName, "undefined", holder, key);
			return val;
		}
		if (original === null) return val;
		if (typeof original === "bigint") throw nonJsonAt(fnName, "BigInt", holder, key);
		if (typeof original === "object") {
			if (original instanceof Map) throw nonJsonAt(fnName, "Map", holder, key);
			if (original instanceof Set) throw nonJsonAt(fnName, "Set", holder, key);
			if (original instanceof Date) throw nonJsonAt(fnName, "Date", holder, key);
			if (Array.isArray(original)) return val;
			const proto = Object.getPrototypeOf(original);
			if (proto !== null && proto !== Object.prototype) throw nonJsonAt(fnName, original.constructor?.name ?? "class instance", holder, key);
		}
		return val;
	});
}
function nonJsonAt(fnName, type, parent, key) {
	const path = formatPath(parent, key);
	return diagnostics.DF0020({
		name: fnName || "<anonymous>",
		type,
		path
	});
}
function formatPath(parent, key) {
	if (Array.isArray(parent)) return `[${key}]`;
	if (key === "") return "<root>";
	return key;
}
//#endregion
//#region src/client/rpc-shared-state.ts
function createRpcSharedStateClientHost(rpc) {
	const sharedState = /* @__PURE__ */ new Map();
	const initialValues = /* @__PURE__ */ new Map();
	const keyAddedListeners = /* @__PURE__ */ new Set();
	const isStaticBackend = rpc.connectionMeta.backend === "static";
	function mergeWithInitialValue(key, serverState) {
		const initial = initialValues.get(key);
		if (initial && typeof initial === "object" && !Array.isArray(initial) && typeof serverState === "object" && !Array.isArray(serverState)) return {
			...initial,
			...serverState
		};
		return serverState;
	}
	rpc.client.register({
		name: "devframe:rpc:client-state:updated",
		type: "event",
		handler: (key, fullState, syncId) => {
			const state = sharedState.get(key);
			if (!state || state.syncIds.has(syncId)) return;
			state.mutate(() => mergeWithInitialValue(key, fullState), syncId);
		}
	});
	rpc.client.register({
		name: "devframe:rpc:client-state:patch",
		type: "event",
		handler: (key, patches, syncId) => {
			const state = sharedState.get(key);
			if (!state || state.syncIds.has(syncId)) return;
			state.patch(patches, syncId);
		}
	});
	function registerSharedState(key, state) {
		const offs = [];
		offs.push(state.on("updated", (fullState, patches, syncId) => {
			if (isStaticBackend) return;
			if (patches) rpc.callEvent("devframe:rpc:server-state:patch", key, patches, syncId);
			else rpc.callEvent("devframe:rpc:server-state:set", key, fullState, syncId);
		}));
		return () => {
			for (const off of offs) off();
		};
	}
	return {
		keys: () => Array.from(sharedState.keys()),
		onKeyAdded(fn) {
			keyAddedListeners.add(fn);
			return () => {
				keyAddedListeners.delete(fn);
			};
		},
		get: async (key, options) => {
			if (options?.initialValue !== void 0) initialValues.set(key, options.initialValue);
			if (sharedState.has(key)) return sharedState.get(key);
			const state = createSharedState({
				initialValue: options?.initialValue,
				enablePatches: false
			});
			async function initSharedState() {
				if (!isStaticBackend) rpc.callEvent("devframe:rpc:server-state:subscribe", key);
				if (options?.initialValue !== void 0) {
					sharedState.set(key, state);
					for (const fn of keyAddedListeners) fn(key);
					rpc.call("devframe:rpc:server-state:get", key).then((serverState) => {
						if (serverState !== void 0) state.mutate(() => mergeWithInitialValue(key, serverState));
					}).catch((error) => {
						console.error("Error getting server state", error);
					});
					registerSharedState(key, state);
					return state;
				} else {
					const serverValue = await rpc.call("devframe:rpc:server-state:get", key);
					state.mutate(() => mergeWithInitialValue(key, serverValue));
					sharedState.set(key, state);
					for (const fn of keyAddedListeners) fn(key);
					registerSharedState(key, state);
					return state;
				}
			}
			return new Promise((resolve) => {
				if (!rpc.isTrusted) {
					resolve(state);
					rpc.events.on("rpc:is-trusted:updated", (isTrusted) => {
						if (isTrusted) initSharedState();
					});
				} else initSharedState().then(resolve);
			});
		}
	};
}
//#endregion
//#region src/client/static-rpc.ts
function isStaticEntry(value) {
	return typeof value === "object" && value !== null && value.type === "static" && typeof value.path === "string";
}
function isQueryEntry(value) {
	return typeof value === "object" && value !== null && value.type === "query" && typeof value.records === "object" && value.records !== null;
}
function isRecord(value) {
	return typeof value === "object" && value !== null && ("output" in value || "error" in value);
}
function resolveRecordOutput(record) {
	if (record.error) throw reviveDumpError(record.error);
	return record.output;
}
function hasMeaningfulArgs(args) {
	return args.some((arg) => arg !== null && arg !== void 0);
}
function unwrapEnvelope(raw) {
	if (raw !== null && typeof raw === "object" && "serialization" in raw && "data" in raw) return raw.data;
	return raw;
}
function createStaticRpcCaller(manifest, fetchJson) {
	const staticCache = /* @__PURE__ */ new Map();
	const queryRecordCache = /* @__PURE__ */ new Map();
	function reviveIfStructuredClone(value, serialization) {
		if (serialization === "structured-clone" && Array.isArray(value)) return structuredCloneDeserialize(value);
		return value;
	}
	function decode(raw, serialization) {
		return reviveIfStructuredClone(unwrapEnvelope(raw), serialization);
	}
	async function loadStatic(entry) {
		if (!staticCache.has(entry.path)) staticCache.set(entry.path, fetchJson(entry.path).then((raw) => decode(raw, entry.serialization)));
		const data = await staticCache.get(entry.path);
		if (isRecord(data)) return resolveRecordOutput(data);
		return data;
	}
	async function loadQueryRecord(path, serialization) {
		if (!queryRecordCache.has(path)) queryRecordCache.set(path, fetchJson(path).then((raw) => decode(raw, serialization)));
		return await queryRecordCache.get(path);
	}
	async function call(functionName, args) {
		if (!(functionName in manifest)) throw new Error(`[devframe-rpc] Function "${functionName}" not found in dump store`);
		const entry = manifest[functionName];
		if (isStaticEntry(entry)) {
			if (hasMeaningfulArgs(args)) throw new Error(`[devframe-rpc] No dump match for "${functionName}" with args: ${JSON.stringify(args)}`);
			return await loadStatic(entry);
		}
		if (isQueryEntry(entry)) {
			const argsHash = hash(args);
			const recordPath = entry.records[argsHash];
			if (recordPath) return resolveRecordOutput(await loadQueryRecord(recordPath, entry.serialization));
			if (entry.fallback) return resolveRecordOutput(await loadQueryRecord(entry.fallback, entry.serialization));
			throw new Error(`[devframe-rpc] No dump match for "${functionName}" with args: ${JSON.stringify(args)}`);
		}
		if (!hasMeaningfulArgs(args)) return entry;
		throw new Error(`[devframe-rpc] No dump match for "${functionName}" with args: ${JSON.stringify(args)}`);
	}
	return {
		call: async (functionName, args) => await call(functionName, args),
		callOptional: async (functionName, args) => {
			if (!(functionName in manifest)) return void 0;
			return await call(functionName, args);
		},
		callEvent: async (_functionName, _args) => {}
	};
}
//#endregion
//#region src/client/rpc-static.ts
async function createStaticRpcClientMode(options) {
	const staticCaller = createStaticRpcCaller(await options.fetchJsonFromBases(DEVFRAME_RPC_DUMP_MANIFEST_FILENAME), options.fetchJsonFromBases);
	return {
		isTrusted: true,
		requestTrust: async () => true,
		requestTrustWithToken: async () => true,
		ensureTrusted: async () => true,
		call: (...args) => staticCaller.call(args[0], args.slice(1)),
		callEvent: (...args) => staticCaller.callEvent(args[0], args.slice(1)),
		callOptional: (...args) => staticCaller.callOptional(args[0], args.slice(1))
	};
}
//#endregion
//#region src/client/rpc-streaming.ts
const STREAM_KEY_SEPARATOR = "";
function streamKey(channel, id) {
	return `${channel}${STREAM_KEY_SEPARATOR}${id}`;
}
/**
* Client-side streaming host. Mirrors `createRpcSharedStateClientHost`:
* registers the two `:chunk` / `:end` event handlers once, then per-stream
* state lives in a `Map<streamKey, StreamReader>`.
*/
function createRpcStreamingClientHost(rpc) {
	const readers = /* @__PURE__ */ new Map();
	const uploads = /* @__PURE__ */ new Map();
	rpc.client.register({
		name: "devframe:streaming:chunk",
		type: "event",
		handler(channel, id, seq, chunk) {
			readers.get(streamKey(channel, id))?._push(seq, chunk);
		}
	});
	rpc.client.register({
		name: "devframe:streaming:end",
		type: "event",
		handler(channel, id, error) {
			const key = streamKey(channel, id);
			const reader = readers.get(key);
			if (!reader) return;
			reader._end(error);
			readers.delete(key);
		}
	});
	rpc.client.register({
		name: "devframe:streaming:upload-cancel",
		type: "event",
		handler(channel, id) {
			const key = streamKey(channel, id);
			const sink = uploads.get(key);
			if (!sink) return;
			sink.abort("server cancelled upload");
			uploads.delete(key);
		}
	});
	rpc.events.on("rpc:is-trusted:updated", (isTrusted) => {
		if (!isTrusted) return;
		for (const [key, reader] of readers) {
			if (reader.cancelled || reader.done) continue;
			const sepIdx = key.indexOf(STREAM_KEY_SEPARATOR);
			if (sepIdx < 0) continue;
			const channel = key.slice(0, sepIdx);
			const id = key.slice(sepIdx + 1);
			rpc.callEvent("devframe:streaming:subscribe", channel, id, { afterSeq: reader.lastSeenSeq });
		}
	});
	function subscribe(channel, id, options = {}) {
		const key = streamKey(channel, id);
		const existing = readers.get(key);
		if (existing) return existing;
		const reader = createStreamReader({
			id,
			highWaterMark: options.highWaterMark,
			onOverflow(dropped) {
				console.warn(`[devframe] DF0029: Stream "${channel}#${id}" dropped ${dropped} chunk(s) after exceeding the client high-water mark.`);
			},
			onCancel() {
				rpc.callEvent("devframe:streaming:cancel", channel, id);
				readers.delete(key);
			}
		});
		readers.set(key, reader);
		if (rpc.isTrusted) rpc.callEvent("devframe:streaming:subscribe", channel, id, { afterSeq: 0 });
		else {
			const off = rpc.events.on("rpc:is-trusted:updated", (trusted) => {
				if (trusted) {
					off();
					if (readers.has(key) && !reader.cancelled && !reader.done) rpc.callEvent("devframe:streaming:subscribe", channel, id, { afterSeq: reader.lastSeenSeq });
				}
			});
		}
		return reader;
	}
	function upload(channel, id) {
		const key = streamKey(channel, id);
		const existing = uploads.get(key);
		if (existing) return existing;
		const sink = createStreamSink({ id });
		sink.events.on("chunk", (seq, chunk) => {
			rpc.callEvent("devframe:streaming:upload-chunk", channel, id, seq, chunk);
		});
		sink.events.on("end", (error) => {
			rpc.callEvent("devframe:streaming:upload-end", channel, id, error);
			uploads.delete(key);
		});
		uploads.set(key, sink);
		return sink;
	}
	return {
		subscribe,
		upload
	};
}
//#endregion
//#region src/rpc/client.ts
function createRpcClient(functions, options) {
	const { channel, rpcOptions = {} } = options;
	return createBirpc(functions, {
		...channel,
		timeout: -1,
		...rpcOptions,
		proxify: false
	});
}
//#endregion
//#region src/rpc/transports/ws-client.ts
function NOOP() {}
const EMPTY_DEFS = /* @__PURE__ */ new Map();
/**
* Build a birpc `ChannelOptions` object backed by a browser `WebSocket`.
* Pass the result straight to `createRpcClient`'s `channel` option.
*/
function createWsRpcChannel(options) {
	let url = options.url;
	if (options.authToken) url = `${url}?devframe_auth_token=${encodeURIComponent(options.authToken)}`;
	const ws = new WebSocket(url);
	const { onConnected = NOOP, onError = NOOP, onDisconnected = NOOP, definitions = EMPTY_DEFS } = options;
	ws.addEventListener("open", (e) => {
		onConnected(e);
	});
	ws.addEventListener("error", (e) => {
		onError(e instanceof Error ? e : new Error(e.type));
	});
	ws.addEventListener("close", (e) => {
		onDisconnected(e);
	});
	const pendingRequestMethods = /* @__PURE__ */ new Map();
	return {
		on: (handler) => {
			ws.addEventListener("message", (e) => {
				handler(e.data);
			});
		},
		post: (data) => {
			if (ws.readyState === WebSocket.OPEN) ws.send(data);
			else {
				function handler() {
					ws.send(data);
					ws.removeEventListener("open", handler);
				}
				ws.addEventListener("open", handler);
			}
		},
		serialize: (msg) => {
			let method;
			if (msg.t === "q") method = msg.m;
			else {
				method = pendingRequestMethods.get(msg.i);
				pendingRequestMethods.delete(msg.i);
			}
			if (!(msg.t === "s" && "e" in msg) && !!method && definitions.get(method)?.jsonSerializable === true) return strictJsonStringify(msg, method ?? "");
			return `s:${structuredCloneStringify(msg)}`;
		},
		deserialize: (raw) => {
			const msg = raw.startsWith("s:") ? structuredCloneParse(raw.slice(2)) : JSON.parse(raw);
			if (msg.t === "q" && msg.i && msg.m) pendingRequestMethods.set(msg.i, msg.m);
			return msg;
		}
	};
}
//#endregion
//#region ../../node_modules/.pnpm/ua-parser-modern@0.1.1/node_modules/ua-parser-modern/dist/index.mjs
const EMPTY = "";
const UNKNOWN = "?";
const MAJOR = "major";
const MODEL = "model";
const NAME = "name";
const TYPE = "type";
const VENDOR = "vendor";
const VERSION = "version";
const ARCHITECTURE = "architecture";
const CONSOLE = "console";
const MOBILE = "mobile";
const TABLET = "tablet";
const SMARTTV = "smarttv";
const WEARABLE = "wearable";
const EMBEDDED = "embedded";
const UA_MAX_LENGTH = 500;
const AMAZON = "Amazon";
const APPLE = "Apple";
const ASUS = "ASUS";
const BLACKBERRY = "BlackBerry";
const BROWSER_LABEL = "Browser";
const CHROME = "Chrome";
const EDGE = "Edge";
const FIREFOX = "Firefox";
const GOOGLE = "Google";
const HUAWEI = "Huawei";
const LG = "LG";
const MICROSOFT = "Microsoft";
const MOTOROLA = "Motorola";
const OPERA = "Opera";
const SAMSUNG = "Samsung";
const SHARP = "Sharp";
const SONY = "Sony";
const XIAOMI = "Xiaomi";
const ZEBRA = "Zebra";
const FACEBOOK = "Facebook";
const CHROMIUM_OS = "Chromium OS";
const MAC_OS = "Mac OS";
function extend(regexes, extensions = {}) {
	const mergedRegexes = {};
	for (const key of Object.keys(regexes)) {
		const extension = extensions[key];
		if (extension && extension.length % 2 === 0) mergedRegexes[key] = [...extension, ...regexes[key]];
		else mergedRegexes[key] = regexes[key];
	}
	return mergedRegexes;
}
function enumerize(arr) {
	const enums = {};
	for (let i = 0; i < arr.length; i++) enums[arr[i].toUpperCase()] = arr[i];
	return enums;
}
function lowerize(str) {
	return str.toLowerCase();
}
function has(str1, str2) {
	return typeof str1 === "string" ? lowerize(str2).includes(lowerize(str1)) : false;
}
function majorize(version) {
	return typeof version === "string" ? version.replace(/[^\d.]/g, EMPTY).split(".")[0] : void 0;
}
function trim(str, len) {
	const result = str.replace(/^\s+/, EMPTY);
	return typeof len === "undefined" ? result : result.substring(0, len);
}
function rgxMapper(ua, arrays, augment) {
	const result = augment ?? {};
	for (let i = 0; i < arrays.length; i += 2) {
		const regex = arrays[i];
		const props = arrays[i + 1];
		for (let j = 0; j < regex.length; j++) {
			const matcher = regex[j];
			if (!matcher) break;
			const matches = matcher.exec(ua);
			if (!matches) continue;
			for (let p = 0; p < props.length; p++) {
				const match = matches[p + 1];
				const q = props[p];
				if (Array.isArray(q)) {
					const key = q[0];
					const length = q.length;
					if (length === 2) {
						const value = q[1];
						if (typeof value === "function") result[key] = value(match, void 0, result);
						else result[key] = value;
					} else if (length === 3) {
						const arg1 = q[1];
						const arg2 = q[2];
						if (typeof arg1 === "function" && !("exec" in arg1) && !("test" in arg1)) result[key] = match ? arg1(match, arg2, result) : void 0;
						else result[key] = match ? match.replace(arg1, arg2) : void 0;
					} else if (length === 4) result[key] = match ? q[3](match.replace(q[1], q[2]), void 0, result) : void 0;
				} else result[q] = match || void 0;
			}
			return result;
		}
	}
	return result;
}
function strMapper(str, map) {
	for (const i in map) {
		const value = map[i];
		if (Array.isArray(value) && value.length > 0) {
			for (let j = 0; j < value.length; j++) if (has(value[j], str)) return i === UNKNOWN ? void 0 : i;
		} else if (has(value, str)) return i === UNKNOWN ? void 0 : i;
	}
	return str;
}
const oldSafariMap = {
	"1.0": "/8",
	"1.2": "/1",
	"1.3": "/3",
	"2.0": "/412",
	"2.0.2": "/416",
	"2.0.3": "/417",
	"2.0.4": "/419",
	"?": "/"
};
const windowsVersionMap = {
	"ME": "4.90",
	"NT 3.11": "NT3.51",
	"NT 4.0": "NT4.0",
	"2000": "NT 5.0",
	"XP": ["NT 5.1", "NT 5.2"],
	"Vista": "NT 6.0",
	"7": "NT 6.1",
	"8": "NT 6.2",
	"8.1": "NT 6.3",
	"10": ["NT 6.4", "NT 10.0"],
	"RT": "ARM"
};
const regexes = {
	browser: [
		[/\b(?:crmo|crios)\/([\w.]+)/i],
		[VERSION, [NAME, "Chrome"]],
		[/edg(?:e|ios|a)?\/([\w.]+)/i],
		[VERSION, [NAME, "Edge"]],
		[
			/(opera mini)\/([-\w.]+)/i,
			/(opera [mobileta]{3,6})\b.+version\/([-\w.]+)/i,
			/(opera)(?:.+version\/|[/ ]+)([\w.]+)/i
		],
		[NAME, VERSION],
		[/opios[/ ]+([\w.]+)/i],
		[VERSION, [NAME, `${OPERA} Mini`]],
		[/\bop(?:rg)?x\/([\w.]+)/i],
		[VERSION, [NAME, `${OPERA} GX`]],
		[/\bopr\/([\w.]+)/i],
		[VERSION, [NAME, OPERA]],
		[/\bb[ai]*d(?:uhd|[ub]*[aekoprswx]{5,6})[/ ]?([\w.]+)/i],
		[VERSION, [NAME, "Baidu"]],
		[
			/(kindle)\/([\w.]+)/i,
			/(lunascape|maxthon|netfront|jasmine|blazer)[/ ]?([\w.]*)/i,
			/(avant|iemobile|slim)\s?(?:browser)?[/ ]?([\w.]*)/i,
			/(?:ms|\()(ie) ([\w.]+)/i,
			/(flock|rockmelt|midori|epiphany|silk|skyfire|bolt|iron|vivaldi|iridium|phantomjs|bowser|quark|qupzilla|falkon|rekonq|puffin|brave|whale(?!.+naver)|qqbrowserlite|qq|duckduckgo)\/([-\w.]+)/i,
			/(heytap|ovi)browser\/([\d.]+)/i,
			/(weibo)__([\d.]+)/i
		],
		[NAME, VERSION],
		[/\bddg\/([\w.]+)/i],
		[VERSION, [NAME, "DuckDuckGo"]],
		[/(?:\buc? ?browser|juc.+ucweb)[/ ]?([\w.]+)/i],
		[VERSION, [NAME, `UC${BROWSER_LABEL}`]],
		[
			/microm.+\bqbcore\/([\w.]+)/i,
			/\bqbcore\/([\w.]+).+microm/i,
			/micromessenger\/([\w.]+)/i
		],
		[VERSION, [NAME, "WeChat"]],
		[/konqueror\/([\w.]+)/i],
		[VERSION, [NAME, "Konqueror"]],
		[/trident.+rv[: ]([\w.]{1,9})\b.+like gecko/i],
		[VERSION, [NAME, "IE"]],
		[/ya(?:search)?browser\/([\w.]+)/i],
		[VERSION, [NAME, "Yandex"]],
		[/slbrowser\/([\w.]+)/i],
		[VERSION, [NAME, `Smart Lenovo ${BROWSER_LABEL}`]],
		[/(avast|avg)\/([\w.]+)/i],
		[[
			NAME,
			/(.+)/,
			`$1 Secure ${BROWSER_LABEL}`
		], VERSION],
		[/\bfocus\/([\w.]+)/i],
		[VERSION, [NAME, `${FIREFOX} Focus`]],
		[/\bopt\/([\w.]+)/i],
		[VERSION, [NAME, `${OPERA} Touch`]],
		[/coc_coc\w+\/([\w.]+)/i],
		[VERSION, [NAME, "Coc Coc"]],
		[/dolfin\/([\w.]+)/i],
		[VERSION, [NAME, "Dolphin"]],
		[/coast\/([\w.]+)/i],
		[VERSION, [NAME, `${OPERA} Coast`]],
		[/miuibrowser\/([\w.]+)/i],
		[VERSION, [NAME, `MIUI ${BROWSER_LABEL}`]],
		[/fxios\/([-\w.]+)/i],
		[VERSION, [NAME, FIREFOX]],
		[/\bqihu|(qi?ho{0,2}|360)browser/i],
		[[NAME, `360 ${BROWSER_LABEL}`]],
		[/(oculus|sailfish|huawei|vivo)browser\/([\w.]+)/i],
		[[
			NAME,
			/(.+)/,
			`$1 ${BROWSER_LABEL}`
		], VERSION],
		[/samsungbrowser\/([\w.]+)/i],
		[VERSION, [NAME, `${SAMSUNG} Internet`]],
		[/(comodo_dragon)\/([\w.]+)/i],
		[[
			NAME,
			/_/g,
			" "
		], VERSION],
		[/metasr[/ ]?([\d.]+)/i],
		[VERSION, [NAME, "Sogou Explorer"]],
		[/(sogou)mo\w+\/([\d.]+)/i],
		[[NAME, "Sogou Mobile"], VERSION],
		[
			/(electron)\/([\w.]+) safari/i,
			/(tesla)(?: qtcarbrowser|\/(20\d\d\.[-\w.]+))/i,
			/m?(qqbrowser|2345Explorer)[/ ]?([\w.]+)/i
		],
		[NAME, VERSION],
		[/(lbbrowser)/i, /\[(linkedin)app\]/i],
		[NAME],
		[/((?:fban\/fbios|fb_iab\/fb4a)(?!.+fbav)|;fbav\/([\w.]+);)/i],
		[[NAME, FACEBOOK], VERSION],
		[
			/(Klarna)\/([\w.]+)/i,
			/(kakao(?:talk|story))[/ ]([\w.]+)/i,
			/(naver)\(.*?(\d+\.[\w.]+).*\)/i,
			/safari (line)\/([\w.]+)/i,
			/\b(line)\/([\w.]+)\/iab/i,
			/(alipay)client\/([\w.]+)/i,
			/(twitter)(?:and| f.+e\/([\w.]+))/i,
			/(chromium|instagram|snapchat)[/ ]([-\w.]+)/i
		],
		[NAME, VERSION],
		[/\bgsa\/([\w.]+) .*safari\//i],
		[VERSION, [NAME, "GSA"]],
		[/musical_ly(?:.+app_?version\/|_)([\w.]+)/i],
		[VERSION, [NAME, "TikTok"]],
		[/headlesschrome(?:\/([\w.]+)| )/i],
		[VERSION, [NAME, `${CHROME} Headless`]],
		[/ wv\).+(chrome)\/([\w.]+)/i],
		[[NAME, `${CHROME} WebView`], VERSION],
		[/droid.+ version\/([\w.]+)\b.+(?:mobile safari|safari)/i],
		[VERSION, [NAME, `Android ${BROWSER_LABEL}`]],
		[/(chrome|omniweb|arora|[tizenoka]{5} ?browser)\/v?([\w.]+)/i],
		[NAME, VERSION],
		[/version\/([\w.,]+) .*mobile\/\w+ (safari)/i],
		[VERSION, [NAME, "Mobile Safari"]],
		[/version\/([\w(.|,)]+) .*(mobile ?safari|safari)/i],
		[VERSION, NAME],
		[/webkit.+?(mobile ?safari|safari)(\/[\w.]+)/i],
		[NAME, [
			VERSION,
			strMapper,
			oldSafariMap
		]],
		[/(webkit|khtml)\/([\w.]+)/i],
		[NAME, VERSION],
		[/(navigator|netscape\d?)\/([-\w.]+)/i],
		[[NAME, "Netscape"], VERSION],
		[/mobile vr; rv:([\w.]+)\).+firefox/i],
		[VERSION, [NAME, `${FIREFOX} Reality`]],
		[
			/ekiohf.+(flow)\/([\w.]+)/i,
			/(swiftfox)/i,
			/(icedragon|iceweasel|camino|chimera|fennec|maemo browser|minimo|conkeror|klar)[/ ]?([\w.+]+)/i,
			/(seamonkey|k-meleon|icecat|iceape|firebird|phoenix|palemoon|basilisk|waterfox)\/([-\w.]+)$/i,
			/(firefox)\/([\w.]+)/i,
			/(mozilla)\/([\w.]+) .+rv:.+gecko\/\d+/i,
			/(polaris|lynx|dillo|icab|doris|amaya|w3m|netsurf|sleipnir|obigo|mosaic|(?:go|ice|up)[. ]?browser)[-/ ]?v?([\w.]+)/i,
			/(links) \(([\w.]+)/i,
			/panasonic;(viera)/i
		],
		[NAME, VERSION],
		[/(cobalt)\/([\w.]+)/i],
		[NAME, [
			VERSION,
			/master.|lts./,
			""
		]]
	],
	cpu: [
		[/(amd|x(?:(?:86|64)[-_])?|wow|win)64[;)]/i],
		[[ARCHITECTURE, "amd64"]],
		[/(ia32(?=;))/i],
		[[ARCHITECTURE, lowerize]],
		[/((?:i[346]|x)86)[;)]/i],
		[[ARCHITECTURE, "ia32"]],
		[/\b(aarch64|arm(v?8e?l?|_?64))\b/i],
		[[ARCHITECTURE, "arm64"]],
		[/\b(arm(?:v[67])?ht?n?[fl]p?)\b/i],
		[[ARCHITECTURE, "armhf"]],
		[/windows (ce|mobile); ppc;/i],
		[[ARCHITECTURE, "arm"]],
		[/((?:ppc|powerpc)(?:64)?)(?: mac|;|\))/i],
		[[
			ARCHITECTURE,
			/ower/,
			EMPTY,
			lowerize
		]],
		[/(sun4\w)[;)]/i],
		[[ARCHITECTURE, "sparc"]],
		[/(avr32|ia64(?=;)|68k(?=\))|\barm(?=v(?:[1-7]|[5-7]1)l?|;|eabi)|(?=atmel )avr|(?:irix|mips|sparc)(?:64)?\b|pa-risc)/i],
		[[ARCHITECTURE, lowerize]]
	],
	device: [
		[/\b(sch-i[89]0\d|shw-m380s|sm-[ptx]\w{2,4}|gt-[pn]\d{2,4}|sgh-t8[56]9|nexus 10)/i],
		[
			MODEL,
			[VENDOR, SAMSUNG],
			[TYPE, TABLET]
		],
		[
			/\b((?:s[cgp]h|gt|sm)-\w+|sc[g-]?\d+a?|galaxy nexus)/i,
			/samsung[- ]([-\w]+)/i,
			/sec-(sgh\w+)/i
		],
		[
			MODEL,
			[VENDOR, SAMSUNG],
			[TYPE, MOBILE]
		],
		[/(?:\/|\()(ip(?:hone|od)[\w, ]*)(?:\/|;)/i],
		[
			MODEL,
			[VENDOR, APPLE],
			[TYPE, MOBILE]
		],
		[
			/\((ipad);[-\w),; ]+apple/i,
			/applecoremedia\/[\w.]+ \((ipad)/i,
			/\b(ipad)\d\d?,\d\d?[;\]].+ios/i
		],
		[
			MODEL,
			[VENDOR, APPLE],
			[TYPE, TABLET]
		],
		[/(macintosh);/i],
		[MODEL, [VENDOR, APPLE]],
		[/\b(sh-?[altvz]?\d\d[a-ekm]?)/i],
		[
			MODEL,
			[VENDOR, SHARP],
			[TYPE, MOBILE]
		],
		[/\b((?:ag[rs][23]?|bah2?|sht?|btv)-a?[lw]\d{2})\b(?!.+d\/s)/i],
		[
			MODEL,
			[VENDOR, HUAWEI],
			[TYPE, TABLET]
		],
		[/(?:huawei|honor)([-\w ]+)[;)]/i, /\b(nexus 6p|\w{2,4}e?-[atu]?[ln][\dx][0-359c][adn]?)\b(?!.+d\/s)/i],
		[
			MODEL,
			[VENDOR, HUAWEI],
			[TYPE, MOBILE]
		],
		[
			/\b(poco[\w ]+|m2\d{3}j\d\d[a-z]{2})(?: bui|\))/i,
			/\b; (\w+) build\/hm\1/i,
			/\b(hm[-_ ]?note?[_ ]?(?:\d\w)?) bui/i,
			/\b(redmi[\-_ ]?[\w ]+)(?: bui|\))/i,
			/oid[^)]+; (m?[12][0-389][01]\w{3,6}[c-y])( bui|; wv|\))/i,
			/\b(mi[-_ ]?(?:a\d|one|one[_ ]plus|note lte|max|cc)?[_ ]?\d?\w?[_ ]?(?:plus|se|lite)?)(?: bui|\))/i
		],
		[
			[
				MODEL,
				/_/g,
				" "
			],
			[VENDOR, XIAOMI],
			[TYPE, MOBILE]
		],
		[/oid[^)]+; (2\d{4}(283|rpbf)[cgl])( bui|\))/i, /\b(mi[-_ ]?pad[\w ]+)(?: bui|\))/i],
		[
			[
				MODEL,
				/_/g,
				" "
			],
			[VENDOR, XIAOMI],
			[TYPE, TABLET]
		],
		[/; (\w+) bui.+ oppo/i, /\b(cph[12]\d{3}|p(?:af|c[al]|d\w|e[ar])[mt]\d0|x9007|a101op)\b/i],
		[
			MODEL,
			[VENDOR, "OPPO"],
			[TYPE, MOBILE]
		],
		[/\b(opd2\d{3}a?) bui/i],
		[
			MODEL,
			[VENDOR, "OPPO"],
			[TYPE, TABLET]
		],
		[/vivo (\w+)(?: bui|\))/i, /\b(v[12]\d{3}\w?[at])(?: bui|;)/i],
		[
			MODEL,
			[VENDOR, "Vivo"],
			[TYPE, MOBILE]
		],
		[/\b(rmx[1-3]\d{3})(?: bui|;|\))/i],
		[
			MODEL,
			[VENDOR, "Realme"],
			[TYPE, MOBILE]
		],
		[
			/\b(milestone|droid(?:[2-4x]| (?:bionic|x2|pro|razr))?:?( 4g)?)\b[\w ]+build\//i,
			/\bmot(?:orola)?[- ](\w*)/i,
			/((?:moto[\w() ]+|xt\d{3,4}|nexus 6)(?= bui|\)))/i
		],
		[
			MODEL,
			[VENDOR, MOTOROLA],
			[TYPE, MOBILE]
		],
		[/\b(mz60\d|xoom[2 ]{0,2}) build\//i],
		[
			MODEL,
			[VENDOR, MOTOROLA],
			[TYPE, TABLET]
		],
		[/((?=lg)?[vl]k-?\d{3}) bui| 3\.[-\w; ]{10}lg?-([06cv9]{3,4})/i],
		[
			MODEL,
			[VENDOR, LG],
			[TYPE, TABLET]
		],
		[
			/(lm(?:-?f100[nv]?|-[\w.]+)(?= bui|\))|nexus [45])/i,
			/\blg[-e;/ ]+((?!browser|netcast|android tv)\w+)/i,
			/\blg-?(\w+) bui/i
		],
		[
			MODEL,
			[VENDOR, LG],
			[TYPE, MOBILE]
		],
		[/(ideatab[-\w ]+)/i, /lenovo ?(s[56]000[-\w]+|tab[\w ]+|yt[-\w]{6}|tb[-\w]{6})/i],
		[
			MODEL,
			[VENDOR, "Lenovo"],
			[TYPE, TABLET]
		],
		[/(?:maemo|nokia).*(n900|lumia \d+)/i, /nokia[-_ ]?([-\w.]*)/i],
		[
			[
				MODEL,
				/_/g,
				" "
			],
			[VENDOR, "Nokia"],
			[TYPE, MOBILE]
		],
		[/(pixel c)\b/i],
		[
			MODEL,
			[VENDOR, GOOGLE],
			[TYPE, TABLET]
		],
		[/droid.+; (pixel[\daxl ]{0,6})(?: bui|\))/i],
		[
			MODEL,
			[VENDOR, GOOGLE],
			[TYPE, MOBILE]
		],
		[/droid.+ (a?\d[0-2]{2}so|[c-g]\d{4}|so[-gl]\w+|xq-a\w[4-7][12])(?= bui|\).+chrome\/(?![1-6]?\d\.))/i],
		[
			MODEL,
			[VENDOR, SONY],
			[TYPE, MOBILE]
		],
		[/sony tablet [ps]/i, /\b(?:sony)?sgp\w+(?: bui|\))/i],
		[
			[MODEL, "Xperia Tablet"],
			[VENDOR, SONY],
			[TYPE, TABLET]
		],
		[/ (kb2005|in20[12]5|be20[12][59])\b/i, /(?:one)?(?:plus)? (a\d0\d\d)(?: b|\))/i],
		[
			MODEL,
			[VENDOR, "OnePlus"],
			[TYPE, MOBILE]
		],
		[
			/(alexa)webm/i,
			/(kf[a-z]{2}wi|aeo[c-r]{2})( bui|\))/i,
			/(kf[a-z]+)( bui|\)).+silk\//i
		],
		[
			MODEL,
			[VENDOR, AMAZON],
			[TYPE, TABLET]
		],
		[/((?:sd|kf)[0349hijor-uw]+)( bui|\)).+silk\//i],
		[
			[
				MODEL,
				/(.+)/g,
				"Fire Phone $1"
			],
			[VENDOR, AMAZON],
			[TYPE, MOBILE]
		],
		[/(playbook);[-\w),; ]+(rim)/i],
		[
			MODEL,
			VENDOR,
			[TYPE, TABLET]
		],
		[/\b((?:bb[a-f]|st[hv])100-\d)/i, /\(bb10; (\w+)/i],
		[
			MODEL,
			[VENDOR, BLACKBERRY],
			[TYPE, MOBILE]
		],
		[/(?:\b|asus_)(transfo[prime ]{4,10} \w+|eeepc|slider \w+|nexus 7|padfone|p00[cj])/i],
		[
			MODEL,
			[VENDOR, ASUS],
			[TYPE, TABLET]
		],
		[/ (z[bes]6[027][012][km][ls]|zenfone \d\w?)\b/i],
		[
			MODEL,
			[VENDOR, ASUS],
			[TYPE, MOBILE]
		],
		[/(nexus 9)/i],
		[
			MODEL,
			[VENDOR, "HTC"],
			[TYPE, TABLET]
		],
		[
			/(htc)[-;_ ]{1,2}([\w ]+(?=\)| bui)|\w+)/i,
			/(zte)[- ]([\w ]+?)(?: bui|\/|\))/i,
			/(alcatel|geeksphone|nexian|panasonic(?!;|\.)|sony(?!-bra))[-_ ]?([-\w]*)/i
		],
		[
			VENDOR,
			[
				MODEL,
				/_/g,
				" "
			],
			[TYPE, MOBILE]
		],
		[/droid.+; ([ab][1-7]-?[0178a]\d\d?)/i],
		[
			MODEL,
			[VENDOR, "Acer"],
			[TYPE, TABLET]
		],
		[/droid.+; (m[1-5] note) bui/i, /\bmz-([-\w]{2,})/i],
		[
			MODEL,
			[VENDOR, "Meizu"],
			[TYPE, MOBILE]
		],
		[/; ((?:power )?armor[\w ]{0,8})(?: bui|\))/i],
		[
			MODEL,
			[VENDOR, "Ulefone"],
			[TYPE, MOBILE]
		],
		[
			/(blackberry|benq|palm(?=-)|sonyericsson|acer|asus|dell|meizu|motorola|polytron|infinix|tecno)[-_ ]?([-\w]*)/i,
			/(hp) ([\w ]+\w)/i,
			/(asus)-?(\w+)/i,
			/(microsoft); (lumia[\w ]+)/i,
			/(lenovo)[-_ ]?([-\w]+)/i,
			/(jolla)/i,
			/(oppo) ?([\w ]+) bui/i
		],
		[
			VENDOR,
			MODEL,
			[TYPE, MOBILE]
		],
		[
			/(kobo)\s(ereader|touch)/i,
			/(archos) (gamepad2?)/i,
			/(hp).+(touchpad(?!.+tablet)|tablet)/i,
			/(kindle)\/([\w.]+)/i,
			/(nook)[\w ]+build\/(\w+)/i,
			/(dell) (strea[kpr\d ]*[\dko])/i,
			/(le[- ]+pan)[- ]+(\w{1,9}) bui/i,
			/(trinity)[- ]*(t\d{3}) bui/i,
			/(gigaset)[- ]+(q\w{1,9}) bui/i,
			/(vodafone) ([\w ]+)(?:\)| bui)/i
		],
		[
			VENDOR,
			MODEL,
			[TYPE, TABLET]
		],
		[/(surface duo)/i],
		[
			MODEL,
			[VENDOR, MICROSOFT],
			[TYPE, TABLET]
		],
		[/droid [\d.]+; (fp\du?)(?: b|\))/i],
		[
			MODEL,
			[VENDOR, "Fairphone"],
			[TYPE, MOBILE]
		],
		[/(u304aa)/i],
		[
			MODEL,
			[VENDOR, "AT&T"],
			[TYPE, MOBILE]
		],
		[/\bsie-(\w*)/i],
		[
			MODEL,
			[VENDOR, "Siemens"],
			[TYPE, MOBILE]
		],
		[/\b(rct\w+) b/i],
		[
			MODEL,
			[VENDOR, "RCA"],
			[TYPE, TABLET]
		],
		[/\b(venue[\d ]{2,7}) b/i],
		[
			MODEL,
			[VENDOR, "Dell"],
			[TYPE, TABLET]
		],
		[/\b(q(?:mv|ta)\w+) b/i],
		[
			MODEL,
			[VENDOR, "Verizon"],
			[TYPE, TABLET]
		],
		[/\b(?:barnes[& ]+noble |bn[rt])([\w+ ]*) b/i],
		[
			MODEL,
			[VENDOR, "Barnes & Noble"],
			[TYPE, TABLET]
		],
		[/\b(tm\d{3}\w+) b/i],
		[
			MODEL,
			[VENDOR, "NuVision"],
			[TYPE, TABLET]
		],
		[/\b(k88) b/i],
		[
			MODEL,
			[VENDOR, "ZTE"],
			[TYPE, TABLET]
		],
		[/\b(nx\d{3}j) b/i],
		[
			MODEL,
			[VENDOR, "ZTE"],
			[TYPE, MOBILE]
		],
		[/\b(gen\d{3}) b.+49h/i],
		[
			MODEL,
			[VENDOR, "Swiss"],
			[TYPE, MOBILE]
		],
		[/\b(zur\d{3}) b/i],
		[
			MODEL,
			[VENDOR, "Swiss"],
			[TYPE, TABLET]
		],
		[/\b((zeki)?tb.*\b) b/i],
		[
			MODEL,
			[VENDOR, "Zeki"],
			[TYPE, TABLET]
		],
		[/\b([yr]\d{2}) b/i, /\b(dragon[- ]+touch |dt)(\w{5}) b/i],
		[
			[VENDOR, "Dragon Touch"],
			MODEL,
			[TYPE, TABLET]
		],
		[/\b(ns-?\w{0,9}) b/i],
		[
			MODEL,
			[VENDOR, "Insignia"],
			[TYPE, TABLET]
		],
		[/\b((nxa|next)-?\w{0,9}) b/i],
		[
			MODEL,
			[VENDOR, "NextBook"],
			[TYPE, TABLET]
		],
		[/\b(xtreme_)?(v(1[045]|2[015]|[3469]0|7[05])) b/i],
		[
			[VENDOR, "Voice"],
			MODEL,
			[TYPE, MOBILE]
		],
		[/\b(lvtel-)?(v1[12]) b/i],
		[
			[VENDOR, "LvTel"],
			MODEL,
			[TYPE, MOBILE]
		],
		[/\b(ph-1) /i],
		[
			MODEL,
			[VENDOR, "Essential"],
			[TYPE, MOBILE]
		],
		[/\b(v(100md|700na|7011|917g).*\b) b/i],
		[
			MODEL,
			[VENDOR, "Envizen"],
			[TYPE, TABLET]
		],
		[/\b(trio[-\w. ]+) b/i],
		[
			MODEL,
			[VENDOR, "MachSpeed"],
			[TYPE, TABLET]
		],
		[/\btu_(1491) b/i],
		[
			MODEL,
			[VENDOR, "Rotor"],
			[TYPE, TABLET]
		],
		[/(shield[\w ]+) b/i],
		[
			MODEL,
			[VENDOR, "Nvidia"],
			[TYPE, TABLET]
		],
		[/(sprint) (\w+)/i],
		[
			VENDOR,
			MODEL,
			[TYPE, MOBILE]
		],
		[/(kin\.[onetw]{3})/i],
		[
			[
				MODEL,
				/\./g,
				" "
			],
			[VENDOR, MICROSOFT],
			[TYPE, MOBILE]
		],
		[/droid.+; (cc6666?|et5[16]|mc[239][23]x?|vc8[03]x?)\)/i],
		[
			MODEL,
			[VENDOR, ZEBRA],
			[TYPE, TABLET]
		],
		[/droid.+; (ec30|ps20|tc[2-8]\d[kx])\)/i],
		[
			MODEL,
			[VENDOR, ZEBRA],
			[TYPE, MOBILE]
		],
		[/smart-tv.+(samsung)/i],
		[VENDOR, [TYPE, SMARTTV]],
		[/hbbtv.+maple;(\d+)/i],
		[
			[
				MODEL,
				/^/,
				"SmartTV"
			],
			[VENDOR, SAMSUNG],
			[TYPE, SMARTTV]
		],
		[/(nux; netcast.+smarttv|lg (netcast\.tv-201\d|android tv))/i],
		[[VENDOR, LG], [TYPE, SMARTTV]],
		[/(apple) ?tv/i],
		[
			VENDOR,
			[MODEL, `${APPLE} TV`],
			[TYPE, SMARTTV]
		],
		[/crkey/i],
		[
			[MODEL, `${CHROME}cast`],
			[VENDOR, GOOGLE],
			[TYPE, SMARTTV]
		],
		[/droid.+aft(\w+)( bui|\))/i],
		[
			MODEL,
			[VENDOR, AMAZON],
			[TYPE, SMARTTV]
		],
		[/\(dtv[);].+(aquos)/i, /(aquos-tv[\w ]+)\)/i],
		[
			MODEL,
			[VENDOR, SHARP],
			[TYPE, SMARTTV]
		],
		[/(bravia[\w ]+)( bui|\))/i],
		[
			MODEL,
			[VENDOR, SONY],
			[TYPE, SMARTTV]
		],
		[/(mitv-\w{5}) bui/i],
		[
			MODEL,
			[VENDOR, XIAOMI],
			[TYPE, SMARTTV]
		],
		[/Hbbtv.*(technisat) (.*);/i],
		[
			VENDOR,
			MODEL,
			[TYPE, SMARTTV]
		],
		[/\b(roku)[\dx]*[)/]((?:dvp-)?[\d.]*)/i, /hbbtv\/\d+\.\d+\.\d+ +\([\w+ ]*; *(\w[^;]*);([^;]*)/i],
		[
			[VENDOR, trim],
			[MODEL, trim],
			[TYPE, SMARTTV]
		],
		[/\b(android tv|smart[- ]?tv|opera tv|tv; rv:)\b/i],
		[[TYPE, SMARTTV]],
		[/(ouya)/i, /(nintendo) ([wids3utch]+)/i],
		[
			VENDOR,
			MODEL,
			[TYPE, CONSOLE]
		],
		[/droid.+; (shield) bui/i],
		[
			MODEL,
			[VENDOR, "Nvidia"],
			[TYPE, CONSOLE]
		],
		[/(playstation [345portablevi]+)/i],
		[
			MODEL,
			[VENDOR, SONY],
			[TYPE, CONSOLE]
		],
		[/\b(xbox(?: one)?(?!; xbox))[); ]/i],
		[
			MODEL,
			[VENDOR, MICROSOFT],
			[TYPE, CONSOLE]
		],
		[/((pebble))app/i],
		[
			VENDOR,
			MODEL,
			[TYPE, WEARABLE]
		],
		[/(watch)(?: ?os[,/]|\d,\d\/)[\d.]+/i],
		[
			MODEL,
			[VENDOR, APPLE],
			[TYPE, WEARABLE]
		],
		[/droid.+; (glass) \d/i],
		[
			MODEL,
			[VENDOR, GOOGLE],
			[TYPE, WEARABLE]
		],
		[/droid.+; (wt63?0{2,3})\)/i],
		[
			MODEL,
			[VENDOR, ZEBRA],
			[TYPE, WEARABLE]
		],
		[/(quest( \d| pro)?)/i],
		[
			MODEL,
			[VENDOR, FACEBOOK],
			[TYPE, WEARABLE]
		],
		[/(tesla)(?: qtcarbrowser|\/[-\w.]+)/i],
		[VENDOR, [TYPE, EMBEDDED]],
		[/(aeobc)\b/i],
		[
			MODEL,
			[VENDOR, AMAZON],
			[TYPE, EMBEDDED]
		],
		[/droid .+?; ([^;]+?)(?: bui|; wv\)|\) applew).+? mobile safari/i],
		[MODEL, [TYPE, MOBILE]],
		[/droid .+?; ([^;]+?)(?: bui|\) applew).+?(?! mobile) safari/i],
		[MODEL, [TYPE, TABLET]],
		[/\b((tablet|tab)[;/]|focus\/\d(?!.+mobile))/i],
		[[TYPE, TABLET]],
		[/(phone|mobile(?:[;/]| [ \w/.]*safari)|pda(?=.+windows ce))/i],
		[[TYPE, MOBILE]],
		[/(android[-\w. ]{0,9});.+buil/i],
		[MODEL, [VENDOR, "Generic"]]
	],
	engine: [
		[/windows.+ edge\/([\w.]+)/i],
		[VERSION, [NAME, `${EDGE}HTML`]],
		[/webkit\/537\.36.+chrome\/(?!27)([\w.]+)/i],
		[VERSION, [NAME, "Blink"]],
		[
			/(presto)\/([\w.]+)/i,
			/(webkit|trident|netfront|netsurf|amaya|lynx|w3m|goanna)\/([\w.]+)/i,
			/ekioh(flow)\/([\w.]+)/i,
			/(khtml|tasman|links)[/ ]\(?([\w.]+)/i,
			/(icab)[/ ]([23]\.[\d.]+)/i,
			/\b(libweb)/i
		],
		[NAME, VERSION],
		[/rv:([\w.]{1,9})\b.+(gecko)/i],
		[VERSION, NAME]
	],
	os: [
		[/microsoft (windows) (vista|xp)/i],
		[NAME, VERSION],
		[/(windows (?:phone(?: os)?|mobile))[/ ]?([.\w ]*)/i],
		[NAME, [
			VERSION,
			strMapper,
			windowsVersionMap
		]],
		[
			/windows nt 6\.2; (arm)/i,
			/windows[/ ]?([ntce\d. ]+\w)(?!.+xbox)/i,
			/(?:win(?=[39n])|win 9x )([nt\d.]+)/i
		],
		[[
			VERSION,
			strMapper,
			windowsVersionMap
		], [NAME, "Windows"]],
		[
			/ip[honead]{2,4}\b(?:.*os (\w+) like mac|; opera)/i,
			/(?:ios;fbsv\/|iphone.+ios[/ ])([\d.]+)/i,
			/cfnetwork\/.+darwin/i
		],
		[[
			VERSION,
			/_/g,
			"."
		], [NAME, "iOS"]],
		[/(mac os x) ?([\w. ]*)/i, /(macintosh|mac_powerpc\b)(?!.+haiku)/i],
		[[NAME, MAC_OS], [
			VERSION,
			/_/g,
			"."
		]],
		[/droid ([\w.]+)\b.+(android[- ]x86|harmonyos)/i],
		[VERSION, NAME],
		[
			/(android|webos|qnx|bada|rim tablet os|maemo|meego|sailfish)[-/ ]?([\w.]*)/i,
			/(blackberry)\w*\/([\w.]*)/i,
			/(tizen|kaios)[/ ]([\w.]+)/i,
			/\((series40);/i
		],
		[NAME, VERSION],
		[/\(bb(10);/i],
		[VERSION, [NAME, BLACKBERRY]],
		[/(?:symbian ?os|symbos|s60(?=;)|series60)[-/ ]?([\w.]*)/i],
		[VERSION, [NAME, "Symbian"]],
		[/mozilla\/[\d.]+ \((?:mobile|tablet|tv|mobile; [\w ]+); rv:.+ gecko\/([\w.]+)/i],
		[VERSION, [NAME, `${FIREFOX} OS`]],
		[/web0s;.+rt(tv)/i, /\b(?:hp)?wos(?:browser)?\/([\w.]+)/i],
		[VERSION, [NAME, "webOS"]],
		[/watch(?: ?os[,/]|\d,\d\/)([\d.]+)/i],
		[VERSION, [NAME, "watchOS"]],
		[/crkey\/([\d.]+)/i],
		[VERSION, [NAME, `${CHROME}cast`]],
		[/(cros) \w+(?:\)| ([\w.]+)\b)/i],
		[[NAME, CHROMIUM_OS], VERSION],
		[
			/panasonic;(viera)/i,
			/(netrange)mmh/i,
			/(nettv)\/(\d+\.[\w.]+)/i,
			/(nintendo|playstation) ([wids345portablevuch]+)/i,
			/(xbox); +xbox ([^);]+)/i,
			/\b(joli|palm)\b ?(?:os)?\/?([\w.]*)/i,
			/(mint)[/() ]?(\w*)/i,
			/(mageia|vectorlinux)[; ]/i,
			/([kxln]?ubuntu|debian|suse|opensuse|gentoo|arch(?= linux)|slackware|fedora|mandriva|centos|pclinuxos|red ?hat|zenwalk|linpus|raspbian|plan 9|minix|risc os|contiki|deepin|manjaro|elementary os|sabayon|linspire)(?: gnu\/linux)?(?: enterprise)?(?:[- ]linux)?(?:-gnu)?[-/ ]?(?!chrom|package)([-\w.]*)/i,
			/(hurd|linux) ?([\w.]*)/i,
			/(gnu) ?([\w.]*)/i,
			/\b([-e-hrntopcs]{0,5}bsd|dragonfly)[/ ]?(?!amd|[ix346]{1,2}86)([\w.]*)/i,
			/(haiku) (\w+)/i
		],
		[NAME, VERSION],
		[/(sunos) ?([\w.]*)/i],
		[[NAME, "Solaris"], VERSION],
		[
			/((?:open)?solaris)[-/ ]?([\w.]*)/i,
			/(aix) ((\d)(?=[.) ])[\w.])*/i,
			/\b(beos|os\/2|amigaos|morphos|openvms|fuchsia|hp-ux|serenityos)/i,
			/(unix) ?([\w.]*)/i
		],
		[NAME, VERSION]
	]
};
function isExtensionsInput(input) {
	return typeof input === "object" && input !== null;
}
Object.freeze(enumerize([
	NAME,
	VERSION,
	MAJOR
]));
Object.freeze(enumerize([ARCHITECTURE]));
Object.freeze(enumerize([
	MODEL,
	VENDOR,
	TYPE,
	CONSOLE,
	MOBILE,
	SMARTTV,
	TABLET,
	WEARABLE,
	EMBEDDED
]));
Object.freeze(enumerize([NAME, VERSION]));
Object.freeze(enumerize([NAME, VERSION]));
function getRuntimeNavigator() {
	if (typeof window === "undefined" || !window.navigator) return void 0;
	return window.navigator;
}
function normalizeUA(ua) {
	return ua.length > UA_MAX_LENGTH ? trim(ua, UA_MAX_LENGTH) : ua;
}
function createDefaultParserContext(uastring) {
	if (typeof uastring === "string") return {
		ua: normalizeUA(uastring),
		isSelfNavigator: false,
		navigator: void 0,
		userAgentData: void 0,
		regexMap: regexes
	};
	const navigator = getRuntimeNavigator();
	const ua = normalizeUA(navigator && navigator.userAgent ? navigator.userAgent : EMPTY);
	return {
		ua,
		isSelfNavigator: Boolean(navigator && navigator.userAgent === ua),
		navigator,
		userAgentData: navigator?.userAgentData,
		regexMap: regexes
	};
}
function createParserContext(uastringOrExtensions, extensions) {
	if (!extensions && !isExtensionsInput(uastringOrExtensions)) return createDefaultParserContext(uastringOrExtensions);
	const navigator = getRuntimeNavigator();
	const uastring = isExtensionsInput(uastringOrExtensions) ? void 0 : uastringOrExtensions;
	const resolvedExtensions = isExtensionsInput(uastringOrExtensions) ? uastringOrExtensions : extensions;
	const ua = normalizeUA(uastring || (navigator && navigator.userAgent ? navigator.userAgent : EMPTY));
	const userAgentData = navigator?.userAgentData;
	return {
		ua,
		isSelfNavigator: Boolean(navigator && navigator.userAgent === ua),
		navigator,
		userAgentData,
		regexMap: resolvedExtensions ? extend(regexes, resolvedExtensions) : regexes
	};
}
function parseBrowserFromContext(context) {
	const browser = {
		name: void 0,
		version: void 0,
		major: void 0
	};
	rgxMapper(context.ua, context.regexMap.browser, browser);
	browser.major = majorize(browser.version);
	if (context.isSelfNavigator && context.navigator?.brave && typeof context.navigator.brave.isBrave === "function") browser.name = "Brave";
	return browser;
}
function parseCPUFromContext(context) {
	const cpu = { architecture: void 0 };
	rgxMapper(context.ua, context.regexMap.cpu, cpu);
	return cpu;
}
function parseDeviceFromContext(context) {
	const device = {
		vendor: void 0,
		model: void 0,
		type: void 0
	};
	rgxMapper(context.ua, context.regexMap.device, device);
	if (context.isSelfNavigator && !device.type && context.userAgentData?.mobile) device.type = MOBILE;
	if (context.isSelfNavigator && device.model === "Macintosh" && context.navigator && typeof context.navigator.standalone !== "undefined" && context.navigator.maxTouchPoints && context.navigator.maxTouchPoints > 2) {
		device.model = "iPad";
		device.type = TABLET;
	}
	return device;
}
function parseEngineFromContext(context) {
	const engine = {
		name: void 0,
		version: void 0
	};
	rgxMapper(context.ua, context.regexMap.engine, engine);
	return engine;
}
function parseOSFromContext(context) {
	const os = {
		name: void 0,
		version: void 0
	};
	rgxMapper(context.ua, context.regexMap.os, os);
	if (context.isSelfNavigator && !os.name && context.userAgentData?.platform && context.userAgentData.platform !== "Unknown") os.name = context.userAgentData.platform.replace(/chrome os/i, CHROMIUM_OS).replace(/macos/i, MAC_OS);
	return os;
}
function parseUA(uastringOrExtensions, extensions) {
	let context;
	let browserCache;
	let engineCache;
	let osCache;
	let deviceCache;
	let cpuCache;
	const getContext = () => context ||= createParserContext(uastringOrExtensions, extensions);
	return {
		get ua() {
			return getContext().ua;
		},
		get browser() {
			browserCache ||= parseBrowserFromContext(getContext());
			return browserCache;
		},
		get engine() {
			engineCache ||= parseEngineFromContext(getContext());
			return engineCache;
		},
		get os() {
			osCache ||= parseOSFromContext(getContext());
			return osCache;
		},
		get device() {
			deviceCache ||= parseDeviceFromContext(getContext());
			return deviceCache;
		},
		get cpu() {
			cpuCache ||= parseCPUFromContext(getContext());
			return cpuCache;
		}
	};
}
//#endregion
//#region src/client/rpc-ws.ts
function isNumeric(str) {
	if (str == null) return false;
	return `${+str}` === `${str}`;
}
function createWsRpcClientMode(options) {
	const { authToken, connectionMeta, events, clientRpc, rpcOptions = {}, wsOptions = {} } = options;
	let isTrusted = false;
	const trustedPromise = promiseWithResolver();
	const url = isNumeric(connectionMeta.websocket) ? `${location.protocol.replace("http", "ws")}//${location.hostname}:${connectionMeta.websocket}` : connectionMeta.websocket;
	const definitions = /* @__PURE__ */ new Map();
	for (const name of connectionMeta.jsonSerializableMethods ?? []) definitions.set(name, { jsonSerializable: true });
	const serverRpc = createRpcClient(clientRpc.functions, {
		channel: createWsRpcChannel({
			url,
			authToken,
			definitions,
			...wsOptions
		}),
		rpcOptions
	});
	clientRpc.register({
		name: "devframe:auth:revoked",
		type: "event",
		handler: () => {
			isTrusted = false;
			events.emit("rpc:is-trusted:updated", false);
		}
	});
	let currentAuthToken = authToken;
	async function requestTrustWithToken(token) {
		currentAuthToken = token;
		const info = parseUA(navigator.userAgent);
		const ua = [
			info.browser.name,
			info.browser.version,
			"|",
			info.os.name,
			info.os.version,
			info.device.type
		].filter((i) => i).join(" ");
		const result = await serverRpc.$call("devframe:anonymous:auth", {
			authToken: token,
			ua,
			origin: location.origin
		});
		isTrusted = result.isTrusted;
		trustedPromise.resolve(isTrusted);
		events.emit("rpc:is-trusted:updated", isTrusted);
		return result.isTrusted;
	}
	async function requestTrust() {
		if (isTrusted) return true;
		return requestTrustWithToken(currentAuthToken);
	}
	async function ensureTrusted(timeout = 6e4) {
		if (isTrusted) trustedPromise.resolve(true);
		if (timeout <= 0) return trustedPromise.promise;
		let clear = () => {};
		await Promise.race([trustedPromise.promise.then(clear), new Promise((resolve, reject) => {
			const id = setTimeout(() => {
				reject(/* @__PURE__ */ new Error("[devframe] Timeout waiting for rpc to be trusted"));
			}, timeout);
			clear = () => clearTimeout(id);
		})]);
		return isTrusted;
	}
	return {
		get isTrusted() {
			return isTrusted;
		},
		requestTrust,
		requestTrustWithToken,
		ensureTrusted,
		call: (...args) => {
			return serverRpc.$call(...args);
		},
		callEvent: (...args) => {
			return serverRpc.$callEvent(...args);
		},
		callOptional: (...args) => {
			return serverRpc.$callOptional(...args);
		}
	};
}
//#endregion
//#region src/client/rpc.ts
const CONNECTION_META_KEY = "__DEVFRAME_CONNECTION_META__";
const CONNECTION_AUTH_TOKEN_KEY = "__DEVFRAME_CONNECTION_AUTH_TOKEN__";
function getConnectionAuthTokenFromWindows(userAuthToken) {
	const getters = [
		() => userAuthToken,
		() => localStorage.getItem(CONNECTION_AUTH_TOKEN_KEY),
		() => window?.[CONNECTION_AUTH_TOKEN_KEY],
		() => globalThis?.[CONNECTION_AUTH_TOKEN_KEY],
		() => parent.window?.[CONNECTION_AUTH_TOKEN_KEY]
	];
	let value;
	for (const getter of getters) try {
		value = getter();
		if (value) break;
	} catch {}
	if (!value) value = humanId();
	localStorage.setItem(CONNECTION_AUTH_TOKEN_KEY, value);
	globalThis[CONNECTION_AUTH_TOKEN_KEY] = value;
	return value;
}
function findConnectionMetaFromWindows() {
	const getters = [
		() => window?.[CONNECTION_META_KEY],
		() => globalThis?.[CONNECTION_META_KEY],
		() => parent.window?.[CONNECTION_META_KEY]
	];
	for (const getter of getters) try {
		const value = getter();
		if (value) return value;
	} catch {}
}
async function getDevframeRpcClient(options = {}) {
	const { baseURL = "./", rpcOptions = {}, cacheOptions = false } = options;
	const events = createEventEmitter();
	const bases = Array.isArray(baseURL) ? baseURL : [baseURL];
	let connectionMeta = options.connectionMeta || findConnectionMetaFromWindows();
	let resolvedBaseURL = bases[0] ?? "./";
	function normalizeBase(base) {
		return base.endsWith("/") ? base : `${base}/`;
	}
	function resolveBasePath(base, path) {
		if (/^https?:\/\//.test(path)) return path;
		if (path.startsWith("/")) return path;
		return `${normalizeBase(base)}${path}`;
	}
	if (!connectionMeta) {
		const errors = [];
		for (const base of bases) try {
			connectionMeta = await fetch(resolveBasePath(base, DEVFRAME_CONNECTION_META_FILENAME)).then((r) => r.json());
			resolvedBaseURL = base;
			globalThis[CONNECTION_META_KEY] = connectionMeta;
			break;
		} catch (e) {
			errors.push(e);
		}
		if (!connectionMeta) throw new Error(`Failed to get connection meta from ${bases.join(", ")}`, { cause: errors });
	}
	const cacheManager = new RpcCacheManager({
		functions: [],
		...typeof options.cacheOptions === "object" ? options.cacheOptions : {}
	});
	const context = { rpc: void 0 };
	const authToken = getConnectionAuthTokenFromWindows(options.authToken);
	const clientRpc = new RpcFunctionsCollectorBase(context);
	async function fetchJsonFromBases(path) {
		const candidates = [resolvedBaseURL, ...bases.filter((base) => base !== resolvedBaseURL)].filter((x) => x != null);
		const errors = [];
		for (const base of candidates) try {
			return await fetch(resolveBasePath(base, path)).then((r) => {
				if (!r.ok) throw new Error(`Failed to fetch ${path} from ${base}: ${r.status}`);
				return r.json();
			});
		} catch (error) {
			errors.push(error);
		}
		throw new Error(`Failed to load ${path} from ${candidates.join(", ")}`, { cause: errors });
	}
	const mode = connectionMeta.backend === "static" ? await createStaticRpcClientMode({ fetchJsonFromBases }) : createWsRpcClientMode({
		authToken,
		connectionMeta,
		events,
		clientRpc,
		rpcOptions: {
			...rpcOptions,
			async onRequest(req, next, resolve) {
				await rpcOptions.onRequest?.call(this, req, next, resolve);
				if (cacheOptions && cacheManager?.validate(req.m)) {
					const cached = cacheManager.cached(req.m, req.a);
					if (cached) return resolve(cached);
					else {
						const res = await next(req);
						cacheManager?.apply(req, res);
					}
				} else await next(req);
			}
		},
		wsOptions: options.wsOptions
	});
	const rpc = {
		events,
		get isTrusted() {
			return mode.isTrusted;
		},
		connectionMeta,
		ensureTrusted: mode.ensureTrusted,
		requestTrust: mode.requestTrust,
		requestTrustWithToken: async (token) => {
			localStorage.setItem(CONNECTION_AUTH_TOKEN_KEY, token);
			globalThis[CONNECTION_AUTH_TOKEN_KEY] = token;
			return mode.requestTrustWithToken(token);
		},
		call: mode.call,
		callEvent: mode.callEvent,
		callOptional: mode.callOptional,
		client: clientRpc,
		sharedState: void 0,
		streaming: void 0,
		cacheManager
	};
	rpc.sharedState = createRpcSharedStateClientHost(rpc);
	rpc.streaming = createRpcStreamingClientHost(rpc);
	context.rpc = rpc;
	mode.requestTrust();
	try {
		const bc = new BroadcastChannel("devframe-auth");
		bc.onmessage = (event) => {
			if (event.data?.type === "auth-update" && event.data.authToken) rpc.requestTrustWithToken(event.data.authToken);
		};
	} catch {}
	return rpc;
}
//#endregion
//#region src/client/index.ts
const connectDevframe = getDevframeRpcClient;
//#endregion
export { connectDevframe, createRpcStreamingClientHost, getDevframeRpcClient };
