// unhead@3.1.4/dist/index.mjs
import { S as SafeInputPlugin, F as FlatMetaPlugin } from './shared/unhead.Doc2Fa-a.mjs';
export { c as createUnhead } from './shared/unhead.CfgPMHXt.mjs';
export { u as useScript } from './shared/unhead.DCW7mbD5.mjs';
import './shared/unhead.CBkhoTGw.mjs';
import './shared/unhead.fg-0ge_u.mjs';
import './shared/unhead.CUXLLRtV.mjs';
import './shared/unhead.mB5lMBMV.mjs';
import 'hookable';

function useHead(unhead, input, options = {}) {
  return unhead.push(input || {}, options);
}
function useHeadSafe(unhead, input = {}, options = {}) {
  unhead.use(SafeInputPlugin);
  return useHead(unhead, input, Object.assign(options, { _safe: true }));
}
function useSeoMeta(unhead, input = {}, options) {
  unhead.use(FlatMetaPlugin);
  function normalize(input2) {
    if (input2._flatMeta) {
      return input2;
    }
    const { title, titleTemplate, ...meta } = input2 || {};
    return {
      title,
      titleTemplate,
      _flatMeta: meta
    };
  }
  const entry = unhead.push(normalize(input), options);
  const corePatch = entry.patch;
  if (!entry.__patched) {
    entry.patch = (input2) => corePatch(normalize(input2));
    entry.__patched = true;
  }
  return entry;
}

function defineLink(link) {
  return link;
}
function defineScript(script) {
  return script;
}

export { defineLink, defineScript, useHead, useHeadSafe, useSeoMeta };
