// unhead@3.1.4/dist/validate.mjs
export { D as DEPRECATED_PROPS, K as KNOWN_META_NAMES, a as KNOWN_META_PROPERTIES, T as TAG_PRIORITY_ALIASES, U as URL_META_KEYS, d as deferOnModuleScript, e as emptyMetaContent, f as findClosestMatch, h as headInputPredicates, l as levenshtein, m as migrationTagPredicates, n as noDeprecatedProps, b as noHtmlInTitle, c as noUnknownMeta, g as nonAbsoluteCanonical, i as numericTagPriority, p as preferDefineHelpers, j as preloadFontCrossorigin, k as preloadMissingAs, r as robotsConflict, s as scriptSrcWithContent, t as tagInputFromRuntime, o as tagPredicates, q as titleInputFromRuntime, u as twitterHandleMissingAt, v as viewportUserScalable } from './shared/unhead.WR8XDfSi.mjs';

const VALIDATION_RULE_IDS = [
  "canonical-og-url-mismatch",
  "charset-not-early",
  "defer-on-module-script",
  "deprecated-option-mode",
  "deprecated-prop-body",
  "deprecated-prop-children",
  "deprecated-prop-hid-vmid",
  "duplicate-resource-hint",
  "empty-meta-content",
  "empty-title",
  "html-in-title",
  "inline-script-size",
  "inline-style-size",
  "meta-beyond-1mb",
  "missing-alias-sorting-plugin",
  "missing-description",
  "missing-template-params-plugin",
  "missing-title",
  "non-absolute-canonical",
  "non-absolute-og-url",
  "numeric-tag-priority",
  "og-image-missing-dimensions",
  "og-missing-description",
  "og-missing-title",
  "possible-typo",
  "preconnect-missing-crossorigin",
  "prefer-define-helpers",
  "prefetch-preload-conflict",
  "preload-async-defer-conflict",
  "preload-fetchpriority-conflict",
  "preload-font-crossorigin",
  "preload-missing-as",
  "preload-not-modulepreload",
  "redundant-dns-prefetch",
  "render-blocking-script",
  "robots-conflict",
  "script-src-with-content",
  "too-many-fetchpriority-high",
  "too-many-preconnects",
  "too-many-preloads",
  "twitter-handle-missing-at",
  "unresolved-template-param",
  "viewport-user-scalable"
];

export { VALIDATION_RULE_IDS };
