// unhead@3.1.4/dist/stream/client.mjs
import { r as registerPlugin } from '../shared/unhead.CfgPMHXt.mjs';
import { a as createHooks } from '../shared/unhead.mB5lMBMV.mjs';
import 'hookable';

const DEFAULT_STREAM_KEY = "__unhead__";
function createStreamableHead(options = {}) {
  const { streamKey = DEFAULT_STREAM_KEY, ...rest } = options;
  const win = typeof window !== "undefined" ? window : void 0;
  const streamQueue = win?.[streamKey];
  const core = streamQueue?._head;
  if (!core)
    return void 0;
  if (core._wrapped)
    return core;
  const hooks = createHooks(rest.hooks);
  const coreWithDirty = core;
  const isHydrationLocked = () => streamQueue?._hydrationLocked?.() ?? false;
  const head = {
    ...coreWithDirty,
    hooks,
    use: (p) => registerPlugin(head, p),
    render() {
      return core.render();
    },
    invalidate() {
      for (const entry of core.entries.values())
        delete entry._tags;
      coreWithDirty.dirty = true;
      hooks.callHook("entries:updated", head);
    },
    push(input, _options) {
      if (isHydrationLocked()) {
        return {
          _i: -1,
          patch: () => {
          },
          dispose: () => {
          }
        };
      }
      const active = core.push(input, _options);
      const entry = core.entries.get(active._i);
      if (entry)
        entry._o = input;
      coreWithDirty.dirty = true;
      hooks.callHook("entries:updated", head);
      const coreDispose = active.dispose;
      return {
        _i: active._i,
        patch(input2) {
          if (isHydrationLocked())
            return;
          active.patch(input2);
          coreWithDirty.dirty = true;
          hooks.callHook("entries:updated", head);
        },
        dispose() {
          if (core.entries.has(active._i)) {
            coreDispose();
            head.invalidate();
          }
        }
      };
    }
  };
  head._wrapped = true;
  (rest.plugins || []).forEach((p) => registerPlugin(head, p));
  registerPlugin(head, {
    key: "client",
    hooks: {
      "entries:updated": () => {
        head.render();
      }
    }
  });
  const initialPayload = rest.document?.head.querySelector('script[id="unhead:payload"]')?.innerHTML || false;
  const initEntries = [
    initialPayload ? JSON.parse(initialPayload) : false,
    ...rest.init || []
  ];
  initEntries.forEach((e) => e && head.push(e));
  if (streamQueue)
    streamQueue._head = head;
  return head;
}

export { DEFAULT_STREAM_KEY, createStreamableHead };
