// unhead@3.1.4/dist/stream/server.mjs
import { parseHtmlForIndexes, applyHeadToHtml } from '../parser.mjs';
import { a as createHead } from '../shared/unhead.yYqx02R-.mjs';
import { DEFAULT_STREAM_KEY } from './client.mjs';
import '../shared/unhead.CfgPMHXt.mjs';
import '../shared/unhead.mB5lMBMV.mjs';
import 'hookable';
import '../shared/unhead.Cdnk2khL.mjs';
import '../shared/unhead.fg-0ge_u.mjs';

const LT_RE = /</g;
const GT_RE = />/g;
const AMP_RE = /&/g;
const VALID_STREAM_KEY_RE = /^[$_a-z][$\w]*$/i;
function assertValidStreamKey(streamKey) {
  if (typeof streamKey !== "string" || !VALID_STREAM_KEY_RE.test(streamKey)) {
    throw new Error(
      `[unhead] Invalid streamKey: must be a valid JavaScript identifier matching ${VALID_STREAM_KEY_RE}. Received: ${JSON.stringify(streamKey)}`
    );
  }
}
// @__NO_SIDE_EFFECTS__
function createStreamableHead(options = {}) {
  const { streamKey, ...rest } = options;
  if (streamKey !== void 0)
    assertValidStreamKey(streamKey);
  const head = createHead({
    ...rest,
    experimentalStreamKey: streamKey
  });
  let resolveShellReady;
  const shellReady = new Promise((resolve) => {
    resolveShellReady = resolve;
  });
  return {
    head,
    onShellReady: () => resolveShellReady(),
    shellReady
  };
}
function getStreamKey(head) {
  const key = head.resolvedOptions.experimentalStreamKey || DEFAULT_STREAM_KEY;
  assertValidStreamKey(key);
  return key;
}
function createBootstrapScript(streamKey = DEFAULT_STREAM_KEY, nonce) {
  assertValidStreamKey(streamKey);
  const nonceAttr = nonce ? ` nonce="${nonce.replace(/"/g, "&quot;")}"` : "";
  return `<script${nonceAttr}>window.${streamKey}={_q:[],push(e){this._q.push(e)}}<\/script>`;
}
function renderShell(head) {
  const result = head.render();
  head.entries.clear();
  return result;
}
function renderSSRHeadShell(head, template) {
  const ssr = head.render();
  head.entries.clear();
  const bootstrapScript = createBootstrapScript(getStreamKey(head));
  const parsed = parseHtmlForIndexes(template);
  return applyHeadToHtml(parsed, {
    htmlAttrs: ssr.htmlAttrs,
    headTags: bootstrapScript + ssr.headTags,
    bodyAttrs: ssr.bodyAttrs,
    bodyTags: ssr.bodyTags
  });
}
function renderSSRHeadSuspenseChunk(head) {
  if (!head.entries.size)
    return "";
  const streamKey = getStreamKey(head);
  const inputs = Array.from(head.entries.values(), (e) => e.input);
  head.entries.clear();
  return `window.${streamKey}.push(${safeJsonStringify(inputs)})`;
}
function safeJsonStringify(obj) {
  return JSON.stringify(obj).replace(LT_RE, "\\u003c").replace(GT_RE, "\\u003e").replace(AMP_RE, "\\u0026");
}
function wrapStream(head, stream, template, preRenderedState) {
  const encoder = new TextEncoder();
  return new ReadableStream({
    async start(controller) {
      try {
        const { shell, end } = prepareStreamingTemplate(head, template, preRenderedState);
        controller.enqueue(encoder.encode(shell));
        const reader = stream.getReader();
        try {
          while (true) {
            const { done, value } = await reader.read();
            if (done)
              break;
            controller.enqueue(value);
          }
        } finally {
          reader.releaseLock();
        }
        controller.enqueue(encoder.encode(end));
        controller.close();
      } catch (error) {
        controller.error(error);
      }
    }
  });
}
function prepareStreamingTemplate(head, template, preRenderedState) {
  const ssr = preRenderedState ?? head.render();
  if (!preRenderedState) {
    head.entries.clear();
  }
  const bootstrapScript = createBootstrapScript(getStreamKey(head));
  const parsed = parseHtmlForIndexes(template);
  const bodyEnd = parsed.indexes.bodyTagEnd;
  const bodyCloseStart = parsed.indexes.bodyCloseTagStart;
  if (bodyEnd >= 0 && bodyCloseStart >= 0) {
    const bodyInterior = template.substring(bodyEnd, bodyCloseStart);
    const markerMatch = bodyInterior.match(/<!--\s*(?:app-html|ssr-outlet)\s*-->/);
    let beforeStream;
    let afterStream;
    if (markerMatch) {
      beforeStream = bodyInterior.substring(0, markerMatch.index);
      afterStream = bodyInterior.substring(markerMatch.index + markerMatch[0].length);
    } else {
      beforeStream = "";
      afterStream = bodyInterior;
    }
    const shellPart = template.substring(0, bodyEnd) + beforeStream;
    const endPart = template.substring(bodyCloseStart);
    const shellParsed = parseHtmlForIndexes(`${shellPart}</body></html>`);
    const shell = applyHeadToHtml(shellParsed, {
      htmlAttrs: ssr.htmlAttrs,
      headTags: bootstrapScript + ssr.headTags,
      bodyAttrs: ssr.bodyAttrs,
      bodyTags: ""
    }).replace("</body></html>", "");
    return {
      shell,
      end: afterStream + ssr.bodyTags + endPart
    };
  }
  return {
    shell: applyHeadToHtml(parsed, {
      htmlAttrs: ssr.htmlAttrs,
      headTags: bootstrapScript + ssr.headTags,
      bodyAttrs: ssr.bodyAttrs,
      bodyTags: ssr.bodyTags
    }),
    end: ""
  };
}

export { createBootstrapScript, createStreamableHead, prepareStreamingTemplate, renderSSRHeadShell, renderSSRHeadSuspenseChunk, renderShell, wrapStream };
