// @shikijs/twoslash@4.2.0/dist/index.mjs
import { a as rendererClassic, c as ShikiTwoslashError, i as rendererRich, n as defaultTwoslashOptions, o as defaultCompletionIcons, r as defaultHoverInfoProcessor, s as defaultCustomTagIcons, t as createTransformerFactory } from "./core-BNNwO1ON.mjs";
import { createTwoslasher } from "twoslash";
//#region src/index.ts
/**
* Factory function to create a Shiki transformer for twoslash integrations.
*/
function transformerTwoslash(options = {}) {
	const twoslashOptions = {
		cache: options.cache,
		compilerOptions: { moduleResolution: 100 }
	};
	const tsModule = options.tsModule || options.twoslashOptions?.tsModule;
	if (tsModule) twoslashOptions.tsModule = tsModule;
	return createTransformerFactory(createTwoslasher(twoslashOptions), rendererRich(options.rendererRich))(options);
}
//#endregion
export { ShikiTwoslashError, createTransformerFactory, defaultCompletionIcons, defaultCustomTagIcons, defaultHoverInfoProcessor, defaultTwoslashOptions, rendererClassic, rendererRich, transformerTwoslash };
