// @slidev/cli@52.16.0/dist/resolver-CZx9LFrt.mjs
import { dirname, join, relative, resolve, sep } from "node:path";
import process from "node:process";
import { existsSync, lstatSync, readdirSync, readlinkSync } from "node:fs";
import { copyFile, readFile } from "node:fs/promises";
import { fileURLToPath, pathToFileURL } from "node:url";
import { parseNi, run } from "@antfu/ni";
import { ensurePrefix, slash } from "@antfu/utils";
import { underline, yellow } from "ansis";
import globalDirs from "global-directory";
import { resolvePath } from "mlly";
import prompts from "prompts";
import { resolveGlobal } from "resolve-global";
import { findClosestPkgJsonPath, findDepPkgJsonPath } from "vitefu";
//#region node/resolver.ts
const RE_PATH_SEPARATOR = /[/\\]/;
const RE_SAFE_PKG_NAME = /^(?:@[a-z0-9-~][a-z0-9-._~]*\/)?[a-z0-9-~][a-z0-9-._~]*$/;
const cliRoot = fileURLToPath(new URL("..", import.meta.url));
/**
* Detect the `node_modules/` directory the running slidev binary was
* dispatched from. The simple `findDepPkgJsonPath(name, cliRoot)` walk fails
* in pnpm v11+ global installs because `cliRoot` is the realpath inside the
* content-addressable store, with no path back to the install directory. The
* invocation path (`process.argv[1]`), however, is one of:
*
* - `{PNPM_HOME}/v11/{hash}/node_modules/@slidev/cli/bin/slidev.mjs` — pnpm
*   v11 globals dispatch through a shell shim in `{PNPM_HOME}/bin/` that exec's
*   `node` with this argv[1].
* - `{globals}/node_modules/@slidev/cli/bin/slidev.mjs` — npm/yarn globals,
*   where the bin is a fs symlink and the resolved script path keeps the
*   `node_modules` segment.
* - `./node_modules/.bin/slidev` — local project, where Node resolves the
*   `.bin` symlink before populating argv[1].
*
* Returns the trailing `node_modules` directory of that path, or `undefined`
* if `argv1` doesn't pass through one.
*
* Exported for tests.
*/
function findInvocationNodeModulesPath(argv1) {
	if (!argv1) return void 0;
	const direct = resolve(argv1);
	const segment = `${sep}node_modules${sep}`;
	const directIdx = direct.lastIndexOf(segment);
	if (directIdx >= 0) return direct.slice(0, directIdx + 1 + 12);
	let current = direct;
	for (let i = 0; i < 16; i++) {
		let stat;
		try {
			stat = lstatSync(current);
		} catch {
			return;
		}
		if (!stat.isSymbolicLink()) return void 0;
		const target = readlinkSync(current);
		current = resolve(dirname(current), target);
		const idx = current.lastIndexOf(segment);
		if (idx >= 0) return current.slice(0, idx + 1 + 12);
	}
}
/**
* Candidate `node_modules` directories to search when slidev is installed
* globally. Always includes the invocation's own `node_modules` (the install
* group `@slidev/cli` lives in), plus every sibling install group's
* `node_modules` so cross-group lookups can succeed.
*
* pnpm v11 lays globals out as `{root}/v11/{hash}/node_modules/...`; each
* `pnpm add -g <pkg>` invocation creates its own `{hash}` directory even
* when multiple packages are listed in one call. Walking the parent of the
* cli's install group lets us discover packages that landed in sibling
* groups (e.g. a theme installed alongside `@slidev/cli`).
*
* Exported for tests.
*/
function computeInvocationSearchPaths(ownNodeModules) {
	if (!ownNodeModules) return [];
	const paths = [ownNodeModules];
	const installDir = dirname(ownNodeModules);
	const globalRoot = dirname(installDir);
	if (!globalRoot || globalRoot === installDir) return paths;
	let siblings;
	try {
		siblings = readdirSync(globalRoot);
	} catch {
		return paths;
	}
	for (const sib of siblings) {
		const sibInstall = join(globalRoot, sib);
		if (sibInstall === installDir) continue;
		const sibNm = join(sibInstall, "node_modules");
		try {
			if (lstatSync(sibNm).isDirectory()) paths.push(sibNm);
		} catch {}
	}
	return paths;
}
const invocationNodeModules = findInvocationNodeModulesPath(process.argv[1]);
const invocationSearchPaths = computeInvocationSearchPaths(invocationNodeModules);
const isInstalledGlobally = {};
/**
* Resolve path for import url on Vite client side
*/
async function resolveImportUrl(id) {
	return toAtFS(await resolveImportPath(id, true));
}
function toAtFS(path) {
	return `/@fs${ensurePrefix("/", slash(path))}`;
}
async function resolveImportPath(importName, ensure = false) {
	try {
		return await resolvePath(importName, { url: import.meta.url });
	} catch {}
	if (isInstalledGlobally.value) {
		for (const nm of invocationSearchPaths) try {
			return await resolvePath(importName, { url: pathToFileURL(`${nm}${sep}`) });
		} catch {}
		try {
			return resolveGlobal(importName);
		} catch {}
	}
	if (ensure) throw new Error(`Failed to resolve package "${importName}"`);
}
async function findPkgRoot(dep, parent, ensure = false) {
	const pkgJsonPath = await findDepPkgJsonPath(dep, parent);
	const path = pkgJsonPath ? dirname(pkgJsonPath) : isInstalledGlobally.value ? await findGlobalPkgRoot(dep, false) : void 0;
	if (ensure && !path) throw new Error(`Failed to resolve package "${dep}"`);
	return path;
}
async function findGlobalPkgRoot(name, ensure = false) {
	const localPath = await findDepPkgJsonPath(name, cliRoot);
	if (localPath) return dirname(localPath);
	for (const nm of invocationSearchPaths) {
		const direct = join(nm, ...name.split("/"), "package.json");
		if (existsSync(direct)) return dirname(direct);
		const walked = await findDepPkgJsonPath(name, nm);
		if (walked) return dirname(walked);
	}
	const yarnPath = join(globalDirs.yarn.packages, name);
	if (existsSync(`${yarnPath}/package.json`)) return yarnPath;
	const npmPath = join(globalDirs.npm.packages, name);
	if (existsSync(`${npmPath}/package.json`)) return npmPath;
	if (ensure) throw new Error(`Failed to resolve global package "${name}"`);
}
async function resolveEntry(entryRaw) {
	if (!existsSync(entryRaw) && !entryRaw.endsWith(".md") && !RE_PATH_SEPARATOR.test(entryRaw)) entryRaw += ".md";
	const entry = resolve(entryRaw);
	if (!existsSync(entry)) {
		if (!process.stdin.isTTY) {
			console.error(`Entry file "${entry}" does not exist and cannot prompt for confirmation`);
			process.exit(1);
		}
		const { create } = await prompts({
			name: "create",
			type: "confirm",
			initial: "Y",
			message: `Entry file ${yellow(`"${entry}"`)} does not exist, do you want to create it?`
		});
		if (create) await copyFile(resolve(cliRoot, "template.md"), entry);
		else process.exit(0);
	}
	return slash(entry);
}
/**
* Create a resolver for theme or addon
*/
function createResolver(type, officials) {
	async function promptForInstallation(pkgName) {
		if (!process.stdin.isTTY) {
			console.error(`The ${type} "${pkgName}" was not found and cannot prompt for installation`);
			process.exit(1);
		}
		const { confirm } = await prompts({
			name: "confirm",
			initial: "Y",
			type: "confirm",
			message: `The ${type} "${pkgName}" was not found ${underline(isInstalledGlobally.value ? "globally" : "in your project")}, do you want to install it now?`
		});
		if (!confirm) process.exit(1);
		if (isInstalledGlobally.value) await run(parseNi, ["-g", pkgName]);
		else await run(parseNi, [pkgName]);
	}
	return async function(name, importer) {
		const { userRoot } = await getRoots();
		if (name === "none") return ["", null];
		if (name[0] === "/") return [name, name];
		if (name.startsWith("@/")) return [name, resolve(userRoot, name.slice(2))];
		if (name[0] === "." || name[0] !== "@" && name.includes("/")) return [name, resolve(dirname(importer), name)];
		if (!RE_SAFE_PKG_NAME.test(name)) throw new Error(`Invalid ${type} name "${name}". Only valid npm package names are allowed.`);
		{
			const possiblePkgNames = [name];
			if (!name.includes("/") && !name.startsWith("@")) possiblePkgNames.unshift(`@slidev/${type}-${name}`, `slidev-${type}-${name}`);
			for (const pkgName of possiblePkgNames) {
				const pkgRoot = await findPkgRoot(pkgName, importer);
				if (pkgRoot) return [pkgName, pkgRoot];
			}
		}
		const pkgName = officials[name] ?? (name[0] === "@" ? name : `slidev-${type}-${name}`);
		await promptForInstallation(pkgName);
		return [pkgName, await findPkgRoot(pkgName, importer, true)];
	};
}
async function getUserPkgJson(userRoot) {
	const path = resolve(userRoot, "package.json");
	if (existsSync(path)) return JSON.parse(await readFile(path, "utf-8"));
	return {};
}
async function hasWorkspacePackageJSON(root) {
	const path = join(root, "package.json");
	if (!existsSync(path)) return false;
	return !!(JSON.parse(await readFile(path, "utf-8")) || {}).workspaces;
}
function hasRootFile(root) {
	return ["pnpm-workspace.yaml"].some((file) => existsSync(join(root, file)));
}
/**
* Search up for the nearest workspace root
*/
async function searchForWorkspaceRoot(current, root = current) {
	if (hasRootFile(current)) return current;
	if (await hasWorkspacePackageJSON(current)) return current;
	const dir = dirname(current);
	if (!dir || dir === current) return root;
	return searchForWorkspaceRoot(dir, root);
}
let rootsInfo = null;
async function getRoots(entry) {
	if (rootsInfo) return rootsInfo;
	if (!entry) throw new Error("[slidev] Cannot find roots without entry");
	const userRoot = dirname(entry);
	isInstalledGlobally.value = slash(relative(userRoot, process.argv[1])).includes("/.pnpm/") || invocationNodeModules != null && slash(relative(userRoot, invocationNodeModules)).startsWith("..") || (await import("is-installed-globally")).default;
	const clientRoot = await findPkgRoot("@slidev/client", cliRoot, true);
	const closestPkgRoot = dirname(await findClosestPkgJsonPath(userRoot) || userRoot);
	rootsInfo = {
		cliRoot,
		clientRoot,
		userRoot,
		userPkgJson: await getUserPkgJson(closestPkgRoot),
		userWorkspaceRoot: await searchForWorkspaceRoot(closestPkgRoot)
	};
	return rootsInfo;
}
function resolveSourceFiles(roots, subpath, extensions = [
	".mjs",
	".js",
	".mts",
	".ts"
]) {
	const results = [];
	for (const root of roots) for (const ext of extensions) {
		const fullPath = join(root, subpath + ext);
		if (existsSync(fullPath)) {
			results.push(fullPath);
			break;
		}
	}
	return results;
}
//#endregion
export { resolveImportPath as a, toAtFS as c, resolveEntry as i, getRoots as n, resolveImportUrl as o, isInstalledGlobally as r, resolveSourceFiles as s, createResolver as t };
