// @antfu/ni@30.1.0/dist/index.mjs
import { n as __exportAll, r as __reExport } from "./chunk-CMuxRQOh.mjs";
import { C as cmdExists, D as remove, E as limitText, O as terminalLink, S as CLI_TEMP_DIR, T as formatPackageWithUrl, _ as getDefaultAgent, a as UnsupportedCommand, b as getUseSfw, c as parseNd, d as parseNr, f as parseNun, g as getConfig, h as getCatalog, i as runCli, k as writeFileSafe, l as parseNi, m as serializeCommand, n as getCliCommand, o as getCommand, p as parseNup, r as run, s as parseNa, t as src_exports, u as parseNlx, v as getGlobalAgent, w as exclude, x as detect, y as getRunAgent } from "./src-Dc2SIy_D.mjs";
export * from "package-manager-detector/constants";
export * from "package-manager-detector/commands";
__reExport(/* @__PURE__ */ __exportAll({
	CLI_TEMP_DIR: () => CLI_TEMP_DIR,
	UnsupportedCommand: () => UnsupportedCommand,
	cmdExists: () => cmdExists,
	detect: () => detect,
	exclude: () => exclude,
	formatPackageWithUrl: () => formatPackageWithUrl,
	getCatalog: () => getCatalog,
	getCliCommand: () => getCliCommand,
	getCommand: () => getCommand,
	getConfig: () => getConfig,
	getDefaultAgent: () => getDefaultAgent,
	getGlobalAgent: () => getGlobalAgent,
	getRunAgent: () => getRunAgent,
	getUseSfw: () => getUseSfw,
	limitText: () => limitText,
	parseNa: () => parseNa,
	parseNd: () => parseNd,
	parseNi: () => parseNi,
	parseNlx: () => parseNlx,
	parseNr: () => parseNr,
	parseNun: () => parseNun,
	parseNup: () => parseNup,
	remove: () => remove,
	run: () => run,
	runCli: () => runCli,
	serializeCommand: () => serializeCommand,
	terminalLink: () => terminalLink,
	writeFileSafe: () => writeFileSafe
}), src_exports);
//#endregion
export { CLI_TEMP_DIR, UnsupportedCommand, cmdExists, detect, exclude, formatPackageWithUrl, getCatalog, getCliCommand, getCommand, getConfig, getDefaultAgent, getGlobalAgent, getRunAgent, getUseSfw, limitText, parseNa, parseNd, parseNi, parseNlx, parseNr, parseNun, parseNup, remove, run, runCli, serializeCommand, terminalLink, writeFileSafe };
