// @antfu/ni@30.1.0/dist/src-Dc2SIy_D.mjs
import { a as __toESM, i as __require, n as __exportAll, r as __reExport, t as __commonJSMin } from "./chunk-CMuxRQOh.mjs";
import fs, { existsSync, promises } from "node:fs";
import path, { dirname, join, resolve } from "node:path";
import process$1 from "node:process";
import { AGENTS, detect } from "package-manager-detector";
import { INSTALL_PAGE } from "package-manager-detector/constants";
import { x } from "tinyexec";
import os from "node:os";
import { styleText } from "node:util";
//#region node_modules/.pnpm/ini@6.0.0/node_modules/ini/lib/ini.js
var require_ini = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	const { hasOwnProperty } = Object.prototype;
	const encode = (obj, opt = {}) => {
		if (typeof opt === "string") opt = { section: opt };
		opt.align = opt.align === true;
		opt.newline = opt.newline === true;
		opt.sort = opt.sort === true;
		opt.whitespace = opt.whitespace === true || opt.align === true;
		/* istanbul ignore next */
		opt.platform = opt.platform || typeof process !== "undefined" && process.platform;
		opt.bracketedArray = opt.bracketedArray !== false;
		/* istanbul ignore next */
		const eol = opt.platform === "win32" ? "\r\n" : "\n";
		const separator = opt.whitespace ? " = " : "=";
		const children = [];
		const keys = opt.sort ? Object.keys(obj).sort() : Object.keys(obj);
		let padToChars = 0;
		if (opt.align) padToChars = safe(keys.filter((k) => obj[k] === null || Array.isArray(obj[k]) || typeof obj[k] !== "object").map((k) => Array.isArray(obj[k]) ? `${k}[]` : k).concat([""]).reduce((a, b) => safe(a).length >= safe(b).length ? a : b)).length;
		let out = "";
		const arraySuffix = opt.bracketedArray ? "[]" : "";
		for (const k of keys) {
			const val = obj[k];
			if (val && Array.isArray(val)) for (const item of val) out += safe(`${k}${arraySuffix}`).padEnd(padToChars, " ") + separator + safe(item) + eol;
			else if (val && typeof val === "object") children.push(k);
			else out += safe(k).padEnd(padToChars, " ") + separator + safe(val) + eol;
		}
		if (opt.section && out.length) out = "[" + safe(opt.section) + "]" + (opt.newline ? eol + eol : eol) + out;
		for (const k of children) {
			const nk = splitSections(k, ".").join("\\.");
			const section = (opt.section ? opt.section + "." : "") + nk;
			const child = encode(obj[k], {
				...opt,
				section
			});
			if (out.length && child.length) out += eol;
			out += child;
		}
		return out;
	};
	function splitSections(str, separator) {
		var lastMatchIndex = 0;
		var lastSeparatorIndex = 0;
		var nextIndex = 0;
		var sections = [];
		do {
			nextIndex = str.indexOf(separator, lastMatchIndex);
			if (nextIndex !== -1) {
				lastMatchIndex = nextIndex + separator.length;
				if (nextIndex > 0 && str[nextIndex - 1] === "\\") continue;
				sections.push(str.slice(lastSeparatorIndex, nextIndex));
				lastSeparatorIndex = nextIndex + separator.length;
			}
		} while (nextIndex !== -1);
		sections.push(str.slice(lastSeparatorIndex));
		return sections;
	}
	const decode = (str, opt = {}) => {
		opt.bracketedArray = opt.bracketedArray !== false;
		const out = Object.create(null);
		let p = out;
		let section = null;
		const re = /^\[([^\]]*)\]\s*$|^([^=]+)(=(.*))?$/i;
		const lines = str.split(/[\r\n]+/g);
		const duplicates = {};
		for (const line of lines) {
			if (!line || line.match(/^\s*[;#]/) || line.match(/^\s*$/)) continue;
			const match = line.match(re);
			if (!match) continue;
			if (match[1] !== void 0) {
				section = unsafe(match[1]);
				if (section === "__proto__") {
					p = Object.create(null);
					continue;
				}
				p = out[section] = out[section] || Object.create(null);
				continue;
			}
			const keyRaw = unsafe(match[2]);
			let isArray;
			if (opt.bracketedArray) isArray = keyRaw.length > 2 && keyRaw.slice(-2) === "[]";
			else {
				duplicates[keyRaw] = (duplicates?.[keyRaw] || 0) + 1;
				isArray = duplicates[keyRaw] > 1;
			}
			const key = isArray && keyRaw.endsWith("[]") ? keyRaw.slice(0, -2) : keyRaw;
			if (key === "__proto__") continue;
			const valueRaw = match[3] ? unsafe(match[4]) : true;
			const value = valueRaw === "true" || valueRaw === "false" || valueRaw === "null" ? JSON.parse(valueRaw) : valueRaw;
			if (isArray) {
				if (!hasOwnProperty.call(p, key)) p[key] = [];
				else if (!Array.isArray(p[key])) p[key] = [p[key]];
			}
			if (Array.isArray(p[key])) p[key].push(value);
			else p[key] = value;
		}
		const remove = [];
		for (const k of Object.keys(out)) {
			if (!hasOwnProperty.call(out, k) || typeof out[k] !== "object" || Array.isArray(out[k])) continue;
			const parts = splitSections(k, ".");
			p = out;
			const l = parts.pop();
			const nl = l.replace(/\\\./g, ".");
			for (const part of parts) {
				if (part === "__proto__") continue;
				if (!hasOwnProperty.call(p, part) || typeof p[part] !== "object") p[part] = Object.create(null);
				p = p[part];
			}
			if (p === out && nl === l) continue;
			p[nl] = out[k];
			remove.push(k);
		}
		for (const del of remove) delete out[del];
		return out;
	};
	const isQuoted = (val) => {
		return val.startsWith("\"") && val.endsWith("\"") || val.startsWith("'") && val.endsWith("'");
	};
	const safe = (val) => {
		if (typeof val !== "string" || val.match(/[=\r\n]/) || val.match(/^\[/) || val.length > 1 && isQuoted(val) || val !== val.trim()) return JSON.stringify(val);
		return val.split(";").join("\\;").split("#").join("\\#");
	};
	const unsafe = (val) => {
		val = (val || "").trim();
		if (isQuoted(val)) {
			if (val.charAt(0) === "'") val = val.slice(1, -1);
			try {
				val = JSON.parse(val);
			} catch {}
		} else {
			let esc = false;
			let unesc = "";
			for (let i = 0, l = val.length; i < l; i++) {
				const c = val.charAt(i);
				if (esc) {
					if ("\\;#".indexOf(c) !== -1) unesc += c;
					else unesc += "\\" + c;
					esc = false;
				} else if (";#".indexOf(c) !== -1) break;
				else if (c === "\\") esc = true;
				else unesc += c;
			}
			if (esc) unesc += "\\";
			return unesc.trim();
		}
		return val;
	};
	module.exports = {
		parse: decode,
		decode,
		stringify: encode,
		encode,
		safe,
		unsafe
	};
}));
//#endregion
//#region node_modules/.pnpm/kleur@4.1.5/node_modules/kleur/index.js
var require_kleur = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	let FORCE_COLOR, NODE_DISABLE_COLORS, NO_COLOR, TERM, isTTY = true;
	if (typeof process !== "undefined") {
		({FORCE_COLOR, NODE_DISABLE_COLORS, NO_COLOR, TERM} = process.env || {});
		isTTY = process.stdout && process.stdout.isTTY;
	}
	const $ = {
		enabled: !NODE_DISABLE_COLORS && NO_COLOR == null && TERM !== "dumb" && (FORCE_COLOR != null && FORCE_COLOR !== "0" || isTTY),
		reset: init(0, 0),
		bold: init(1, 22),
		dim: init(2, 22),
		italic: init(3, 23),
		underline: init(4, 24),
		inverse: init(7, 27),
		hidden: init(8, 28),
		strikethrough: init(9, 29),
		black: init(30, 39),
		red: init(31, 39),
		green: init(32, 39),
		yellow: init(33, 39),
		blue: init(34, 39),
		magenta: init(35, 39),
		cyan: init(36, 39),
		white: init(37, 39),
		gray: init(90, 39),
		grey: init(90, 39),
		bgBlack: init(40, 49),
		bgRed: init(41, 49),
		bgGreen: init(42, 49),
		bgYellow: init(43, 49),
		bgBlue: init(44, 49),
		bgMagenta: init(45, 49),
		bgCyan: init(46, 49),
		bgWhite: init(47, 49)
	};
	function run(arr, str) {
		let i = 0, tmp, beg = "", end = "";
		for (; i < arr.length; i++) {
			tmp = arr[i];
			beg += tmp.open;
			end += tmp.close;
			if (!!~str.indexOf(tmp.close)) str = str.replace(tmp.rgx, tmp.close + tmp.open);
		}
		return beg + str + end;
	}
	function chain(has, keys) {
		let ctx = {
			has,
			keys
		};
		ctx.reset = $.reset.bind(ctx);
		ctx.bold = $.bold.bind(ctx);
		ctx.dim = $.dim.bind(ctx);
		ctx.italic = $.italic.bind(ctx);
		ctx.underline = $.underline.bind(ctx);
		ctx.inverse = $.inverse.bind(ctx);
		ctx.hidden = $.hidden.bind(ctx);
		ctx.strikethrough = $.strikethrough.bind(ctx);
		ctx.black = $.black.bind(ctx);
		ctx.red = $.red.bind(ctx);
		ctx.green = $.green.bind(ctx);
		ctx.yellow = $.yellow.bind(ctx);
		ctx.blue = $.blue.bind(ctx);
		ctx.magenta = $.magenta.bind(ctx);
		ctx.cyan = $.cyan.bind(ctx);
		ctx.white = $.white.bind(ctx);
		ctx.gray = $.gray.bind(ctx);
		ctx.grey = $.grey.bind(ctx);
		ctx.bgBlack = $.bgBlack.bind(ctx);
		ctx.bgRed = $.bgRed.bind(ctx);
		ctx.bgGreen = $.bgGreen.bind(ctx);
		ctx.bgYellow = $.bgYellow.bind(ctx);
		ctx.bgBlue = $.bgBlue.bind(ctx);
		ctx.bgMagenta = $.bgMagenta.bind(ctx);
		ctx.bgCyan = $.bgCyan.bind(ctx);
		ctx.bgWhite = $.bgWhite.bind(ctx);
		return ctx;
	}
	function init(open, close) {
		let blk = {
			open: `\x1b[${open}m`,
			close: `\x1b[${close}m`,
			rgx: new RegExp(`\\x1b\\[${close}m`, "g")
		};
		return function(txt) {
			if (this !== void 0 && this.has !== void 0) {
				~this.has.indexOf(open) || (this.has.push(open), this.keys.push(blk));
				return txt === void 0 ? this : $.enabled ? run(this.keys, txt + "") : txt + "";
			}
			return txt === void 0 ? chain([open], [blk]) : $.enabled ? run([blk], txt + "") : txt + "";
		};
	}
	module.exports = $;
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/util/action.js
var require_action = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	module.exports = (key, isSelect) => {
		if (key.meta && key.name !== "escape") return;
		if (key.ctrl) {
			if (key.name === "a") return "first";
			if (key.name === "c") return "abort";
			if (key.name === "d") return "abort";
			if (key.name === "e") return "last";
			if (key.name === "g") return "reset";
			if (key.name === "n") return "down";
			if (key.name === "p") return "up";
			return;
		}
		if (isSelect) {
			if (key.name === "j") return "down";
			if (key.name === "k") return "up";
		}
		if (key.name === "return") return "submit";
		if (key.name === "enter") return "submit";
		if (key.name === "backspace") return "delete";
		if (key.name === "delete") return "deleteForward";
		if (key.name === "abort") return "abort";
		if (key.name === "escape") return "exit";
		if (key.name === "tab") return "next";
		if (key.name === "pagedown") return "nextPage";
		if (key.name === "pageup") return "prevPage";
		if (key.name === "home") return "home";
		if (key.name === "end") return "end";
		if (key.name === "up") return "up";
		if (key.name === "down") return "down";
		if (key.name === "right") return "right";
		if (key.name === "left") return "left";
		return false;
	};
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/util/strip.js
var require_strip = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	module.exports = (str) => {
		const pattern = ["[\\u001B\\u009B][[\\]()#;?]*(?:(?:(?:(?:;[-a-zA-Z\\d\\/#&.:=?%@~_]+)*|[a-zA-Z\\d]+(?:;[-a-zA-Z\\d\\/#&.:=?%@~_]*)*)?\\u0007)", "(?:(?:\\d{1,4}(?:;\\d{0,4})*)?[\\dA-PRZcf-ntqry=><~]))"].join("|");
		const RGX = new RegExp(pattern, "g");
		return typeof str === "string" ? str.replace(RGX, "") : str;
	};
}));
//#endregion
//#region node_modules/.pnpm/sisteransi@1.0.5/node_modules/sisteransi/src/index.js
var require_src = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	const ESC = "\x1B";
	const CSI = `${ESC}[`;
	const beep = "\x07";
	const cursor = {
		to(x, y) {
			if (!y) return `${CSI}${x + 1}G`;
			return `${CSI}${y + 1};${x + 1}H`;
		},
		move(x, y) {
			let ret = "";
			if (x < 0) ret += `${CSI}${-x}D`;
			else if (x > 0) ret += `${CSI}${x}C`;
			if (y < 0) ret += `${CSI}${-y}A`;
			else if (y > 0) ret += `${CSI}${y}B`;
			return ret;
		},
		up: (count = 1) => `${CSI}${count}A`,
		down: (count = 1) => `${CSI}${count}B`,
		forward: (count = 1) => `${CSI}${count}C`,
		backward: (count = 1) => `${CSI}${count}D`,
		nextLine: (count = 1) => `${CSI}E`.repeat(count),
		prevLine: (count = 1) => `${CSI}F`.repeat(count),
		left: `${CSI}G`,
		hide: `${CSI}?25l`,
		show: `${CSI}?25h`,
		save: `${ESC}7`,
		restore: `${ESC}8`
	};
	module.exports = {
		cursor,
		scroll: {
			up: (count = 1) => `${CSI}S`.repeat(count),
			down: (count = 1) => `${CSI}T`.repeat(count)
		},
		erase: {
			screen: `${CSI}2J`,
			up: (count = 1) => `${CSI}1J`.repeat(count),
			down: (count = 1) => `${CSI}J`.repeat(count),
			line: `${CSI}2K`,
			lineEnd: `${CSI}K`,
			lineStart: `${CSI}1K`,
			lines(count) {
				let clear = "";
				for (let i = 0; i < count; i++) clear += this.line + (i < count - 1 ? cursor.up() : "");
				if (count) clear += cursor.left;
				return clear;
			}
		},
		beep
	};
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/util/clear.js
var require_clear = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	const strip = require_strip();
	const { erase, cursor } = require_src();
	const width = (str) => [...strip(str)].length;
	/**
	* @param {string} prompt
	* @param {number} perLine
	*/
	module.exports = function(prompt, perLine) {
		if (!perLine) return erase.line + cursor.to(0);
		let rows = 0;
		const lines = prompt.split(/\r?\n/);
		for (let line of lines) rows += 1 + Math.floor(Math.max(width(line) - 1, 0) / perLine);
		return erase.lines(rows);
	};
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/util/figures.js
var require_figures = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	const main = {
		arrowUp: "↑",
		arrowDown: "↓",
		arrowLeft: "←",
		arrowRight: "→",
		radioOn: "◉",
		radioOff: "◯",
		tick: "✔",
		cross: "✖",
		ellipsis: "…",
		pointerSmall: "›",
		line: "─",
		pointer: "❯"
	};
	const win = {
		arrowUp: main.arrowUp,
		arrowDown: main.arrowDown,
		arrowLeft: main.arrowLeft,
		arrowRight: main.arrowRight,
		radioOn: "(*)",
		radioOff: "( )",
		tick: "√",
		cross: "×",
		ellipsis: "...",
		pointerSmall: "»",
		line: "─",
		pointer: ">"
	};
	module.exports = process.platform === "win32" ? win : main;
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/util/style.js
var require_style = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	const c = require_kleur();
	const figures = require_figures();
	const styles = Object.freeze({
		password: {
			scale: 1,
			render: (input) => "*".repeat(input.length)
		},
		emoji: {
			scale: 2,
			render: (input) => "😃".repeat(input.length)
		},
		invisible: {
			scale: 0,
			render: (input) => ""
		},
		default: {
			scale: 1,
			render: (input) => `${input}`
		}
	});
	const render = (type) => styles[type] || styles.default;
	const symbols = Object.freeze({
		aborted: c.red(figures.cross),
		done: c.green(figures.tick),
		exited: c.yellow(figures.cross),
		default: c.cyan("?")
	});
	const symbol = (done, aborted, exited) => aborted ? symbols.aborted : exited ? symbols.exited : done ? symbols.done : symbols.default;
	const delimiter = (completing) => c.gray(completing ? figures.ellipsis : figures.pointerSmall);
	const item = (expandable, expanded) => c.gray(expandable ? expanded ? figures.pointerSmall : "+" : figures.line);
	module.exports = {
		styles,
		render,
		symbols,
		symbol,
		delimiter,
		item
	};
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/util/lines.js
var require_lines = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	const strip = require_strip();
	/**
	* @param {string} msg
	* @param {number} perLine
	*/
	module.exports = function(msg, perLine) {
		let lines = String(strip(msg) || "").split(/\r?\n/);
		if (!perLine) return lines.length;
		return lines.map((l) => Math.ceil(l.length / perLine)).reduce((a, b) => a + b);
	};
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/util/wrap.js
var require_wrap = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	/**
	* @param {string} msg The message to wrap
	* @param {object} opts
	* @param {number|string} [opts.margin] Left margin
	* @param {number} opts.width Maximum characters per line including the margin
	*/
	module.exports = (msg, opts = {}) => {
		const tab = Number.isSafeInteger(parseInt(opts.margin)) ? new Array(parseInt(opts.margin)).fill(" ").join("") : opts.margin || "";
		const width = opts.width;
		return (msg || "").split(/\r?\n/g).map((line) => line.split(/\s+/g).reduce((arr, w) => {
			if (w.length + tab.length >= width || arr[arr.length - 1].length + w.length + 1 < width) arr[arr.length - 1] += ` ${w}`;
			else arr.push(`${tab}${w}`);
			return arr;
		}, [tab]).join("\n")).join("\n");
	};
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/util/entriesToDisplay.js
var require_entriesToDisplay = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	/**
	* Determine what entries should be displayed on the screen, based on the
	* currently selected index and the maximum visible. Used in list-based
	* prompts like `select` and `multiselect`.
	*
	* @param {number} cursor the currently selected entry
	* @param {number} total the total entries available to display
	* @param {number} [maxVisible] the number of entries that can be displayed
	*/
	module.exports = (cursor, total, maxVisible) => {
		maxVisible = maxVisible || total;
		let startIndex = Math.min(total - maxVisible, cursor - Math.floor(maxVisible / 2));
		if (startIndex < 0) startIndex = 0;
		let endIndex = Math.min(startIndex + maxVisible, total);
		return {
			startIndex,
			endIndex
		};
	};
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/util/index.js
var require_util = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	module.exports = {
		action: require_action(),
		clear: require_clear(),
		style: require_style(),
		strip: require_strip(),
		figures: require_figures(),
		lines: require_lines(),
		wrap: require_wrap(),
		entriesToDisplay: require_entriesToDisplay()
	};
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/elements/prompt.js
var require_prompt = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	const readline = __require("readline");
	const { action } = require_util();
	const EventEmitter = __require("events");
	const { beep, cursor } = require_src();
	const color = require_kleur();
	/**
	* Base prompt skeleton
	* @param {Stream} [opts.stdin] The Readable stream to listen to
	* @param {Stream} [opts.stdout] The Writable stream to write readline data to
	*/
	var Prompt = class extends EventEmitter {
		constructor(opts = {}) {
			super();
			this.firstRender = true;
			this.in = opts.stdin || process.stdin;
			this.out = opts.stdout || process.stdout;
			this.onRender = (opts.onRender || (() => void 0)).bind(this);
			const rl = readline.createInterface({
				input: this.in,
				escapeCodeTimeout: 50
			});
			readline.emitKeypressEvents(this.in, rl);
			if (this.in.isTTY) this.in.setRawMode(true);
			const isSelect = ["SelectPrompt", "MultiselectPrompt"].indexOf(this.constructor.name) > -1;
			const keypress = (str, key) => {
				let a = action(key, isSelect);
				if (a === false) this._ && this._(str, key);
				else if (typeof this[a] === "function") this[a](key);
				else this.bell();
			};
			this.close = () => {
				this.out.write(cursor.show);
				this.in.removeListener("keypress", keypress);
				if (this.in.isTTY) this.in.setRawMode(false);
				rl.close();
				this.emit(this.aborted ? "abort" : this.exited ? "exit" : "submit", this.value);
				this.closed = true;
			};
			this.in.on("keypress", keypress);
		}
		fire() {
			this.emit("state", {
				value: this.value,
				aborted: !!this.aborted,
				exited: !!this.exited
			});
		}
		bell() {
			this.out.write(beep);
		}
		render() {
			this.onRender(color);
			if (this.firstRender) this.firstRender = false;
		}
	};
	module.exports = Prompt;
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/elements/text.js
var require_text = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	const color = require_kleur();
	const Prompt = require_prompt();
	const { erase, cursor } = require_src();
	const { style, clear, lines, figures } = require_util();
	/**
	* TextPrompt Base Element
	* @param {Object} opts Options
	* @param {String} opts.message Message
	* @param {String} [opts.style='default'] Render style
	* @param {String} [opts.initial] Default value
	* @param {Function} [opts.validate] Validate function
	* @param {Stream} [opts.stdin] The Readable stream to listen to
	* @param {Stream} [opts.stdout] The Writable stream to write readline data to
	* @param {String} [opts.error] The invalid error label
	*/
	var TextPrompt = class extends Prompt {
		constructor(opts = {}) {
			super(opts);
			this.transform = style.render(opts.style);
			this.scale = this.transform.scale;
			this.msg = opts.message;
			this.initial = opts.initial || ``;
			this.validator = opts.validate || (() => true);
			this.value = ``;
			this.errorMsg = opts.error || `Please Enter A Valid Value`;
			this.cursor = Number(!!this.initial);
			this.cursorOffset = 0;
			this.clear = clear(``, this.out.columns);
			this.render();
		}
		set value(v) {
			if (!v && this.initial) {
				this.placeholder = true;
				this.rendered = color.gray(this.transform.render(this.initial));
			} else {
				this.placeholder = false;
				this.rendered = this.transform.render(v);
			}
			this._value = v;
			this.fire();
		}
		get value() {
			return this._value;
		}
		reset() {
			this.value = ``;
			this.cursor = Number(!!this.initial);
			this.cursorOffset = 0;
			this.fire();
			this.render();
		}
		exit() {
			this.abort();
		}
		abort() {
			this.value = this.value || this.initial;
			this.done = this.aborted = true;
			this.error = false;
			this.red = false;
			this.fire();
			this.render();
			this.out.write("\n");
			this.close();
		}
		async validate() {
			let valid = await this.validator(this.value);
			if (typeof valid === `string`) {
				this.errorMsg = valid;
				valid = false;
			}
			this.error = !valid;
		}
		async submit() {
			this.value = this.value || this.initial;
			this.cursorOffset = 0;
			this.cursor = this.rendered.length;
			await this.validate();
			if (this.error) {
				this.red = true;
				this.fire();
				this.render();
				return;
			}
			this.done = true;
			this.aborted = false;
			this.fire();
			this.render();
			this.out.write("\n");
			this.close();
		}
		next() {
			if (!this.placeholder) return this.bell();
			this.value = this.initial;
			this.cursor = this.rendered.length;
			this.fire();
			this.render();
		}
		moveCursor(n) {
			if (this.placeholder) return;
			this.cursor = this.cursor + n;
			this.cursorOffset += n;
		}
		_(c, key) {
			let s1 = this.value.slice(0, this.cursor);
			this.value = `${s1}${c}${this.value.slice(this.cursor)}`;
			this.red = false;
			this.cursor = this.placeholder ? 0 : s1.length + 1;
			this.render();
		}
		delete() {
			if (this.isCursorAtStart()) return this.bell();
			this.value = `${this.value.slice(0, this.cursor - 1)}${this.value.slice(this.cursor)}`;
			this.red = false;
			if (this.isCursorAtStart()) this.cursorOffset = 0;
			else {
				this.cursorOffset++;
				this.moveCursor(-1);
			}
			this.render();
		}
		deleteForward() {
			if (this.cursor * this.scale >= this.rendered.length || this.placeholder) return this.bell();
			this.value = `${this.value.slice(0, this.cursor)}${this.value.slice(this.cursor + 1)}`;
			this.red = false;
			if (this.isCursorAtEnd()) this.cursorOffset = 0;
			else this.cursorOffset++;
			this.render();
		}
		first() {
			this.cursor = 0;
			this.render();
		}
		last() {
			this.cursor = this.value.length;
			this.render();
		}
		left() {
			if (this.cursor <= 0 || this.placeholder) return this.bell();
			this.moveCursor(-1);
			this.render();
		}
		right() {
			if (this.cursor * this.scale >= this.rendered.length || this.placeholder) return this.bell();
			this.moveCursor(1);
			this.render();
		}
		isCursorAtStart() {
			return this.cursor === 0 || this.placeholder && this.cursor === 1;
		}
		isCursorAtEnd() {
			return this.cursor === this.rendered.length || this.placeholder && this.cursor === this.rendered.length + 1;
		}
		render() {
			if (this.closed) return;
			if (!this.firstRender) {
				if (this.outputError) this.out.write(cursor.down(lines(this.outputError, this.out.columns) - 1) + clear(this.outputError, this.out.columns));
				this.out.write(clear(this.outputText, this.out.columns));
			}
			super.render();
			this.outputError = "";
			this.outputText = [
				style.symbol(this.done, this.aborted),
				color.bold(this.msg),
				style.delimiter(this.done),
				this.red ? color.red(this.rendered) : this.rendered
			].join(` `);
			if (this.error) this.outputError += this.errorMsg.split(`\n`).reduce((a, l, i) => a + `\n${i ? " " : figures.pointerSmall} ${color.red().italic(l)}`, ``);
			this.out.write(erase.line + cursor.to(0) + this.outputText + cursor.save + this.outputError + cursor.restore + cursor.move(this.cursorOffset, 0));
		}
	};
	module.exports = TextPrompt;
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/elements/select.js
var require_select = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	const color = require_kleur();
	const Prompt = require_prompt();
	const { style, clear, figures, wrap, entriesToDisplay } = require_util();
	const { cursor } = require_src();
	/**
	* SelectPrompt Base Element
	* @param {Object} opts Options
	* @param {String} opts.message Message
	* @param {Array} opts.choices Array of choice objects
	* @param {String} [opts.hint] Hint to display
	* @param {Number} [opts.initial] Index of default value
	* @param {Stream} [opts.stdin] The Readable stream to listen to
	* @param {Stream} [opts.stdout] The Writable stream to write readline data to
	* @param {Number} [opts.optionsPerPage=10] Max options to display at once
	*/
	var SelectPrompt = class extends Prompt {
		constructor(opts = {}) {
			super(opts);
			this.msg = opts.message;
			this.hint = opts.hint || "- Use arrow-keys. Return to submit.";
			this.warn = opts.warn || "- This option is disabled";
			this.cursor = opts.initial || 0;
			this.choices = opts.choices.map((ch, idx) => {
				if (typeof ch === "string") ch = {
					title: ch,
					value: idx
				};
				return {
					title: ch && (ch.title || ch.value || ch),
					value: ch && (ch.value === void 0 ? idx : ch.value),
					description: ch && ch.description,
					selected: ch && ch.selected,
					disabled: ch && ch.disabled
				};
			});
			this.optionsPerPage = opts.optionsPerPage || 10;
			this.value = (this.choices[this.cursor] || {}).value;
			this.clear = clear("", this.out.columns);
			this.render();
		}
		moveCursor(n) {
			this.cursor = n;
			this.value = this.choices[n].value;
			this.fire();
		}
		reset() {
			this.moveCursor(0);
			this.fire();
			this.render();
		}
		exit() {
			this.abort();
		}
		abort() {
			this.done = this.aborted = true;
			this.fire();
			this.render();
			this.out.write("\n");
			this.close();
		}
		submit() {
			if (!this.selection.disabled) {
				this.done = true;
				this.aborted = false;
				this.fire();
				this.render();
				this.out.write("\n");
				this.close();
			} else this.bell();
		}
		first() {
			this.moveCursor(0);
			this.render();
		}
		last() {
			this.moveCursor(this.choices.length - 1);
			this.render();
		}
		up() {
			if (this.cursor === 0) this.moveCursor(this.choices.length - 1);
			else this.moveCursor(this.cursor - 1);
			this.render();
		}
		down() {
			if (this.cursor === this.choices.length - 1) this.moveCursor(0);
			else this.moveCursor(this.cursor + 1);
			this.render();
		}
		next() {
			this.moveCursor((this.cursor + 1) % this.choices.length);
			this.render();
		}
		_(c, key) {
			if (c === " ") return this.submit();
		}
		get selection() {
			return this.choices[this.cursor];
		}
		render() {
			if (this.closed) return;
			if (this.firstRender) this.out.write(cursor.hide);
			else this.out.write(clear(this.outputText, this.out.columns));
			super.render();
			let { startIndex, endIndex } = entriesToDisplay(this.cursor, this.choices.length, this.optionsPerPage);
			this.outputText = [
				style.symbol(this.done, this.aborted),
				color.bold(this.msg),
				style.delimiter(false),
				this.done ? this.selection.title : this.selection.disabled ? color.yellow(this.warn) : color.gray(this.hint)
			].join(" ");
			if (!this.done) {
				this.outputText += "\n";
				for (let i = startIndex; i < endIndex; i++) {
					let title, prefix, desc = "", v = this.choices[i];
					if (i === startIndex && startIndex > 0) prefix = figures.arrowUp;
					else if (i === endIndex - 1 && endIndex < this.choices.length) prefix = figures.arrowDown;
					else prefix = " ";
					if (v.disabled) {
						title = this.cursor === i ? color.gray().underline(v.title) : color.strikethrough().gray(v.title);
						prefix = (this.cursor === i ? color.bold().gray(figures.pointer) + " " : "  ") + prefix;
					} else {
						title = this.cursor === i ? color.cyan().underline(v.title) : v.title;
						prefix = (this.cursor === i ? color.cyan(figures.pointer) + " " : "  ") + prefix;
						if (v.description && this.cursor === i) {
							desc = ` - ${v.description}`;
							if (prefix.length + title.length + desc.length >= this.out.columns || v.description.split(/\r?\n/).length > 1) desc = "\n" + wrap(v.description, {
								margin: 3,
								width: this.out.columns
							});
						}
					}
					this.outputText += `${prefix} ${title}${color.gray(desc)}\n`;
				}
			}
			this.out.write(this.outputText);
		}
	};
	module.exports = SelectPrompt;
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/elements/toggle.js
var require_toggle = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	const color = require_kleur();
	const Prompt = require_prompt();
	const { style, clear } = require_util();
	const { cursor, erase } = require_src();
	/**
	* TogglePrompt Base Element
	* @param {Object} opts Options
	* @param {String} opts.message Message
	* @param {Boolean} [opts.initial=false] Default value
	* @param {String} [opts.active='no'] Active label
	* @param {String} [opts.inactive='off'] Inactive label
	* @param {Stream} [opts.stdin] The Readable stream to listen to
	* @param {Stream} [opts.stdout] The Writable stream to write readline data to
	*/
	var TogglePrompt = class extends Prompt {
		constructor(opts = {}) {
			super(opts);
			this.msg = opts.message;
			this.value = !!opts.initial;
			this.active = opts.active || "on";
			this.inactive = opts.inactive || "off";
			this.initialValue = this.value;
			this.render();
		}
		reset() {
			this.value = this.initialValue;
			this.fire();
			this.render();
		}
		exit() {
			this.abort();
		}
		abort() {
			this.done = this.aborted = true;
			this.fire();
			this.render();
			this.out.write("\n");
			this.close();
		}
		submit() {
			this.done = true;
			this.aborted = false;
			this.fire();
			this.render();
			this.out.write("\n");
			this.close();
		}
		deactivate() {
			if (this.value === false) return this.bell();
			this.value = false;
			this.render();
		}
		activate() {
			if (this.value === true) return this.bell();
			this.value = true;
			this.render();
		}
		delete() {
			this.deactivate();
		}
		left() {
			this.deactivate();
		}
		right() {
			this.activate();
		}
		down() {
			this.deactivate();
		}
		up() {
			this.activate();
		}
		next() {
			this.value = !this.value;
			this.fire();
			this.render();
		}
		_(c, key) {
			if (c === " ") this.value = !this.value;
			else if (c === "1") this.value = true;
			else if (c === "0") this.value = false;
			else return this.bell();
			this.render();
		}
		render() {
			if (this.closed) return;
			if (this.firstRender) this.out.write(cursor.hide);
			else this.out.write(clear(this.outputText, this.out.columns));
			super.render();
			this.outputText = [
				style.symbol(this.done, this.aborted),
				color.bold(this.msg),
				style.delimiter(this.done),
				this.value ? this.inactive : color.cyan().underline(this.inactive),
				color.gray("/"),
				this.value ? color.cyan().underline(this.active) : this.active
			].join(" ");
			this.out.write(erase.line + cursor.to(0) + this.outputText);
		}
	};
	module.exports = TogglePrompt;
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/dateparts/datepart.js
var require_datepart = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	module.exports = class DatePart {
		constructor({ token, date, parts, locales }) {
			this.token = token;
			this.date = date || /* @__PURE__ */ new Date();
			this.parts = parts || [this];
			this.locales = locales || {};
		}
		up() {}
		down() {}
		next() {
			const currentIdx = this.parts.indexOf(this);
			return this.parts.find((part, idx) => idx > currentIdx && part instanceof DatePart);
		}
		setTo(val) {}
		prev() {
			let parts = [].concat(this.parts).reverse();
			const currentIdx = parts.indexOf(this);
			return parts.find((part, idx) => idx > currentIdx && part instanceof DatePart);
		}
		toString() {
			return String(this.date);
		}
	};
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/dateparts/meridiem.js
var require_meridiem = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	const DatePart = require_datepart();
	var Meridiem = class extends DatePart {
		constructor(opts = {}) {
			super(opts);
		}
		up() {
			this.date.setHours((this.date.getHours() + 12) % 24);
		}
		down() {
			this.up();
		}
		toString() {
			let meridiem = this.date.getHours() > 12 ? "pm" : "am";
			return /\A/.test(this.token) ? meridiem.toUpperCase() : meridiem;
		}
	};
	module.exports = Meridiem;
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/dateparts/day.js
var require_day = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	const DatePart = require_datepart();
	const pos = (n) => {
		n = n % 10;
		return n === 1 ? "st" : n === 2 ? "nd" : n === 3 ? "rd" : "th";
	};
	var Day = class extends DatePart {
		constructor(opts = {}) {
			super(opts);
		}
		up() {
			this.date.setDate(this.date.getDate() + 1);
		}
		down() {
			this.date.setDate(this.date.getDate() - 1);
		}
		setTo(val) {
			this.date.setDate(parseInt(val.substr(-2)));
		}
		toString() {
			let date = this.date.getDate();
			let day = this.date.getDay();
			return this.token === "DD" ? String(date).padStart(2, "0") : this.token === "Do" ? date + pos(date) : this.token === "d" ? day + 1 : this.token === "ddd" ? this.locales.weekdaysShort[day] : this.token === "dddd" ? this.locales.weekdays[day] : date;
		}
	};
	module.exports = Day;
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/dateparts/hours.js
var require_hours = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	const DatePart = require_datepart();
	var Hours = class extends DatePart {
		constructor(opts = {}) {
			super(opts);
		}
		up() {
			this.date.setHours(this.date.getHours() + 1);
		}
		down() {
			this.date.setHours(this.date.getHours() - 1);
		}
		setTo(val) {
			this.date.setHours(parseInt(val.substr(-2)));
		}
		toString() {
			let hours = this.date.getHours();
			if (/h/.test(this.token)) hours = hours % 12 || 12;
			return this.token.length > 1 ? String(hours).padStart(2, "0") : hours;
		}
	};
	module.exports = Hours;
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/dateparts/milliseconds.js
var require_milliseconds = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	const DatePart = require_datepart();
	var Milliseconds = class extends DatePart {
		constructor(opts = {}) {
			super(opts);
		}
		up() {
			this.date.setMilliseconds(this.date.getMilliseconds() + 1);
		}
		down() {
			this.date.setMilliseconds(this.date.getMilliseconds() - 1);
		}
		setTo(val) {
			this.date.setMilliseconds(parseInt(val.substr(-this.token.length)));
		}
		toString() {
			return String(this.date.getMilliseconds()).padStart(4, "0").substr(0, this.token.length);
		}
	};
	module.exports = Milliseconds;
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/dateparts/minutes.js
var require_minutes = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	const DatePart = require_datepart();
	var Minutes = class extends DatePart {
		constructor(opts = {}) {
			super(opts);
		}
		up() {
			this.date.setMinutes(this.date.getMinutes() + 1);
		}
		down() {
			this.date.setMinutes(this.date.getMinutes() - 1);
		}
		setTo(val) {
			this.date.setMinutes(parseInt(val.substr(-2)));
		}
		toString() {
			let m = this.date.getMinutes();
			return this.token.length > 1 ? String(m).padStart(2, "0") : m;
		}
	};
	module.exports = Minutes;
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/dateparts/month.js
var require_month = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	const DatePart = require_datepart();
	var Month = class extends DatePart {
		constructor(opts = {}) {
			super(opts);
		}
		up() {
			this.date.setMonth(this.date.getMonth() + 1);
		}
		down() {
			this.date.setMonth(this.date.getMonth() - 1);
		}
		setTo(val) {
			val = parseInt(val.substr(-2)) - 1;
			this.date.setMonth(val < 0 ? 0 : val);
		}
		toString() {
			let month = this.date.getMonth();
			let tl = this.token.length;
			return tl === 2 ? String(month + 1).padStart(2, "0") : tl === 3 ? this.locales.monthsShort[month] : tl === 4 ? this.locales.months[month] : String(month + 1);
		}
	};
	module.exports = Month;
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/dateparts/seconds.js
var require_seconds = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	const DatePart = require_datepart();
	var Seconds = class extends DatePart {
		constructor(opts = {}) {
			super(opts);
		}
		up() {
			this.date.setSeconds(this.date.getSeconds() + 1);
		}
		down() {
			this.date.setSeconds(this.date.getSeconds() - 1);
		}
		setTo(val) {
			this.date.setSeconds(parseInt(val.substr(-2)));
		}
		toString() {
			let s = this.date.getSeconds();
			return this.token.length > 1 ? String(s).padStart(2, "0") : s;
		}
	};
	module.exports = Seconds;
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/dateparts/year.js
var require_year = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	const DatePart = require_datepart();
	var Year = class extends DatePart {
		constructor(opts = {}) {
			super(opts);
		}
		up() {
			this.date.setFullYear(this.date.getFullYear() + 1);
		}
		down() {
			this.date.setFullYear(this.date.getFullYear() - 1);
		}
		setTo(val) {
			this.date.setFullYear(val.substr(-4));
		}
		toString() {
			let year = String(this.date.getFullYear()).padStart(4, "0");
			return this.token.length === 2 ? year.substr(-2) : year;
		}
	};
	module.exports = Year;
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/dateparts/index.js
var require_dateparts = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	module.exports = {
		DatePart: require_datepart(),
		Meridiem: require_meridiem(),
		Day: require_day(),
		Hours: require_hours(),
		Milliseconds: require_milliseconds(),
		Minutes: require_minutes(),
		Month: require_month(),
		Seconds: require_seconds(),
		Year: require_year()
	};
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/elements/date.js
var require_date = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	const color = require_kleur();
	const Prompt = require_prompt();
	const { style, clear, figures } = require_util();
	const { erase, cursor } = require_src();
	const { DatePart, Meridiem, Day, Hours, Milliseconds, Minutes, Month, Seconds, Year } = require_dateparts();
	const regex = /\\(.)|"((?:\\["\\]|[^"])+)"|(D[Do]?|d{3,4}|d)|(M{1,4})|(YY(?:YY)?)|([aA])|([Hh]{1,2})|(m{1,2})|(s{1,2})|(S{1,4})|./g;
	const regexGroups = {
		1: ({ token }) => token.replace(/\\(.)/g, "$1"),
		2: (opts) => new Day(opts),
		3: (opts) => new Month(opts),
		4: (opts) => new Year(opts),
		5: (opts) => new Meridiem(opts),
		6: (opts) => new Hours(opts),
		7: (opts) => new Minutes(opts),
		8: (opts) => new Seconds(opts),
		9: (opts) => new Milliseconds(opts)
	};
	const dfltLocales = {
		months: "January,February,March,April,May,June,July,August,September,October,November,December".split(","),
		monthsShort: "Jan,Feb,Mar,Apr,May,Jun,Jul,Aug,Sep,Oct,Nov,Dec".split(","),
		weekdays: "Sunday,Monday,Tuesday,Wednesday,Thursday,Friday,Saturday".split(","),
		weekdaysShort: "Sun,Mon,Tue,Wed,Thu,Fri,Sat".split(",")
	};
	/**
	* DatePrompt Base Element
	* @param {Object} opts Options
	* @param {String} opts.message Message
	* @param {Number} [opts.initial] Index of default value
	* @param {String} [opts.mask] The format mask
	* @param {object} [opts.locales] The date locales
	* @param {String} [opts.error] The error message shown on invalid value
	* @param {Function} [opts.validate] Function to validate the submitted value
	* @param {Stream} [opts.stdin] The Readable stream to listen to
	* @param {Stream} [opts.stdout] The Writable stream to write readline data to
	*/
	var DatePrompt = class extends Prompt {
		constructor(opts = {}) {
			super(opts);
			this.msg = opts.message;
			this.cursor = 0;
			this.typed = "";
			this.locales = Object.assign(dfltLocales, opts.locales);
			this._date = opts.initial || /* @__PURE__ */ new Date();
			this.errorMsg = opts.error || "Please Enter A Valid Value";
			this.validator = opts.validate || (() => true);
			this.mask = opts.mask || "YYYY-MM-DD HH:mm:ss";
			this.clear = clear("", this.out.columns);
			this.render();
		}
		get value() {
			return this.date;
		}
		get date() {
			return this._date;
		}
		set date(date) {
			if (date) this._date.setTime(date.getTime());
		}
		set mask(mask) {
			let result;
			this.parts = [];
			while (result = regex.exec(mask)) {
				let match = result.shift();
				let idx = result.findIndex((gr) => gr != null);
				this.parts.push(idx in regexGroups ? regexGroups[idx]({
					token: result[idx] || match,
					date: this.date,
					parts: this.parts,
					locales: this.locales
				}) : result[idx] || match);
			}
			let parts = this.parts.reduce((arr, i) => {
				if (typeof i === "string" && typeof arr[arr.length - 1] === "string") arr[arr.length - 1] += i;
				else arr.push(i);
				return arr;
			}, []);
			this.parts.splice(0);
			this.parts.push(...parts);
			this.reset();
		}
		moveCursor(n) {
			this.typed = "";
			this.cursor = n;
			this.fire();
		}
		reset() {
			this.moveCursor(this.parts.findIndex((p) => p instanceof DatePart));
			this.fire();
			this.render();
		}
		exit() {
			this.abort();
		}
		abort() {
			this.done = this.aborted = true;
			this.error = false;
			this.fire();
			this.render();
			this.out.write("\n");
			this.close();
		}
		async validate() {
			let valid = await this.validator(this.value);
			if (typeof valid === "string") {
				this.errorMsg = valid;
				valid = false;
			}
			this.error = !valid;
		}
		async submit() {
			await this.validate();
			if (this.error) {
				this.color = "red";
				this.fire();
				this.render();
				return;
			}
			this.done = true;
			this.aborted = false;
			this.fire();
			this.render();
			this.out.write("\n");
			this.close();
		}
		up() {
			this.typed = "";
			this.parts[this.cursor].up();
			this.render();
		}
		down() {
			this.typed = "";
			this.parts[this.cursor].down();
			this.render();
		}
		left() {
			let prev = this.parts[this.cursor].prev();
			if (prev == null) return this.bell();
			this.moveCursor(this.parts.indexOf(prev));
			this.render();
		}
		right() {
			let next = this.parts[this.cursor].next();
			if (next == null) return this.bell();
			this.moveCursor(this.parts.indexOf(next));
			this.render();
		}
		next() {
			let next = this.parts[this.cursor].next();
			this.moveCursor(next ? this.parts.indexOf(next) : this.parts.findIndex((part) => part instanceof DatePart));
			this.render();
		}
		_(c) {
			if (/\d/.test(c)) {
				this.typed += c;
				this.parts[this.cursor].setTo(this.typed);
				this.render();
			}
		}
		render() {
			if (this.closed) return;
			if (this.firstRender) this.out.write(cursor.hide);
			else this.out.write(clear(this.outputText, this.out.columns));
			super.render();
			this.outputText = [
				style.symbol(this.done, this.aborted),
				color.bold(this.msg),
				style.delimiter(false),
				this.parts.reduce((arr, p, idx) => arr.concat(idx === this.cursor && !this.done ? color.cyan().underline(p.toString()) : p), []).join("")
			].join(" ");
			if (this.error) this.outputText += this.errorMsg.split("\n").reduce((a, l, i) => a + `\n${i ? ` ` : figures.pointerSmall} ${color.red().italic(l)}`, ``);
			this.out.write(erase.line + cursor.to(0) + this.outputText);
		}
	};
	module.exports = DatePrompt;
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/elements/number.js
var require_number = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	const color = require_kleur();
	const Prompt = require_prompt();
	const { cursor, erase } = require_src();
	const { style, figures, clear, lines } = require_util();
	const isNumber = /[0-9]/;
	const isDef = (any) => any !== void 0;
	const round = (number, precision) => {
		let factor = Math.pow(10, precision);
		return Math.round(number * factor) / factor;
	};
	/**
	* NumberPrompt Base Element
	* @param {Object} opts Options
	* @param {String} opts.message Message
	* @param {String} [opts.style='default'] Render style
	* @param {Number} [opts.initial] Default value
	* @param {Number} [opts.max=+Infinity] Max value
	* @param {Number} [opts.min=-Infinity] Min value
	* @param {Boolean} [opts.float=false] Parse input as floats
	* @param {Number} [opts.round=2] Round floats to x decimals
	* @param {Number} [opts.increment=1] Number to increment by when using arrow-keys
	* @param {Function} [opts.validate] Validate function
	* @param {Stream} [opts.stdin] The Readable stream to listen to
	* @param {Stream} [opts.stdout] The Writable stream to write readline data to
	* @param {String} [opts.error] The invalid error label
	*/
	var NumberPrompt = class extends Prompt {
		constructor(opts = {}) {
			super(opts);
			this.transform = style.render(opts.style);
			this.msg = opts.message;
			this.initial = isDef(opts.initial) ? opts.initial : "";
			this.float = !!opts.float;
			this.round = opts.round || 2;
			this.inc = opts.increment || 1;
			this.min = isDef(opts.min) ? opts.min : -Infinity;
			this.max = isDef(opts.max) ? opts.max : Infinity;
			this.errorMsg = opts.error || `Please Enter A Valid Value`;
			this.validator = opts.validate || (() => true);
			this.color = `cyan`;
			this.value = ``;
			this.typed = ``;
			this.lastHit = 0;
			this.render();
		}
		set value(v) {
			if (!v && v !== 0) {
				this.placeholder = true;
				this.rendered = color.gray(this.transform.render(`${this.initial}`));
				this._value = ``;
			} else {
				this.placeholder = false;
				this.rendered = this.transform.render(`${round(v, this.round)}`);
				this._value = round(v, this.round);
			}
			this.fire();
		}
		get value() {
			return this._value;
		}
		parse(x) {
			return this.float ? parseFloat(x) : parseInt(x);
		}
		valid(c) {
			return c === `-` || c === `.` && this.float || isNumber.test(c);
		}
		reset() {
			this.typed = ``;
			this.value = ``;
			this.fire();
			this.render();
		}
		exit() {
			this.abort();
		}
		abort() {
			let x = this.value;
			this.value = x !== `` ? x : this.initial;
			this.done = this.aborted = true;
			this.error = false;
			this.fire();
			this.render();
			this.out.write(`\n`);
			this.close();
		}
		async validate() {
			let valid = await this.validator(this.value);
			if (typeof valid === `string`) {
				this.errorMsg = valid;
				valid = false;
			}
			this.error = !valid;
		}
		async submit() {
			await this.validate();
			if (this.error) {
				this.color = `red`;
				this.fire();
				this.render();
				return;
			}
			let x = this.value;
			this.value = x !== `` ? x : this.initial;
			this.done = true;
			this.aborted = false;
			this.error = false;
			this.fire();
			this.render();
			this.out.write(`\n`);
			this.close();
		}
		up() {
			this.typed = ``;
			if (this.value === "") this.value = this.min - this.inc;
			if (this.value >= this.max) return this.bell();
			this.value += this.inc;
			this.color = `cyan`;
			this.fire();
			this.render();
		}
		down() {
			this.typed = ``;
			if (this.value === "") this.value = this.min + this.inc;
			if (this.value <= this.min) return this.bell();
			this.value -= this.inc;
			this.color = `cyan`;
			this.fire();
			this.render();
		}
		delete() {
			let val = this.value.toString();
			if (val.length === 0) return this.bell();
			this.value = this.parse(val = val.slice(0, -1)) || ``;
			if (this.value !== "" && this.value < this.min) this.value = this.min;
			this.color = `cyan`;
			this.fire();
			this.render();
		}
		next() {
			this.value = this.initial;
			this.fire();
			this.render();
		}
		_(c, key) {
			if (!this.valid(c)) return this.bell();
			const now = Date.now();
			if (now - this.lastHit > 1e3) this.typed = ``;
			this.typed += c;
			this.lastHit = now;
			this.color = `cyan`;
			if (c === `.`) return this.fire();
			this.value = Math.min(this.parse(this.typed), this.max);
			if (this.value > this.max) this.value = this.max;
			if (this.value < this.min) this.value = this.min;
			this.fire();
			this.render();
		}
		render() {
			if (this.closed) return;
			if (!this.firstRender) {
				if (this.outputError) this.out.write(cursor.down(lines(this.outputError, this.out.columns) - 1) + clear(this.outputError, this.out.columns));
				this.out.write(clear(this.outputText, this.out.columns));
			}
			super.render();
			this.outputError = "";
			this.outputText = [
				style.symbol(this.done, this.aborted),
				color.bold(this.msg),
				style.delimiter(this.done),
				!this.done || !this.done && !this.placeholder ? color[this.color]().underline(this.rendered) : this.rendered
			].join(` `);
			if (this.error) this.outputError += this.errorMsg.split(`\n`).reduce((a, l, i) => a + `\n${i ? ` ` : figures.pointerSmall} ${color.red().italic(l)}`, ``);
			this.out.write(erase.line + cursor.to(0) + this.outputText + cursor.save + this.outputError + cursor.restore);
		}
	};
	module.exports = NumberPrompt;
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/elements/multiselect.js
var require_multiselect = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	const color = require_kleur();
	const { cursor } = require_src();
	const Prompt = require_prompt();
	const { clear, figures, style, wrap, entriesToDisplay } = require_util();
	/**
	* MultiselectPrompt Base Element
	* @param {Object} opts Options
	* @param {String} opts.message Message
	* @param {Array} opts.choices Array of choice objects
	* @param {String} [opts.hint] Hint to display
	* @param {String} [opts.warn] Hint shown for disabled choices
	* @param {Number} [opts.max] Max choices
	* @param {Number} [opts.cursor=0] Cursor start position
	* @param {Number} [opts.optionsPerPage=10] Max options to display at once
	* @param {Stream} [opts.stdin] The Readable stream to listen to
	* @param {Stream} [opts.stdout] The Writable stream to write readline data to
	*/
	var MultiselectPrompt = class extends Prompt {
		constructor(opts = {}) {
			super(opts);
			this.msg = opts.message;
			this.cursor = opts.cursor || 0;
			this.scrollIndex = opts.cursor || 0;
			this.hint = opts.hint || "";
			this.warn = opts.warn || "- This option is disabled -";
			this.minSelected = opts.min;
			this.showMinError = false;
			this.maxChoices = opts.max;
			this.instructions = opts.instructions;
			this.optionsPerPage = opts.optionsPerPage || 10;
			this.value = opts.choices.map((ch, idx) => {
				if (typeof ch === "string") ch = {
					title: ch,
					value: idx
				};
				return {
					title: ch && (ch.title || ch.value || ch),
					description: ch && ch.description,
					value: ch && (ch.value === void 0 ? idx : ch.value),
					selected: ch && ch.selected,
					disabled: ch && ch.disabled
				};
			});
			this.clear = clear("", this.out.columns);
			if (!opts.overrideRender) this.render();
		}
		reset() {
			this.value.map((v) => !v.selected);
			this.cursor = 0;
			this.fire();
			this.render();
		}
		selected() {
			return this.value.filter((v) => v.selected);
		}
		exit() {
			this.abort();
		}
		abort() {
			this.done = this.aborted = true;
			this.fire();
			this.render();
			this.out.write("\n");
			this.close();
		}
		submit() {
			const selected = this.value.filter((e) => e.selected);
			if (this.minSelected && selected.length < this.minSelected) {
				this.showMinError = true;
				this.render();
			} else {
				this.done = true;
				this.aborted = false;
				this.fire();
				this.render();
				this.out.write("\n");
				this.close();
			}
		}
		first() {
			this.cursor = 0;
			this.render();
		}
		last() {
			this.cursor = this.value.length - 1;
			this.render();
		}
		next() {
			this.cursor = (this.cursor + 1) % this.value.length;
			this.render();
		}
		up() {
			if (this.cursor === 0) this.cursor = this.value.length - 1;
			else this.cursor--;
			this.render();
		}
		down() {
			if (this.cursor === this.value.length - 1) this.cursor = 0;
			else this.cursor++;
			this.render();
		}
		left() {
			this.value[this.cursor].selected = false;
			this.render();
		}
		right() {
			if (this.value.filter((e) => e.selected).length >= this.maxChoices) return this.bell();
			this.value[this.cursor].selected = true;
			this.render();
		}
		handleSpaceToggle() {
			const v = this.value[this.cursor];
			if (v.selected) {
				v.selected = false;
				this.render();
			} else if (v.disabled || this.value.filter((e) => e.selected).length >= this.maxChoices) return this.bell();
			else {
				v.selected = true;
				this.render();
			}
		}
		toggleAll() {
			if (this.maxChoices !== void 0 || this.value[this.cursor].disabled) return this.bell();
			const newSelected = !this.value[this.cursor].selected;
			this.value.filter((v) => !v.disabled).forEach((v) => v.selected = newSelected);
			this.render();
		}
		_(c, key) {
			if (c === " ") this.handleSpaceToggle();
			else if (c === "a") this.toggleAll();
			else return this.bell();
		}
		renderInstructions() {
			if (this.instructions === void 0 || this.instructions) {
				if (typeof this.instructions === "string") return this.instructions;
				return `
Instructions:
    ${figures.arrowUp}/${figures.arrowDown}: Highlight option\n    ${figures.arrowLeft}/${figures.arrowRight}/[space]: Toggle selection\n` + (this.maxChoices === void 0 ? `    a: Toggle all\n` : "") + `    enter/return: Complete answer`;
			}
			return "";
		}
		renderOption(cursor, v, i, arrowIndicator) {
			const prefix = (v.selected ? color.green(figures.radioOn) : figures.radioOff) + " " + arrowIndicator + " ";
			let title, desc;
			if (v.disabled) title = cursor === i ? color.gray().underline(v.title) : color.strikethrough().gray(v.title);
			else {
				title = cursor === i ? color.cyan().underline(v.title) : v.title;
				if (cursor === i && v.description) {
					desc = ` - ${v.description}`;
					if (prefix.length + title.length + desc.length >= this.out.columns || v.description.split(/\r?\n/).length > 1) desc = "\n" + wrap(v.description, {
						margin: prefix.length,
						width: this.out.columns
					});
				}
			}
			return prefix + title + color.gray(desc || "");
		}
		paginateOptions(options) {
			if (options.length === 0) return color.red("No matches for this query.");
			let { startIndex, endIndex } = entriesToDisplay(this.cursor, options.length, this.optionsPerPage);
			let prefix, styledOptions = [];
			for (let i = startIndex; i < endIndex; i++) {
				if (i === startIndex && startIndex > 0) prefix = figures.arrowUp;
				else if (i === endIndex - 1 && endIndex < options.length) prefix = figures.arrowDown;
				else prefix = " ";
				styledOptions.push(this.renderOption(this.cursor, options[i], i, prefix));
			}
			return "\n" + styledOptions.join("\n");
		}
		renderOptions(options) {
			if (!this.done) return this.paginateOptions(options);
			return "";
		}
		renderDoneOrInstructions() {
			if (this.done) return this.value.filter((e) => e.selected).map((v) => v.title).join(", ");
			const output = [color.gray(this.hint), this.renderInstructions()];
			if (this.value[this.cursor].disabled) output.push(color.yellow(this.warn));
			return output.join(" ");
		}
		render() {
			if (this.closed) return;
			if (this.firstRender) this.out.write(cursor.hide);
			super.render();
			let prompt = [
				style.symbol(this.done, this.aborted),
				color.bold(this.msg),
				style.delimiter(false),
				this.renderDoneOrInstructions()
			].join(" ");
			if (this.showMinError) {
				prompt += color.red(`You must select a minimum of ${this.minSelected} choices.`);
				this.showMinError = false;
			}
			prompt += this.renderOptions(this.value);
			this.out.write(this.clear + prompt);
			this.clear = clear(prompt, this.out.columns);
		}
	};
	module.exports = MultiselectPrompt;
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/elements/autocomplete.js
var require_autocomplete = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	const color = require_kleur();
	const Prompt = require_prompt();
	const { erase, cursor } = require_src();
	const { style, clear, figures, wrap, entriesToDisplay } = require_util();
	const getVal = (arr, i) => arr[i] && (arr[i].value || arr[i].title || arr[i]);
	const getTitle = (arr, i) => arr[i] && (arr[i].title || arr[i].value || arr[i]);
	const getIndex = (arr, valOrTitle) => {
		const index = arr.findIndex((el) => el.value === valOrTitle || el.title === valOrTitle);
		return index > -1 ? index : void 0;
	};
	/**
	* TextPrompt Base Element
	* @param {Object} opts Options
	* @param {String} opts.message Message
	* @param {Array} opts.choices Array of auto-complete choices objects
	* @param {Function} [opts.suggest] Filter function. Defaults to sort by title
	* @param {Number} [opts.limit=10] Max number of results to show
	* @param {Number} [opts.cursor=0] Cursor start position
	* @param {String} [opts.style='default'] Render style
	* @param {String} [opts.fallback] Fallback message - initial to default value
	* @param {String} [opts.initial] Index of the default value
	* @param {Boolean} [opts.clearFirst] The first ESCAPE keypress will clear the input
	* @param {Stream} [opts.stdin] The Readable stream to listen to
	* @param {Stream} [opts.stdout] The Writable stream to write readline data to
	* @param {String} [opts.noMatches] The no matches found label
	*/
	var AutocompletePrompt = class extends Prompt {
		constructor(opts = {}) {
			super(opts);
			this.msg = opts.message;
			this.suggest = opts.suggest;
			this.choices = opts.choices;
			this.initial = typeof opts.initial === "number" ? opts.initial : getIndex(opts.choices, opts.initial);
			this.select = this.initial || opts.cursor || 0;
			this.i18n = { noMatches: opts.noMatches || "no matches found" };
			this.fallback = opts.fallback || this.initial;
			this.clearFirst = opts.clearFirst || false;
			this.suggestions = [];
			this.input = "";
			this.limit = opts.limit || 10;
			this.cursor = 0;
			this.transform = style.render(opts.style);
			this.scale = this.transform.scale;
			this.render = this.render.bind(this);
			this.complete = this.complete.bind(this);
			this.clear = clear("", this.out.columns);
			this.complete(this.render);
			this.render();
		}
		set fallback(fb) {
			this._fb = Number.isSafeInteger(parseInt(fb)) ? parseInt(fb) : fb;
		}
		get fallback() {
			let choice;
			if (typeof this._fb === "number") choice = this.choices[this._fb];
			else if (typeof this._fb === "string") choice = { title: this._fb };
			return choice || this._fb || { title: this.i18n.noMatches };
		}
		moveSelect(i) {
			this.select = i;
			if (this.suggestions.length > 0) this.value = getVal(this.suggestions, i);
			else this.value = this.fallback.value;
			this.fire();
		}
		async complete(cb) {
			const p = this.completing = this.suggest(this.input, this.choices);
			const suggestions = await p;
			if (this.completing !== p) return;
			this.suggestions = suggestions.map((s, i, arr) => ({
				title: getTitle(arr, i),
				value: getVal(arr, i),
				description: s.description
			}));
			this.completing = false;
			const l = Math.max(suggestions.length - 1, 0);
			this.moveSelect(Math.min(l, this.select));
			cb && cb();
		}
		reset() {
			this.input = "";
			this.complete(() => {
				this.moveSelect(this.initial !== void 0 ? this.initial : 0);
				this.render();
			});
			this.render();
		}
		exit() {
			if (this.clearFirst && this.input.length > 0) this.reset();
			else {
				this.done = this.exited = true;
				this.aborted = false;
				this.fire();
				this.render();
				this.out.write("\n");
				this.close();
			}
		}
		abort() {
			this.done = this.aborted = true;
			this.exited = false;
			this.fire();
			this.render();
			this.out.write("\n");
			this.close();
		}
		submit() {
			this.done = true;
			this.aborted = this.exited = false;
			this.fire();
			this.render();
			this.out.write("\n");
			this.close();
		}
		_(c, key) {
			let s1 = this.input.slice(0, this.cursor);
			this.input = `${s1}${c}${this.input.slice(this.cursor)}`;
			this.cursor = s1.length + 1;
			this.complete(this.render);
			this.render();
		}
		delete() {
			if (this.cursor === 0) return this.bell();
			this.input = `${this.input.slice(0, this.cursor - 1)}${this.input.slice(this.cursor)}`;
			this.complete(this.render);
			this.cursor = this.cursor - 1;
			this.render();
		}
		deleteForward() {
			if (this.cursor * this.scale >= this.rendered.length) return this.bell();
			this.input = `${this.input.slice(0, this.cursor)}${this.input.slice(this.cursor + 1)}`;
			this.complete(this.render);
			this.render();
		}
		first() {
			this.moveSelect(0);
			this.render();
		}
		last() {
			this.moveSelect(this.suggestions.length - 1);
			this.render();
		}
		up() {
			if (this.select === 0) this.moveSelect(this.suggestions.length - 1);
			else this.moveSelect(this.select - 1);
			this.render();
		}
		down() {
			if (this.select === this.suggestions.length - 1) this.moveSelect(0);
			else this.moveSelect(this.select + 1);
			this.render();
		}
		next() {
			if (this.select === this.suggestions.length - 1) this.moveSelect(0);
			else this.moveSelect(this.select + 1);
			this.render();
		}
		nextPage() {
			this.moveSelect(Math.min(this.select + this.limit, this.suggestions.length - 1));
			this.render();
		}
		prevPage() {
			this.moveSelect(Math.max(this.select - this.limit, 0));
			this.render();
		}
		left() {
			if (this.cursor <= 0) return this.bell();
			this.cursor = this.cursor - 1;
			this.render();
		}
		right() {
			if (this.cursor * this.scale >= this.rendered.length) return this.bell();
			this.cursor = this.cursor + 1;
			this.render();
		}
		renderOption(v, hovered, isStart, isEnd) {
			let desc;
			let prefix = isStart ? figures.arrowUp : isEnd ? figures.arrowDown : " ";
			let title = hovered ? color.cyan().underline(v.title) : v.title;
			prefix = (hovered ? color.cyan(figures.pointer) + " " : "  ") + prefix;
			if (v.description) {
				desc = ` - ${v.description}`;
				if (prefix.length + title.length + desc.length >= this.out.columns || v.description.split(/\r?\n/).length > 1) desc = "\n" + wrap(v.description, {
					margin: 3,
					width: this.out.columns
				});
			}
			return prefix + " " + title + color.gray(desc || "");
		}
		render() {
			if (this.closed) return;
			if (this.firstRender) this.out.write(cursor.hide);
			else this.out.write(clear(this.outputText, this.out.columns));
			super.render();
			let { startIndex, endIndex } = entriesToDisplay(this.select, this.choices.length, this.limit);
			this.outputText = [
				style.symbol(this.done, this.aborted, this.exited),
				color.bold(this.msg),
				style.delimiter(this.completing),
				this.done && this.suggestions[this.select] ? this.suggestions[this.select].title : this.rendered = this.transform.render(this.input)
			].join(" ");
			if (!this.done) {
				const suggestions = this.suggestions.slice(startIndex, endIndex).map((item, i) => this.renderOption(item, this.select === i + startIndex, i === 0 && startIndex > 0, i + startIndex === endIndex - 1 && endIndex < this.choices.length)).join("\n");
				this.outputText += `\n` + (suggestions || color.gray(this.fallback.title));
			}
			this.out.write(erase.line + cursor.to(0) + this.outputText);
		}
	};
	module.exports = AutocompletePrompt;
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/elements/autocompleteMultiselect.js
var require_autocompleteMultiselect = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	const color = require_kleur();
	const { cursor } = require_src();
	const MultiselectPrompt = require_multiselect();
	const { clear, style, figures } = require_util();
	/**
	* MultiselectPrompt Base Element
	* @param {Object} opts Options
	* @param {String} opts.message Message
	* @param {Array} opts.choices Array of choice objects
	* @param {String} [opts.hint] Hint to display
	* @param {String} [opts.warn] Hint shown for disabled choices
	* @param {Number} [opts.max] Max choices
	* @param {Number} [opts.cursor=0] Cursor start position
	* @param {Stream} [opts.stdin] The Readable stream to listen to
	* @param {Stream} [opts.stdout] The Writable stream to write readline data to
	*/
	var AutocompleteMultiselectPrompt = class extends MultiselectPrompt {
		constructor(opts = {}) {
			opts.overrideRender = true;
			super(opts);
			this.inputValue = "";
			this.clear = clear("", this.out.columns);
			this.filteredOptions = this.value;
			this.render();
		}
		last() {
			this.cursor = this.filteredOptions.length - 1;
			this.render();
		}
		next() {
			this.cursor = (this.cursor + 1) % this.filteredOptions.length;
			this.render();
		}
		up() {
			if (this.cursor === 0) this.cursor = this.filteredOptions.length - 1;
			else this.cursor--;
			this.render();
		}
		down() {
			if (this.cursor === this.filteredOptions.length - 1) this.cursor = 0;
			else this.cursor++;
			this.render();
		}
		left() {
			this.filteredOptions[this.cursor].selected = false;
			this.render();
		}
		right() {
			if (this.value.filter((e) => e.selected).length >= this.maxChoices) return this.bell();
			this.filteredOptions[this.cursor].selected = true;
			this.render();
		}
		delete() {
			if (this.inputValue.length) {
				this.inputValue = this.inputValue.substr(0, this.inputValue.length - 1);
				this.updateFilteredOptions();
			}
		}
		updateFilteredOptions() {
			const currentHighlight = this.filteredOptions[this.cursor];
			this.filteredOptions = this.value.filter((v) => {
				if (this.inputValue) {
					if (typeof v.title === "string") {
						if (v.title.toLowerCase().includes(this.inputValue.toLowerCase())) return true;
					}
					if (typeof v.value === "string") {
						if (v.value.toLowerCase().includes(this.inputValue.toLowerCase())) return true;
					}
					return false;
				}
				return true;
			});
			const newHighlightIndex = this.filteredOptions.findIndex((v) => v === currentHighlight);
			this.cursor = newHighlightIndex < 0 ? 0 : newHighlightIndex;
			this.render();
		}
		handleSpaceToggle() {
			const v = this.filteredOptions[this.cursor];
			if (v.selected) {
				v.selected = false;
				this.render();
			} else if (v.disabled || this.value.filter((e) => e.selected).length >= this.maxChoices) return this.bell();
			else {
				v.selected = true;
				this.render();
			}
		}
		handleInputChange(c) {
			this.inputValue = this.inputValue + c;
			this.updateFilteredOptions();
		}
		_(c, key) {
			if (c === " ") this.handleSpaceToggle();
			else this.handleInputChange(c);
		}
		renderInstructions() {
			if (this.instructions === void 0 || this.instructions) {
				if (typeof this.instructions === "string") return this.instructions;
				return `
Instructions:
    ${figures.arrowUp}/${figures.arrowDown}: Highlight option
    ${figures.arrowLeft}/${figures.arrowRight}/[space]: Toggle selection
    [a,b,c]/delete: Filter choices
    enter/return: Complete answer
`;
			}
			return "";
		}
		renderCurrentInput() {
			return `
Filtered results for: ${this.inputValue ? this.inputValue : color.gray("Enter something to filter")}\n`;
		}
		renderOption(cursor, v, i, arrowIndicator) {
			const prefix = (v.selected ? color.green(figures.radioOn) : figures.radioOff) + " " + arrowIndicator + " ";
			let title;
			if (v.disabled) title = cursor === i ? color.gray().underline(v.title) : color.strikethrough().gray(v.title);
			else title = cursor === i ? color.cyan().underline(v.title) : v.title;
			return prefix + title;
		}
		renderDoneOrInstructions() {
			if (this.done) return this.value.filter((e) => e.selected).map((v) => v.title).join(", ");
			const output = [
				color.gray(this.hint),
				this.renderInstructions(),
				this.renderCurrentInput()
			];
			if (this.filteredOptions.length && this.filteredOptions[this.cursor].disabled) output.push(color.yellow(this.warn));
			return output.join(" ");
		}
		render() {
			if (this.closed) return;
			if (this.firstRender) this.out.write(cursor.hide);
			super.render();
			let prompt = [
				style.symbol(this.done, this.aborted),
				color.bold(this.msg),
				style.delimiter(false),
				this.renderDoneOrInstructions()
			].join(" ");
			if (this.showMinError) {
				prompt += color.red(`You must select a minimum of ${this.minSelected} choices.`);
				this.showMinError = false;
			}
			prompt += this.renderOptions(this.filteredOptions);
			this.out.write(this.clear + prompt);
			this.clear = clear(prompt, this.out.columns);
		}
	};
	module.exports = AutocompleteMultiselectPrompt;
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/elements/confirm.js
var require_confirm = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	const color = require_kleur();
	const Prompt = require_prompt();
	const { style, clear } = require_util();
	const { erase, cursor } = require_src();
	/**
	* ConfirmPrompt Base Element
	* @param {Object} opts Options
	* @param {String} opts.message Message
	* @param {Boolean} [opts.initial] Default value (true/false)
	* @param {Stream} [opts.stdin] The Readable stream to listen to
	* @param {Stream} [opts.stdout] The Writable stream to write readline data to
	* @param {String} [opts.yes] The "Yes" label
	* @param {String} [opts.yesOption] The "Yes" option when choosing between yes/no
	* @param {String} [opts.no] The "No" label
	* @param {String} [opts.noOption] The "No" option when choosing between yes/no
	*/
	var ConfirmPrompt = class extends Prompt {
		constructor(opts = {}) {
			super(opts);
			this.msg = opts.message;
			this.value = opts.initial;
			this.initialValue = !!opts.initial;
			this.yesMsg = opts.yes || "yes";
			this.yesOption = opts.yesOption || "(Y/n)";
			this.noMsg = opts.no || "no";
			this.noOption = opts.noOption || "(y/N)";
			this.render();
		}
		reset() {
			this.value = this.initialValue;
			this.fire();
			this.render();
		}
		exit() {
			this.abort();
		}
		abort() {
			this.done = this.aborted = true;
			this.fire();
			this.render();
			this.out.write("\n");
			this.close();
		}
		submit() {
			this.value = this.value || false;
			this.done = true;
			this.aborted = false;
			this.fire();
			this.render();
			this.out.write("\n");
			this.close();
		}
		_(c, key) {
			if (c.toLowerCase() === "y") {
				this.value = true;
				return this.submit();
			}
			if (c.toLowerCase() === "n") {
				this.value = false;
				return this.submit();
			}
			return this.bell();
		}
		render() {
			if (this.closed) return;
			if (this.firstRender) this.out.write(cursor.hide);
			else this.out.write(clear(this.outputText, this.out.columns));
			super.render();
			this.outputText = [
				style.symbol(this.done, this.aborted),
				color.bold(this.msg),
				style.delimiter(this.done),
				this.done ? this.value ? this.yesMsg : this.noMsg : color.gray(this.initialValue ? this.yesOption : this.noOption)
			].join(" ");
			this.out.write(erase.line + cursor.to(0) + this.outputText);
		}
	};
	module.exports = ConfirmPrompt;
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/elements/index.js
var require_elements = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	module.exports = {
		TextPrompt: require_text(),
		SelectPrompt: require_select(),
		TogglePrompt: require_toggle(),
		DatePrompt: require_date(),
		NumberPrompt: require_number(),
		MultiselectPrompt: require_multiselect(),
		AutocompletePrompt: require_autocomplete(),
		AutocompleteMultiselectPrompt: require_autocompleteMultiselect(),
		ConfirmPrompt: require_confirm()
	};
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/prompts.js
var require_prompts$1 = /* @__PURE__ */ __commonJSMin(((exports) => {
	const $ = exports;
	const el = require_elements();
	const noop = (v) => v;
	function toPrompt(type, args, opts = {}) {
		return new Promise((res, rej) => {
			const p = new el[type](args);
			const onAbort = opts.onAbort || noop;
			const onSubmit = opts.onSubmit || noop;
			const onExit = opts.onExit || noop;
			p.on("state", args.onState || noop);
			p.on("submit", (x) => res(onSubmit(x)));
			p.on("exit", (x) => res(onExit(x)));
			p.on("abort", (x) => rej(onAbort(x)));
		});
	}
	/**
	* Text prompt
	* @param {string} args.message Prompt message to display
	* @param {string} [args.initial] Default string value
	* @param {string} [args.style="default"] Render style ('default', 'password', 'invisible')
	* @param {function} [args.onState] On state change callback
	* @param {function} [args.validate] Function to validate user input
	* @param {Stream} [args.stdin] The Readable stream to listen to
	* @param {Stream} [args.stdout] The Writable stream to write readline data to
	* @returns {Promise} Promise with user input
	*/
	$.text = (args) => toPrompt("TextPrompt", args);
	/**
	* Password prompt with masked input
	* @param {string} args.message Prompt message to display
	* @param {string} [args.initial] Default string value
	* @param {function} [args.onState] On state change callback
	* @param {function} [args.validate] Function to validate user input
	* @param {Stream} [args.stdin] The Readable stream to listen to
	* @param {Stream} [args.stdout] The Writable stream to write readline data to
	* @returns {Promise} Promise with user input
	*/
	$.password = (args) => {
		args.style = "password";
		return $.text(args);
	};
	/**
	* Prompt where input is invisible, like sudo
	* @param {string} args.message Prompt message to display
	* @param {string} [args.initial] Default string value
	* @param {function} [args.onState] On state change callback
	* @param {function} [args.validate] Function to validate user input
	* @param {Stream} [args.stdin] The Readable stream to listen to
	* @param {Stream} [args.stdout] The Writable stream to write readline data to
	* @returns {Promise} Promise with user input
	*/
	$.invisible = (args) => {
		args.style = "invisible";
		return $.text(args);
	};
	/**
	* Number prompt
	* @param {string} args.message Prompt message to display
	* @param {number} args.initial Default number value
	* @param {function} [args.onState] On state change callback
	* @param {number} [args.max] Max value
	* @param {number} [args.min] Min value
	* @param {string} [args.style="default"] Render style ('default', 'password', 'invisible')
	* @param {Boolean} [opts.float=false] Parse input as floats
	* @param {Number} [opts.round=2] Round floats to x decimals
	* @param {Number} [opts.increment=1] Number to increment by when using arrow-keys
	* @param {function} [args.validate] Function to validate user input
	* @param {Stream} [args.stdin] The Readable stream to listen to
	* @param {Stream} [args.stdout] The Writable stream to write readline data to
	* @returns {Promise} Promise with user input
	*/
	$.number = (args) => toPrompt("NumberPrompt", args);
	/**
	* Date prompt
	* @param {string} args.message Prompt message to display
	* @param {number} args.initial Default number value
	* @param {function} [args.onState] On state change callback
	* @param {number} [args.max] Max value
	* @param {number} [args.min] Min value
	* @param {string} [args.style="default"] Render style ('default', 'password', 'invisible')
	* @param {Boolean} [opts.float=false] Parse input as floats
	* @param {Number} [opts.round=2] Round floats to x decimals
	* @param {Number} [opts.increment=1] Number to increment by when using arrow-keys
	* @param {function} [args.validate] Function to validate user input
	* @param {Stream} [args.stdin] The Readable stream to listen to
	* @param {Stream} [args.stdout] The Writable stream to write readline data to
	* @returns {Promise} Promise with user input
	*/
	$.date = (args) => toPrompt("DatePrompt", args);
	/**
	* Classic yes/no prompt
	* @param {string} args.message Prompt message to display
	* @param {boolean} [args.initial=false] Default value
	* @param {function} [args.onState] On state change callback
	* @param {Stream} [args.stdin] The Readable stream to listen to
	* @param {Stream} [args.stdout] The Writable stream to write readline data to
	* @returns {Promise} Promise with user input
	*/
	$.confirm = (args) => toPrompt("ConfirmPrompt", args);
	/**
	* List prompt, split intput string by `seperator`
	* @param {string} args.message Prompt message to display
	* @param {string} [args.initial] Default string value
	* @param {string} [args.style="default"] Render style ('default', 'password', 'invisible')
	* @param {string} [args.separator] String separator
	* @param {function} [args.onState] On state change callback
	* @param {Stream} [args.stdin] The Readable stream to listen to
	* @param {Stream} [args.stdout] The Writable stream to write readline data to
	* @returns {Promise} Promise with user input, in form of an `Array`
	*/
	$.list = (args) => {
		const sep = args.separator || ",";
		return toPrompt("TextPrompt", args, { onSubmit: (str) => str.split(sep).map((s) => s.trim()) });
	};
	/**
	* Toggle/switch prompt
	* @param {string} args.message Prompt message to display
	* @param {boolean} [args.initial=false] Default value
	* @param {string} [args.active="on"] Text for `active` state
	* @param {string} [args.inactive="off"] Text for `inactive` state
	* @param {function} [args.onState] On state change callback
	* @param {Stream} [args.stdin] The Readable stream to listen to
	* @param {Stream} [args.stdout] The Writable stream to write readline data to
	* @returns {Promise} Promise with user input
	*/
	$.toggle = (args) => toPrompt("TogglePrompt", args);
	/**
	* Interactive select prompt
	* @param {string} args.message Prompt message to display
	* @param {Array} args.choices Array of choices objects `[{ title, value }, ...]`
	* @param {number} [args.initial] Index of default value
	* @param {String} [args.hint] Hint to display
	* @param {function} [args.onState] On state change callback
	* @param {Stream} [args.stdin] The Readable stream to listen to
	* @param {Stream} [args.stdout] The Writable stream to write readline data to
	* @returns {Promise} Promise with user input
	*/
	$.select = (args) => toPrompt("SelectPrompt", args);
	/**
	* Interactive multi-select / autocompleteMultiselect prompt
	* @param {string} args.message Prompt message to display
	* @param {Array} args.choices Array of choices objects `[{ title, value, [selected] }, ...]`
	* @param {number} [args.max] Max select
	* @param {string} [args.hint] Hint to display user
	* @param {Number} [args.cursor=0] Cursor start position
	* @param {function} [args.onState] On state change callback
	* @param {Stream} [args.stdin] The Readable stream to listen to
	* @param {Stream} [args.stdout] The Writable stream to write readline data to
	* @returns {Promise} Promise with user input
	*/
	$.multiselect = (args) => {
		args.choices = [].concat(args.choices || []);
		const toSelected = (items) => items.filter((item) => item.selected).map((item) => item.value);
		return toPrompt("MultiselectPrompt", args, {
			onAbort: toSelected,
			onSubmit: toSelected
		});
	};
	$.autocompleteMultiselect = (args) => {
		args.choices = [].concat(args.choices || []);
		const toSelected = (items) => items.filter((item) => item.selected).map((item) => item.value);
		return toPrompt("AutocompleteMultiselectPrompt", args, {
			onAbort: toSelected,
			onSubmit: toSelected
		});
	};
	const byTitle = (input, choices) => Promise.resolve(choices.filter((item) => item.title.slice(0, input.length).toLowerCase() === input.toLowerCase()));
	/**
	* Interactive auto-complete prompt
	* @param {string} args.message Prompt message to display
	* @param {Array} args.choices Array of auto-complete choices objects `[{ title, value }, ...]`
	* @param {Function} [args.suggest] Function to filter results based on user input. Defaults to sort by `title`
	* @param {number} [args.limit=10] Max number of results to show
	* @param {string} [args.style="default"] Render style ('default', 'password', 'invisible')
	* @param {String} [args.initial] Index of the default value
	* @param {boolean} [opts.clearFirst] The first ESCAPE keypress will clear the input
	* @param {String} [args.fallback] Fallback message - defaults to initial value
	* @param {function} [args.onState] On state change callback
	* @param {Stream} [args.stdin] The Readable stream to listen to
	* @param {Stream} [args.stdout] The Writable stream to write readline data to
	* @returns {Promise} Promise with user input
	*/
	$.autocomplete = (args) => {
		args.suggest = args.suggest || byTitle;
		args.choices = [].concat(args.choices || []);
		return toPrompt("AutocompletePrompt", args);
	};
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/lib/index.js
var require_lib$1 = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	const prompts = require_prompts$1();
	const passOn = [
		"suggest",
		"format",
		"onState",
		"validate",
		"onRender",
		"type"
	];
	const noop = () => {};
	/**
	* Prompt for a series of questions
	* @param {Array|Object} questions Single question object or Array of question objects
	* @param {Function} [onSubmit] Callback function called on prompt submit
	* @param {Function} [onCancel] Callback function called on cancel/abort
	* @returns {Object} Object with values from user input
	*/
	async function prompt(questions = [], { onSubmit = noop, onCancel = noop } = {}) {
		const answers = {};
		const override = prompt._override || {};
		questions = [].concat(questions);
		let answer, question, quit, name, type, lastPrompt;
		const getFormattedAnswer = async (question, answer, skipValidation = false) => {
			if (!skipValidation && question.validate && question.validate(answer) !== true) return;
			return question.format ? await question.format(answer, answers) : answer;
		};
		for (question of questions) {
			({name, type} = question);
			if (typeof type === "function") {
				type = await type(answer, { ...answers }, question);
				question["type"] = type;
			}
			if (!type) continue;
			for (let key in question) {
				if (passOn.includes(key)) continue;
				let value = question[key];
				question[key] = typeof value === "function" ? await value(answer, { ...answers }, lastPrompt) : value;
			}
			lastPrompt = question;
			if (typeof question.message !== "string") throw new Error("prompt message is required");
			({name, type} = question);
			if (prompts[type] === void 0) throw new Error(`prompt type (${type}) is not defined`);
			if (override[question.name] !== void 0) {
				answer = await getFormattedAnswer(question, override[question.name]);
				if (answer !== void 0) {
					answers[name] = answer;
					continue;
				}
			}
			try {
				answer = prompt._injected ? getInjectedAnswer(prompt._injected, question.initial) : await prompts[type](question);
				answers[name] = answer = await getFormattedAnswer(question, answer, true);
				quit = await onSubmit(question, answer, answers);
			} catch (err) {
				quit = !await onCancel(question, answers);
			}
			if (quit) return answers;
		}
		return answers;
	}
	function getInjectedAnswer(injected, deafultValue) {
		const answer = injected.shift();
		if (answer instanceof Error) throw answer;
		return answer === void 0 ? deafultValue : answer;
	}
	function inject(answers) {
		prompt._injected = (prompt._injected || []).concat(answers);
	}
	function override(answers) {
		prompt._override = Object.assign({}, answers);
	}
	module.exports = Object.assign(prompt, {
		prompt,
		prompts,
		inject,
		override
	});
}));
//#endregion
//#region node_modules/.pnpm/@posva+prompts@2.4.4/node_modules/@posva/prompts/index.js
var require_prompts = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	module.exports = require_lib$1();
}));
//#endregion
//#region node_modules/.pnpm/isexe@4.0.0/node_modules/isexe/dist/commonjs/index.min.js
var require_index_min = /* @__PURE__ */ __commonJSMin(((exports) => {
	var a = (t, e) => () => (e || t((e = { exports: {} }).exports, e), e.exports);
	var _ = a((i) => {
		"use strict";
		Object.defineProperty(i, "__esModule", { value: !0 });
		i.sync = i.isexe = void 0;
		var M = __require("node:fs"), x = __require("node:fs/promises"), q = async (t, e = {}) => {
			let { ignoreErrors: r = !1 } = e;
			try {
				return d(await (0, x.stat)(t), e);
			} catch (s) {
				let n = s;
				if (r || n.code === "EACCES") return !1;
				throw n;
			}
		};
		i.isexe = q;
		var m = (t, e = {}) => {
			let { ignoreErrors: r = !1 } = e;
			try {
				return d((0, M.statSync)(t), e);
			} catch (s) {
				let n = s;
				if (r || n.code === "EACCES") return !1;
				throw n;
			}
		};
		i.sync = m;
		var d = (t, e) => t.isFile() && A(t, e), A = (t, e) => {
			let r = e.uid ?? process.getuid?.(), s = e.groups ?? process.getgroups?.() ?? [], n = e.gid ?? process.getgid?.() ?? s[0];
			if (r === void 0 || n === void 0) throw new Error("cannot get uid or gid");
			let u = new Set([n, ...s]), c = t.mode, S = t.uid, P = t.gid, f = parseInt("100", 8), l = parseInt("010", 8), j = parseInt("001", 8), C = f | l;
			return !!(c & j || c & l && u.has(P) || c & f && S === r || c & C && r === 0);
		};
	});
	var g = a((o) => {
		"use strict";
		Object.defineProperty(o, "__esModule", { value: !0 });
		o.sync = o.isexe = void 0;
		var T = __require("node:fs"), I = __require("node:fs/promises"), D = __require("node:path"), F = async (t, e = {}) => {
			let { ignoreErrors: r = !1 } = e;
			try {
				return y(await (0, I.stat)(t), t, e);
			} catch (s) {
				let n = s;
				if (r || n.code === "EACCES") return !1;
				throw n;
			}
		};
		o.isexe = F;
		var L = (t, e = {}) => {
			let { ignoreErrors: r = !1 } = e;
			try {
				return y((0, T.statSync)(t), t, e);
			} catch (s) {
				let n = s;
				if (r || n.code === "EACCES") return !1;
				throw n;
			}
		};
		o.sync = L;
		var B = (t, e) => {
			let { pathExt: r = process.env.PATHEXT || "" } = e, s = r.split(D.delimiter);
			if (s.indexOf("") !== -1) return !0;
			for (let n of s) {
				let u = n.toLowerCase(), c = t.substring(t.length - u.length).toLowerCase();
				if (u && c === u) return !0;
			}
			return !1;
		}, y = (t, e, r) => t.isFile() && B(e, r);
	});
	var p = a((h) => {
		"use strict";
		Object.defineProperty(h, "__esModule", { value: !0 });
	});
	var v = exports && exports.__createBinding || (Object.create ? (function(t, e, r, s) {
		s === void 0 && (s = r);
		var n = Object.getOwnPropertyDescriptor(e, r);
		(!n || ("get" in n ? !e.__esModule : n.writable || n.configurable)) && (n = {
			enumerable: !0,
			get: function() {
				return e[r];
			}
		}), Object.defineProperty(t, s, n);
	}) : (function(t, e, r, s) {
		s === void 0 && (s = r), t[s] = e[r];
	})), G = exports && exports.__setModuleDefault || (Object.create ? (function(t, e) {
		Object.defineProperty(t, "default", {
			enumerable: !0,
			value: e
		});
	}) : function(t, e) {
		t.default = e;
	}), w = exports && exports.__importStar || (function() {
		var t = function(e) {
			return t = Object.getOwnPropertyNames || function(r) {
				var s = [];
				for (var n in r) Object.prototype.hasOwnProperty.call(r, n) && (s[s.length] = n);
				return s;
			}, t(e);
		};
		return function(e) {
			if (e && e.__esModule) return e;
			var r = {};
			if (e != null) for (var s = t(e), n = 0; n < s.length; n++) s[n] !== "default" && v(r, e, s[n]);
			return G(r, e), r;
		};
	})(), X = exports && exports.__exportStar || function(t, e) {
		for (var r in t) r !== "default" && !Object.prototype.hasOwnProperty.call(e, r) && v(e, t, r);
	};
	Object.defineProperty(exports, "__esModule", { value: !0 });
	exports.sync = exports.isexe = exports.posix = exports.win32 = void 0;
	var E = w(_());
	exports.posix = E;
	var O = w(g());
	exports.win32 = O;
	X(p(), exports);
	var b = (process.env._ISEXE_TEST_PLATFORM_ || process.platform) === "win32" ? O : E;
	exports.isexe = b.isexe;
	exports.sync = b.sync;
}));
//#endregion
//#region src/utils.ts
var import_lib = /* @__PURE__ */ __toESM((/* @__PURE__ */ __commonJSMin(((exports, module) => {
	const { isexe, sync: isexeSync } = require_index_min();
	const { join: join$1, delimiter, sep, posix } = __require("path");
	const isWindows = process.platform === "win32";
	/* istanbul ignore next */
	const rSlash = new RegExp(`[${posix.sep}${sep === posix.sep ? "" : sep}]`.replace(/(\\)/g, "\\$1"));
	const rRel = new RegExp(`^\\.${rSlash.source}`);
	const getNotFoundError = (cmd) => Object.assign(/* @__PURE__ */ new Error(`not found: ${cmd}`), { code: "ENOENT" });
	const getPathInfo = (cmd, { path: optPath = process.env.PATH, pathExt: optPathExt = process.env.PATHEXT, delimiter: optDelimiter = delimiter }) => {
		const pathEnv = cmd.match(rSlash) ? [""] : [...isWindows ? [process.cwd()] : [], ...(optPath || "").split(optDelimiter)];
		if (isWindows) {
			const pathExtExe = optPathExt || [
				".EXE",
				".CMD",
				".BAT",
				".COM"
			].join(optDelimiter);
			const pathExt = pathExtExe.split(optDelimiter).flatMap((item) => [item, item.toLowerCase()]);
			if (cmd.includes(".") && pathExt[0] !== "") pathExt.unshift("");
			return {
				pathEnv,
				pathExt,
				pathExtExe
			};
		}
		return {
			pathEnv,
			pathExt: [""]
		};
	};
	const getPathPart = (raw, cmd) => {
		const pathPart = /^".*"$/.test(raw) ? raw.slice(1, -1) : raw;
		return (!pathPart && rRel.test(cmd) ? cmd.slice(0, 2) : "") + join$1(pathPart, cmd);
	};
	const which = async (cmd, opt = {}) => {
		const { pathEnv, pathExt, pathExtExe } = getPathInfo(cmd, opt);
		const found = [];
		for (const envPart of pathEnv) {
			const p = getPathPart(envPart, cmd);
			for (const ext of pathExt) {
				const withExt = p + ext;
				if (await isexe(withExt, {
					pathExt: pathExtExe,
					ignoreErrors: true
				})) {
					if (!opt.all) return withExt;
					found.push(withExt);
				}
			}
		}
		if (opt.all && found.length) return found;
		if (opt.nothrow) return null;
		throw getNotFoundError(cmd);
	};
	const whichSync = (cmd, opt = {}) => {
		const { pathEnv, pathExt, pathExtExe } = getPathInfo(cmd, opt);
		const found = [];
		for (const pathEnvPart of pathEnv) {
			const p = getPathPart(pathEnvPart, cmd);
			for (const ext of pathExt) {
				const withExt = p + ext;
				if (isexeSync(withExt, {
					pathExt: pathExtExe,
					ignoreErrors: true
				})) {
					if (!opt.all) return withExt;
					found.push(withExt);
				}
			}
		}
		if (opt.all && found.length) return found;
		if (opt.nothrow) return null;
		throw getNotFoundError(cmd);
	};
	module.exports = which;
	which.sync = whichSync;
})))(), 1);
const CLI_TEMP_DIR = join(os.tmpdir(), "antfu-ni");
function remove(arr, v) {
	const index = arr.indexOf(v);
	if (index >= 0) arr.splice(index, 1);
	return arr;
}
function exclude(arr, ...v) {
	return arr.slice().filter((item) => !v.includes(item));
}
function cmdExists(cmd) {
	return import_lib.default.sync(cmd, { nothrow: true }) !== null;
}
let counter = 0;
async function openTemp() {
	if (!existsSync(CLI_TEMP_DIR)) await promises.mkdir(CLI_TEMP_DIR, { recursive: true });
	const competitivePath = join(CLI_TEMP_DIR, `.${process$1.pid}.${counter}`);
	counter += 1;
	return promises.open(competitivePath, "wx").then((fd) => ({
		fd,
		path: competitivePath,
		cleanup() {
			fd.close().then(() => {
				if (existsSync(competitivePath)) promises.unlink(competitivePath);
			});
		}
	})).catch((error) => {
		if (error && error.code === "EEXIST") return openTemp();
		else return void 0;
	});
}
/**
* Write file safely avoiding conflicts
*/
async function writeFileSafe(path, data = "") {
	const temp = await openTemp();
	if (temp) promises.writeFile(temp.path, data).then(() => {
		const directory = dirname(path);
		if (!existsSync(directory)) promises.mkdir(directory, { recursive: true });
		return promises.rename(temp.path, path).then(() => true).catch(() => false);
	}).catch(() => false).finally(temp.cleanup);
	return false;
}
function limitText(text, maxWidth) {
	if (text.length <= maxWidth) return text;
	return `${text.slice(0, maxWidth)}${styleText("dim", "…")}`;
}
function terminalLink(text, url, options) {
	if (process$1.env.FORCE_HYPERLINK || process$1.stdout.isTTY && !process$1.env.NO_HYPERLINK) return `\u001B]8;;${url}\u0007${text}\u001B]8;;\u0007`;
	if (options?.fallback) return options.fallback(text, url);
	return `${text} (${url})`;
}
function formatPackageWithUrl(pkg, url, limits = 80) {
	return url ? terminalLink(pkg, url, { fallback: (_, url) => pkg.length + url.length > limits ? pkg : `${pkg} ${styleText("dim", `- ${url}`)}` }) : pkg;
}
//#endregion
//#region src/detect.ts
var import_prompts = /* @__PURE__ */ __toESM(require_prompts(), 1);
async function detect$1({ autoInstall, programmatic, cwd } = {}) {
	const targetDir = cwd ?? process$1.cwd();
	if (existsSync(path.join(targetDir, "deno.json")) || existsSync(path.join(targetDir, "deno.jsonc"))) return "deno";
	const { name, agent, version } = await detect({
		cwd,
		onUnknown: (packageManager) => {
			if (!programmatic) console.warn("[ni] Unknown packageManager:", packageManager);
		}
	}) || {};
	if (name && !cmdExists(name) && !programmatic) {
		if (!autoInstall) {
			console.warn(`[ni] Detected ${name} but it doesn't seem to be installed.\n`);
			if (process$1.env.CI) process$1.exit(1);
			const { tryInstall } = await (0, import_prompts.default)({
				name: "tryInstall",
				type: "confirm",
				message: `Would you like to globally install ${terminalLink(name, INSTALL_PAGE[name])}?`
			});
			if (!tryInstall) process$1.exit(1);
		}
		await x("npm", [
			"i",
			"-g",
			`${name}${version ? `@${version}` : ""}`
		], {
			nodeOptions: {
				stdio: "inherit",
				cwd
			},
			throwOnError: true
		});
	}
	return agent;
}
//#endregion
//#region src/config.ts
var import_ini = /* @__PURE__ */ __toESM(require_ini(), 1);
const customRcPath = process$1.env.NI_CONFIG_FILE;
const home = process$1.platform === "win32" ? process$1.env.USERPROFILE : process$1.env.HOME;
const defaultRcPath = path.join(home || "~/", ".nirc");
const rcPath = customRcPath || defaultRcPath;
const defaultConfig = {
	defaultAgent: "prompt",
	globalAgent: "npm",
	runAgent: void 0,
	useSfw: false,
	catalog: true,
	noLastCommand: false
};
let config;
async function getConfig() {
	if (!config) {
		config = {
			...defaultConfig,
			...fs.existsSync(rcPath) ? import_ini.default.parse(fs.readFileSync(rcPath, "utf-8")) : null
		};
		if (process$1.env.NI_DEFAULT_AGENT) config.defaultAgent = process$1.env.NI_DEFAULT_AGENT;
		if (process$1.env.NI_GLOBAL_AGENT) config.globalAgent = process$1.env.NI_GLOBAL_AGENT;
		if (process$1.env.NI_RUN_AGENT === "node") config.runAgent = process$1.env.NI_RUN_AGENT;
		if (process$1.env.NI_USE_SFW !== void 0) config.useSfw = process$1.env.NI_USE_SFW === "true";
		if (process$1.env.NI_CATALOG !== void 0) config.catalog = process$1.env.NI_CATALOG !== "false";
		if (process$1.env.NI_NO_LAST_COMMAND !== void 0) config.noLastCommand = process$1.env.NI_NO_LAST_COMMAND === "true";
		const agent = await detect$1({ programmatic: true });
		if (agent) config.defaultAgent = agent;
	}
	return config;
}
async function getDefaultAgent(programmatic) {
	const { defaultAgent } = await getConfig();
	if (defaultAgent === "prompt" && (programmatic || process$1.env.CI)) return "npm";
	return defaultAgent;
}
async function getGlobalAgent() {
	const { globalAgent } = await getConfig();
	return globalAgent;
}
async function getRunAgent() {
	const { runAgent } = await getConfig();
	return runAgent;
}
async function getUseSfw() {
	const { useSfw } = await getConfig();
	return useSfw;
}
async function getCatalog() {
	const { catalog } = await getConfig();
	return catalog;
}
//#endregion
//#region src/parse.ts
var UnsupportedCommand = class extends Error {
	constructor({ agent, command }) {
		super(`Command "${command}" is not support by agent "${agent}"`);
	}
};
function getCommand(agent, command, args = []) {
	if (!src_exports.COMMANDS[agent]) throw new Error(`Unsupported agent "${agent}"`);
	if (!src_exports.COMMANDS[agent][command]) throw new UnsupportedCommand({
		agent,
		command
	});
	return (0, src_exports.constructCommand)(src_exports.COMMANDS[agent][command], args);
}
const parseNi = ((agent, args, ctx) => {
	if (agent === "bun") args = args.map((i) => i === "-D" ? "-d" : i);
	if (agent === "npm") args = args.map((i) => i === "-P" ? "--omit=dev" : i);
	if (args.includes("-P")) args = args.map((i) => i === "-P" ? "--production" : i);
	if (args.includes("-g")) return getCommand(agent, "global", exclude(args, "-g"));
	if (args.includes("--frozen-if-present")) {
		args = exclude(args, "--frozen-if-present");
		return getCommand(agent, ctx?.hasLock ? "frozen" : "install", args);
	}
	if (args.includes("--frozen")) return getCommand(agent, "frozen", exclude(args, "--frozen"));
	if (args.length === 0 || args.every((i) => i.startsWith("-"))) return getCommand(agent, "install", args);
	return getCommand(agent, "add", args);
});
const parseNr = (async (agent, args, ctx) => {
	if (args.length === 0) args.push("start");
	const runAgent = await getRunAgent();
	let runWithNode = false;
	if (runAgent === "node") {
		const [majorNodeVersion] = process$1.versions.node.split(".").map(Number);
		if (majorNodeVersion < 22) throw new Error("The runAgent \"node\" requires Node.js 22.0.0 or higher");
		runWithNode = true;
		args = ["--run", ...args];
	}
	let hasIfPresent = false;
	if (args.includes("--if-present")) {
		args = exclude(args, "--if-present");
		hasIfPresent = true;
	}
	if (args.includes("-p")) args = exclude(args, "-p");
	const processedArgs = [];
	let i = 0;
	while (i < args.length) {
		const arg = args[i];
		if ((arg === "-w" || arg === "--workspace") && i + 1 < args.length && !args[i + 1].startsWith("-")) {
			processedArgs.push(`${arg}=${args[i + 1]}`);
			i += 2;
		} else {
			processedArgs.push(arg);
			i++;
		}
	}
	args = processedArgs;
	const cmd = runWithNode ? {
		command: "node",
		args
	} : getCommand(agent, "run", args);
	if (ctx?.cwd) cmd.cwd = ctx.cwd;
	if (!cmd) return cmd;
	if (hasIfPresent && !runWithNode) cmd.args.splice(1, 0, "--if-present");
	return cmd;
});
const parseNup = ((agent, args) => {
	if (args.includes("-i")) return getCommand(agent, "upgrade-interactive", exclude(args, "-i"));
	return getCommand(agent, "upgrade", args);
});
const parseNd = ((agent, args) => {
	if (agent === "pnpm") args = args.map((i) => i === "-c" ? "--check" : i);
	if (agent === "npm") args = args.map((i) => i === "-c" ? "--dry-run" : i);
	return getCommand(agent, "dedupe", args);
});
const parseNun = ((agent, args) => {
	if (args.includes("-g")) return getCommand(agent, "global_uninstall", exclude(args, "-g"));
	return getCommand(agent, "uninstall", args);
});
const parseNlx = ((agent, args) => {
	return getCommand(agent, "execute", args);
});
const parseNa = ((agent, args) => {
	return getCommand(agent, "agent", args);
});
function serializeCommand(command) {
	if (!command) return void 0;
	if (command.args.length === 0) return command.command;
	return `${command.command} ${command.args.map((i) => i.includes(" ") ? `"${i}"` : i).join(" ")}`;
}
//#endregion
//#region package.json
var version = "30.1.0";
//#endregion
//#region src/environment.ts
const DEFAULT_ENVIRONMENT_OPTIONS = { autoInstall: false };
function getEnvironmentOptions() {
	return {
		...DEFAULT_ENVIRONMENT_OPTIONS,
		autoInstall: process$1.env.NI_AUTO_INSTALL === "true"
	};
}
//#endregion
//#region src/runner.ts
const DEBUG_SIGN = "?";
const PROGRAMMATIC_SIGN = "--programmatic";
async function runCli(fn, options = {}) {
	options = {
		...getEnvironmentOptions(),
		...options
	};
	const { args = process$1.argv.slice(2).filter(Boolean) } = options;
	try {
		await run(fn, args, options);
	} catch (error) {
		if (error instanceof UnsupportedCommand && !options.programmatic) console.log(styleText("red", `\u2717 ${error.message}`));
		if (!options.programmatic) process$1.exit(1);
		throw error;
	}
}
async function getCliCommand(fn, args, options = {}, cwd = options.cwd ?? process$1.cwd()) {
	if (args.includes("-g")) return await fn(await getGlobalAgent(), args);
	let agent = await detect$1({
		...options,
		cwd
	}) || await getDefaultAgent(options.programmatic);
	if (agent === "prompt") {
		agent = (await (0, import_prompts.default)({
			name: "agent",
			type: "select",
			message: "Choose the agent",
			choices: AGENTS.filter((i) => !i.includes("@")).map((value) => ({
				title: value,
				value
			}))
		})).agent;
		if (!agent) return;
	}
	return await fn(agent, args, {
		programmatic: options.programmatic,
		hasLock: Boolean(agent),
		cwd
	});
}
async function run(fn, args, options = {}) {
	const { detectVolta = true } = options;
	const debug = args.includes(DEBUG_SIGN);
	if (debug) remove(args, DEBUG_SIGN);
	const programmaticFromArgs = args.includes(PROGRAMMATIC_SIGN);
	if (programmaticFromArgs) remove(args, PROGRAMMATIC_SIGN);
	const programmatic = options.programmatic || programmaticFromArgs;
	let cwd = options.cwd ?? process$1.cwd();
	if (args[0] === "-C") {
		cwd = resolve(cwd, args[1]);
		args.splice(0, 2);
	}
	if (args.length === 1 && (args[0]?.toLowerCase() === "-v" || args[0] === "--version")) {
		const getCmd = (a) => AGENTS.includes(a) ? getCommand(a, "agent", ["-v"]) : {
			command: a,
			args: ["-v"]
		};
		const xVersionOptions = {
			nodeOptions: { cwd },
			throwOnError: true
		};
		const getV = (a) => {
			const { command, args } = getCmd(a);
			return x(command, args, xVersionOptions).then((e) => e.stdout).then((e) => e.startsWith("v") ? e : `v${e}`);
		};
		const globalAgentPromise = getGlobalAgent();
		const globalAgentVersionPromise = globalAgentPromise.then(getV);
		const agentPromise = detect$1({
			...options,
			cwd
		}).then((a) => a || "");
		const agentVersionPromise = agentPromise.then((a) => a && getV(a));
		const nodeVersionPromise = getV("node");
		console.log(`@antfu/ni  ${styleText("cyan", `v${version}`)}`);
		console.log(`node       ${styleText("green", await nodeVersionPromise)}`);
		const [agent, agentVersion] = await Promise.all([agentPromise, agentVersionPromise]);
		if (agent) console.log(`${agent.padEnd(10)} ${styleText("blue", agentVersion)}`);
		else console.log("agent      no lock file");
		const [globalAgent, globalAgentVersion] = await Promise.all([globalAgentPromise, globalAgentVersionPromise]);
		console.log(`${`${globalAgent} -g`.padEnd(10)} ${styleText("blue", globalAgentVersion)}`);
		return;
	}
	if (args.length === 1 && args[0] === "--agent") {
		const agent = args.includes("-g") ? await getGlobalAgent() : await detect$1({
			...options,
			cwd
		}) || await getDefaultAgent(programmatic);
		if (agent && agent !== "prompt") process$1.stdout.write(`${agent}\n`);
		else process$1.stdout.write("unknown\n");
		return;
	}
	if (args.length === 1 && ["-h", "--help"].includes(args[0])) {
		const dash = styleText("dim", "-");
		console.log(`${styleText(["green", "bold"], "@antfu/ni")} ${styleText("dim", `use the right package manager v${version}`)}\n`);
		console.log(`ni    ${dash}  install`);
		console.log(`nr    ${dash}  run`);
		console.log(`nlx   ${dash}  execute`);
		console.log(`nup   ${dash}  upgrade`);
		console.log(`nun   ${dash}  uninstall`);
		console.log(`nci   ${dash}  clean install`);
		console.log(`na    ${dash}  agent alias`);
		console.log(`nd    ${dash}  dedupe dependencies`);
		console.log(`ni -v       ${dash}  show used agent`);
		console.log(`ni --agent  ${dash}  print detected agent name (for scripting)`);
		console.log(`ni -i ${dash}  interactive package management`);
		console.log(`ni ?  ${dash}  dry run (print the resolved command)`);
		console.log(styleText("yellow", "\ncheck https://github.com/antfu/ni for more documentation."));
		return;
	}
	if (options.onBeforeCommand) {
		let shouldExit = false;
		await options.onBeforeCommand(args, {
			cwd,
			programmatic,
			exit: () => {
				shouldExit = true;
			}
		});
		if (shouldExit) return;
	}
	const command = await getCliCommand(fn, args, {
		...options,
		programmatic
	}, cwd);
	if (!command) return;
	const useSfw = await getUseSfw();
	if (useSfw && cmdExists("sfw")) {
		command.args = [command.command, ...command.args];
		command.command = "sfw";
	} else if (useSfw) {
		if (programmatic) throw new Error("sfw is enabled but not installed");
		console.error("[ni] sfw is enabled but not installed.");
		console.error("[ni] Install it with: npm install -g sfw");
		process$1.exit(1);
	}
	if (detectVolta && cmdExists("volta")) {
		command.args = [
			"run",
			command.command,
			...command.args
		];
		command.command = "volta";
	}
	if (debug) {
		const commandStr = [command.command, ...command.args].join(" ");
		console.log(commandStr);
		return;
	}
	const proc = x(command.command, command.args, {
		nodeOptions: {
			stdio: "inherit",
			cwd: command.cwd ?? cwd
		},
		throwOnError: true
	});
	process$1.once("SIGINT", async () => {
		await proc;
		process$1.exit(proc.exitCode);
	});
	await proc;
}
//#endregion
//#region src/index.ts
var src_exports = /* @__PURE__ */ __exportAll({
	CLI_TEMP_DIR: () => CLI_TEMP_DIR,
	UnsupportedCommand: () => UnsupportedCommand,
	cmdExists: () => cmdExists,
	detect: () => detect$1,
	exclude: () => exclude,
	formatPackageWithUrl: () => formatPackageWithUrl,
	getCatalog: () => getCatalog,
	getCliCommand: () => getCliCommand,
	getCommand: () => getCommand,
	getConfig: () => getConfig,
	getDefaultAgent: () => getDefaultAgent,
	getGlobalAgent: () => getGlobalAgent,
	getRunAgent: () => getRunAgent,
	getUseSfw: () => getUseSfw,
	limitText: () => limitText,
	parseNa: () => parseNa,
	parseNd: () => parseNd,
	parseNi: () => parseNi,
	parseNlx: () => parseNlx,
	parseNr: () => parseNr,
	parseNun: () => parseNun,
	parseNup: () => parseNup,
	remove: () => remove,
	run: () => run,
	runCli: () => runCli,
	serializeCommand: () => serializeCommand,
	terminalLink: () => terminalLink,
	writeFileSafe: () => writeFileSafe
});
import * as import_package_manager_detector_commands from "package-manager-detector/commands";
__reExport(src_exports, import_package_manager_detector_commands);
import * as import_package_manager_detector_constants from "package-manager-detector/constants";
__reExport(src_exports, import_package_manager_detector_constants);
//#endregion
export { require_prompts as A, cmdExists as C, remove as D, limitText as E, terminalLink as O, CLI_TEMP_DIR as S, formatPackageWithUrl as T, getDefaultAgent as _, UnsupportedCommand as a, getUseSfw as b, parseNd as c, parseNr as d, parseNun as f, getConfig as g, getCatalog as h, runCli as i, writeFileSafe as k, parseNi as l, serializeCommand as m, getCliCommand as n, getCommand as o, parseNup as p, run as r, parseNa as s, src_exports as t, parseNlx as u, getGlobalAgent as v, exclude as w, detect$1 as x, getRunAgent as y };
