// devframe@0.5.4/dist/rpc/transports/ws-server.mjs
import { n as strictJsonStringify } from "../../serialization-BD_qXGjd.mjs";
import { n as structuredCloneStringify, t as structuredCloneParse } from "../../structured-clone-jegjz0hM.mjs";
import { createServer } from "node:https";
import { WebSocketServer } from "ws";
//#region src/rpc/transports/ws-server.ts
let sessionId = 0;
const EMPTY_DEFS = /* @__PURE__ */ new Map();
function NOOP() {}
/**
* Attach a WebSocket transport to an existing RPC group. Either pass an
* existing `WebSocketServer` via `wss`, or let this helper create one from
* `port` / `host` / `https`.
*/
function attachWsRpcTransport(rpcGroup, options = {}) {
	const { wss: externalWss, port, host = "localhost", https, onConnected = NOOP, onDisconnected = NOOP, definitions = EMPTY_DEFS, serialize: serializeOverride, deserialize: deserializeOverride } = options;
	let wss;
	if (externalWss) wss = externalWss;
	else if (https) {
		const httpsServer = createServer(https);
		wss = new WebSocketServer({ server: httpsServer });
		httpsServer.listen(port, host);
	} else wss = new WebSocketServer({
		port,
		host
	});
	wss.on("connection", (ws, req) => {
		const meta = {
			id: sessionId++,
			ws,
			subscribedStates: /* @__PURE__ */ new Set()
		};
		const pendingRequestMethods = /* @__PURE__ */ new Map();
		const channel = {
			post: (data) => {
				ws.send(data);
			},
			on: (fn) => {
				ws.on("message", (data) => {
					fn(data.toString());
				});
			},
			serialize: serializeOverride ?? ((msg) => {
				let method;
				if (msg.t === "q") method = msg.m;
				else {
					method = pendingRequestMethods.get(msg.i);
					pendingRequestMethods.delete(msg.i);
				}
				if (!(msg.t === "s" && "e" in msg) && !!method && definitions.get(method)?.jsonSerializable === true) return strictJsonStringify(msg, method ?? "");
				return `s:${structuredCloneStringify(msg)}`;
			}),
			deserialize: deserializeOverride ?? ((raw) => {
				const msg = raw.startsWith("s:") ? structuredCloneParse(raw.slice(2)) : JSON.parse(raw);
				if (msg.t === "q" && msg.i && msg.m) pendingRequestMethods.set(msg.i, msg.m);
				return msg;
			}),
			meta
		};
		rpcGroup.updateChannels((channels) => {
			channels.push(channel);
		});
		ws.on("close", () => {
			rpcGroup.updateChannels((channels) => {
				const index = channels.indexOf(channel);
				if (index >= 0) channels.splice(index, 1);
			});
			onDisconnected(ws, meta);
		});
		onConnected(ws, req, meta);
	});
	return { wss };
}
//#endregion
export { attachWsRpcTransport };
