// devframe@0.5.4/dist/dump-9lKIJTLh.mjs
import { t as diagnostics } from "./diagnostics-GitRGKbr.mjs";
import { DEVFRAME_RPC_DUMP_DIRNAME } from "./constants.mjs";
import { createHash } from "node:crypto";
//#region ../../node_modules/.pnpm/ohash@2.0.11/node_modules/ohash/dist/shared/ohash.D__AXeF1.mjs
function serialize(o) {
	return typeof o == "string" ? `'${o}'` : new c().serialize(o);
}
const c = /* @__PURE__ */ function() {
	class o {
		#t = /* @__PURE__ */ new Map();
		compare(t, r) {
			const e = typeof t, n = typeof r;
			return e === "string" && n === "string" ? t.localeCompare(r) : e === "number" && n === "number" ? t - r : String.prototype.localeCompare.call(this.serialize(t, true), this.serialize(r, true));
		}
		serialize(t, r) {
			if (t === null) return "null";
			switch (typeof t) {
				case "string": return r ? t : `'${t}'`;
				case "bigint": return `${t}n`;
				case "object": return this.$object(t);
				case "function": return this.$function(t);
			}
			return String(t);
		}
		serializeObject(t) {
			const r = Object.prototype.toString.call(t);
			if (r !== "[object Object]") return this.serializeBuiltInType(r.length < 10 ? `unknown:${r}` : r.slice(8, -1), t);
			const e = t.constructor, n = e === Object || e === void 0 ? "" : e.name;
			if (n !== "" && globalThis[n] === e) return this.serializeBuiltInType(n, t);
			if (typeof t.toJSON == "function") {
				const i = t.toJSON();
				return n + (i !== null && typeof i == "object" ? this.$object(i) : `(${this.serialize(i)})`);
			}
			return this.serializeObjectEntries(n, Object.entries(t));
		}
		serializeBuiltInType(t, r) {
			const e = this["$" + t];
			if (e) return e.call(this, r);
			if (typeof r?.entries == "function") return this.serializeObjectEntries(t, r.entries());
			throw new Error(`Cannot serialize ${t}`);
		}
		serializeObjectEntries(t, r) {
			const e = Array.from(r).sort((i, a) => this.compare(i[0], a[0]));
			let n = `${t}{`;
			for (let i = 0; i < e.length; i++) {
				const [a, l] = e[i];
				n += `${this.serialize(a, true)}:${this.serialize(l)}`, i < e.length - 1 && (n += ",");
			}
			return n + "}";
		}
		$object(t) {
			let r = this.#t.get(t);
			return r === void 0 && (this.#t.set(t, `#${this.#t.size}`), r = this.serializeObject(t), this.#t.set(t, r)), r;
		}
		$function(t) {
			const r = Function.prototype.toString.call(t);
			return r.slice(-15) === "[native code] }" ? `${t.name || ""}()[native]` : `${t.name}(${t.length})${r.replace(/\s*\n\s*/g, "")}`;
		}
		$Array(t) {
			let r = "[";
			for (let e = 0; e < t.length; e++) r += this.serialize(t[e]), e < t.length - 1 && (r += ",");
			return r + "]";
		}
		$Date(t) {
			try {
				return `Date(${t.toISOString()})`;
			} catch {
				return "Date(null)";
			}
		}
		$ArrayBuffer(t) {
			return `ArrayBuffer[${new Uint8Array(t).join(",")}]`;
		}
		$Set(t) {
			return `Set${this.$Array(Array.from(t).sort((r, e) => this.compare(r, e)))}`;
		}
		$Map(t) {
			return this.serializeObjectEntries("Map", t.entries());
		}
	}
	for (const s of [
		"Error",
		"RegExp",
		"URL"
	]) o.prototype["$" + s] = function(t) {
		return `${s}(${t})`;
	};
	for (const s of [
		"Int8Array",
		"Uint8Array",
		"Uint8ClampedArray",
		"Int16Array",
		"Uint16Array",
		"Int32Array",
		"Uint32Array",
		"Float32Array",
		"Float64Array"
	]) o.prototype["$" + s] = function(t) {
		return `${s}[${t.join(",")}]`;
	};
	for (const s of ["BigInt64Array", "BigUint64Array"]) o.prototype["$" + s] = function(t) {
		return `${s}[${t.join("n,")}${t.length > 0 ? "n" : ""}]`;
	};
	return o;
}();
//#endregion
//#region ../../node_modules/.pnpm/ohash@2.0.11/node_modules/ohash/dist/crypto/node/index.mjs
const e = globalThis.process?.getBuiltinModule?.("crypto")?.hash, r = "sha256", s = "base64url";
function digest(t) {
	if (e) return e(r, t, s);
	const o = createHash(r).update(t);
	return globalThis.process?.versions?.webcontainer ? o.digest().toString(s) : o.digest(s);
}
//#endregion
//#region ../../node_modules/.pnpm/ohash@2.0.11/node_modules/ohash/dist/index.mjs
function hash$1(input) {
	return digest(serialize(input));
}
//#endregion
//#region src/utils/hash.ts
/**
* Stable, deterministic hash of any structured-cloneable value.
*/
function hash(value) {
	return hash$1(value);
}
//#endregion
//#region ../../node_modules/.pnpm/yocto-queue@1.2.2/node_modules/yocto-queue/index.js
var Node = class {
	value;
	next;
	constructor(value) {
		this.value = value;
	}
};
var Queue = class {
	#head;
	#tail;
	#size;
	constructor() {
		this.clear();
	}
	enqueue(value) {
		const node = new Node(value);
		if (this.#head) {
			this.#tail.next = node;
			this.#tail = node;
		} else {
			this.#head = node;
			this.#tail = node;
		}
		this.#size++;
	}
	dequeue() {
		const current = this.#head;
		if (!current) return;
		this.#head = this.#head.next;
		this.#size--;
		if (!this.#head) this.#tail = void 0;
		return current.value;
	}
	peek() {
		if (!this.#head) return;
		return this.#head.value;
	}
	clear() {
		this.#head = void 0;
		this.#tail = void 0;
		this.#size = 0;
	}
	get size() {
		return this.#size;
	}
	*[Symbol.iterator]() {
		let current = this.#head;
		while (current) {
			yield current.value;
			current = current.next;
		}
	}
	*drain() {
		while (this.#head) yield this.dequeue();
	}
};
//#endregion
//#region ../../node_modules/.pnpm/p-limit@7.3.0/node_modules/p-limit/index.js
function pLimit(concurrency) {
	let rejectOnClear = false;
	if (typeof concurrency === "object") ({concurrency, rejectOnClear = false} = concurrency);
	validateConcurrency(concurrency);
	if (typeof rejectOnClear !== "boolean") throw new TypeError("Expected `rejectOnClear` to be a boolean");
	const queue = new Queue();
	let activeCount = 0;
	const resumeNext = () => {
		if (activeCount < concurrency && queue.size > 0) {
			activeCount++;
			queue.dequeue().run();
		}
	};
	const next = () => {
		activeCount--;
		resumeNext();
	};
	const run = async (function_, resolve, arguments_) => {
		const result = (async () => function_(...arguments_))();
		resolve(result);
		try {
			await result;
		} catch {}
		next();
	};
	const enqueue = (function_, resolve, reject, arguments_) => {
		const queueItem = { reject };
		new Promise((internalResolve) => {
			queueItem.run = internalResolve;
			queue.enqueue(queueItem);
		}).then(run.bind(void 0, function_, resolve, arguments_));
		if (activeCount < concurrency) resumeNext();
	};
	const generator = (function_, ...arguments_) => new Promise((resolve, reject) => {
		enqueue(function_, resolve, reject, arguments_);
	});
	Object.defineProperties(generator, {
		activeCount: { get: () => activeCount },
		pendingCount: { get: () => queue.size },
		clearQueue: { value() {
			if (!rejectOnClear) {
				queue.clear();
				return;
			}
			const abortError = AbortSignal.abort().reason;
			while (queue.size > 0) queue.dequeue().reject(abortError);
		} },
		concurrency: {
			get: () => concurrency,
			set(newConcurrency) {
				validateConcurrency(newConcurrency);
				concurrency = newConcurrency;
				queueMicrotask(() => {
					while (activeCount < concurrency && queue.size > 0) resumeNext();
				});
			}
		},
		map: { async value(iterable, function_) {
			const promises = Array.from(iterable, (value, index) => this(function_, value, index));
			return Promise.all(promises);
		} }
	});
	return generator;
}
function validateConcurrency(concurrency) {
	if (!((Number.isInteger(concurrency) || concurrency === Number.POSITIVE_INFINITY) && concurrency > 0)) throw new TypeError("Expected `concurrency` to be a number from 1 and up");
}
//#endregion
//#region src/rpc/validation.ts
/**
* Validates RPC function definitions.
* Action and event functions cannot have dumps (side effects should not be cached).
*
* @throws {Error} If an action or event function has a dump configuration
*/
function validateDefinitions(definitions) {
	for (const definition of definitions) {
		const type = definition.type || "query";
		if ((type === "action" || type === "event") && definition.dump) throw diagnostics.DF0027({
			name: definition.name,
			type
		});
		if (definition.snapshot && type !== "query") throw diagnostics.DF0028({
			name: definition.name,
			type
		});
	}
}
/**
* Validates a single RPC function definition.
*
* @throws {Error} If an action or event function has a dump configuration
*/
function validateDefinition(definition) {
	validateDefinitions([definition]);
}
//#endregion
//#region src/rpc/dump/error.ts
/**
* Normalize a thrown value into a plain object suitable for storage in
* a dump record. Preserves `message`, `name`, `cause`, and any own
* enumerable properties of an `Error` so consumers reading the dump can
* reconstruct a richer Error than just `{ message, name }`.
*
* Non-`Error` throws are wrapped as `{ name: 'Error', message: String(thrown) }`.
*/
function serializeDumpError(error) {
	return serializeWithSeen(error, /* @__PURE__ */ new WeakSet());
}
function serializeWithSeen(error, seen) {
	if (!(error instanceof Error)) return {
		name: "Error",
		message: String(error)
	};
	if (seen.has(error)) return {
		name: error.name,
		message: error.message
	};
	seen.add(error);
	const out = {
		name: error.name,
		message: error.message
	};
	const cause = error.cause;
	if (cause !== void 0) out.cause = cause instanceof Error ? serializeWithSeen(cause, seen) : cause;
	for (const key of Object.keys(error)) {
		if (key === "name" || key === "message" || key === "cause") continue;
		out[key] = error[key];
	}
	return out;
}
/**
* Inverse of {@link serializeDumpError}: rebuild a thrown `Error` from
* the plain object stored in a dump record. Preserves `cause`, restores
* the original `name`, and re-attaches any custom own properties.
*/
function reviveDumpError(stored) {
	const cause = stored.cause instanceof Error ? stored.cause : isPlainErrorShape(stored.cause) ? reviveDumpError(stored.cause) : stored.cause;
	const error = cause !== void 0 ? new Error(stored.message, { cause }) : new Error(stored.message);
	error.name = stored.name;
	for (const key of Object.keys(stored)) {
		if (key === "name" || key === "message" || key === "cause") continue;
		error[key] = stored[key];
	}
	return error;
}
function isPlainErrorShape(value) {
	return typeof value === "object" && value !== null && typeof value.message === "string" && typeof value.name === "string";
}
//#endregion
//#region src/rpc/dump/collect.ts
function getDumpRecordKey(functionName, args) {
	return `${functionName}---${hash(args)}`;
}
function getDumpFallbackKey(functionName) {
	return `${functionName}---fallback`;
}
async function resolveGetter(valueOrGetter) {
	return typeof valueOrGetter === "function" ? await valueOrGetter() : valueOrGetter;
}
/**
* Collects pre-computed dumps by executing functions with their defined input combinations.
* Static functions without dump config automatically get `{ inputs: [[]] }`.
*
* @example
* ```ts
* const store = await dumpFunctions([greet], context, { concurrency: 10 })
* ```
*/
async function dumpFunctions(definitions, context, options = {}) {
	validateDefinitions(definitions);
	const concurrency = options.concurrency === true ? 5 : options.concurrency === false || options.concurrency == null ? 1 : options.concurrency;
	const store = {
		definitions: {},
		records: {}
	};
	const tasksResolutions = definitions.map((definition) => async () => {
		if (definition.type === "event" || definition.type === "action") return;
		const setupResult = definition.setup ? await Promise.resolve(definition.setup(context)) : {};
		const handler = setupResult.handler || definition.handler;
		if (!handler) throw diagnostics.DF0024({ name: definition.name });
		let dump = setupResult.dump ?? definition.dump;
		if (!dump && definition.type === "static") dump = { inputs: [[]] };
		if (!dump && definition.snapshot) dump = async (_ctx, h) => {
			const output = await Promise.resolve(h(...[]));
			return {
				records: [{
					inputs: [],
					output
				}],
				fallback: output
			};
		};
		if (!dump) return;
		if (typeof dump === "function") dump = await Promise.resolve(dump(context, handler));
		store.definitions[definition.name] = {
			name: definition.name,
			type: definition.type
		};
		return {
			handler,
			dump,
			definition
		};
	});
	let functionsToDump = [];
	if (concurrency <= 1) for (const task of tasksResolutions) {
		const resolution = await task();
		if (resolution) functionsToDump.push(resolution);
	}
	else {
		const limit = pLimit(concurrency);
		functionsToDump = (await Promise.all(tasksResolutions.map((task) => limit(task)))).filter((x) => !!x);
	}
	const dumpTasks = [];
	for (const { definition, handler, dump } of functionsToDump) {
		const { inputs, records, fallback } = dump;
		if (records) for (const record of records) {
			const recordKey = getDumpRecordKey(definition.name, record.inputs);
			store.records[recordKey] = record;
		}
		if ("fallback" in dump) {
			const fallbackKey = getDumpFallbackKey(definition.name);
			store.records[fallbackKey] = {
				inputs: [],
				output: fallback
			};
		}
		if (inputs) for (const input of inputs) dumpTasks.push(async () => {
			const recordKey = getDumpRecordKey(definition.name, input);
			try {
				const output = await Promise.resolve(handler(...input));
				store.records[recordKey] = {
					inputs: input,
					output
				};
			} catch (error) {
				store.records[recordKey] = {
					inputs: input,
					error: serializeDumpError(error)
				};
			}
		});
	}
	if (concurrency <= 1) for (const task of dumpTasks) await task();
	else {
		const limit = pLimit(concurrency);
		await Promise.all(dumpTasks.map((task) => limit(task)));
	}
	return store;
}
/**
* Creates a client that serves pre-computed results from a dump store.
* Uses argument hashing to match calls to stored records.
*
* @example
* ```ts
* const client = createClientFromDump(store)
* await client.greet('Alice')
* ```
*/
function createClientFromDump(store, options = {}) {
	const { onMiss } = options;
	return new Proxy({}, {
		get(_, functionName) {
			if (!(functionName in store.definitions)) throw diagnostics.DF0025({ name: functionName });
			return async (...args) => {
				const recordKey = getDumpRecordKey(functionName, args);
				const recordOrGetter = store.records[recordKey];
				if (recordOrGetter) {
					const record = await resolveGetter(recordOrGetter);
					if (record.error) throw reviveDumpError(record.error);
					if (typeof record.output === "function") return await record.output();
					return record.output;
				}
				onMiss?.(functionName, args);
				const fallbackKey = getDumpFallbackKey(functionName);
				if (fallbackKey in store.records) {
					const fallbackOrGetter = store.records[fallbackKey];
					const fallbackRecord = await resolveGetter(fallbackOrGetter);
					if (fallbackRecord && typeof fallbackRecord.output === "function") return await fallbackRecord.output();
					if (fallbackRecord) return fallbackRecord.output;
				}
				throw diagnostics.DF0026({
					name: functionName,
					args: JSON.stringify(args)
				});
			};
		},
		has(_, functionName) {
			return functionName in store.definitions;
		},
		ownKeys() {
			return Object.keys(store.definitions);
		},
		getOwnPropertyDescriptor(_, functionName) {
			return functionName in store.definitions ? {
				configurable: true,
				enumerable: true,
				value: void 0
			} : void 0;
		}
	});
}
/**
* Filters function definitions to only those with dump definitions.
* Note: Only checks the definition itself, not setup results.
*/
function getDefinitionsWithDumps(definitions) {
	return definitions.filter((def) => def.dump !== void 0);
}
//#endregion
//#region src/rpc/handler.ts
async function getRpcResolvedSetupResult(definition, context) {
	if (!definition.setup) return {};
	if (typeof context === "object" && context !== null) {
		definition.__cache ??= /* @__PURE__ */ new WeakMap();
		let promise = definition.__cache.get(context);
		if (!promise) {
			promise = Promise.resolve(definition.setup(context));
			definition.__cache.set(context, promise);
		}
		return await promise;
	}
	definition.__promise ??= Promise.resolve(definition.setup(context));
	return await definition.__promise;
}
async function getRpcHandler(definition, context) {
	if (definition.handler) return definition.handler;
	const result = await getRpcResolvedSetupResult(definition, context);
	if (!result.handler) throw diagnostics.DF0024({ name: definition.name });
	return result.handler;
}
//#endregion
//#region src/rpc/dump/static.ts
function makeDumpKey(name) {
	return encodeURIComponent(name.replaceAll(":", "~"));
}
function makeStaticPath(name) {
	return `${DEVFRAME_RPC_DUMP_DIRNAME}/${makeDumpKey(name)}.static.json`;
}
function makeQueryRecordPath(name, hash) {
	return `${DEVFRAME_RPC_DUMP_DIRNAME}/${makeDumpKey(name)}.record.${hash}.json`;
}
function makeQueryFallbackPath(name) {
	return `${DEVFRAME_RPC_DUMP_DIRNAME}/${makeDumpKey(name)}.fallback.json`;
}
async function resolveRecord(record) {
	return typeof record === "function" ? await record() : record;
}
async function collectStaticRpcDump(definitions, context) {
	const manifest = {};
	const files = {};
	for (const definition of definitions) {
		const type = definition.type ?? "query";
		const serialization = definition.jsonSerializable === true ? "json" : "structured-clone";
		if (type === "static") {
			const handler = await getRpcHandler(definition, context);
			const path = makeStaticPath(definition.name);
			files[path] = {
				serialization,
				fnName: definition.name,
				data: { output: await Promise.resolve(handler()) }
			};
			manifest[definition.name] = {
				type: "static",
				path,
				serialization
			};
			continue;
		}
		if (type !== "query") continue;
		const store = await dumpFunctions([definition], context);
		if (!(definition.name in store.definitions)) continue;
		const queryEntry = {
			type: "query",
			records: {},
			serialization
		};
		const prefix = `${definition.name}---`;
		for (const [recordKey, recordOrGetter] of Object.entries(store.records)) {
			if (!recordKey.startsWith(prefix)) continue;
			const key = recordKey.slice(prefix.length);
			const record = await resolveRecord(recordOrGetter);
			if (key === "fallback") {
				const path = makeQueryFallbackPath(definition.name);
				files[path] = {
					serialization,
					fnName: definition.name,
					data: record
				};
				queryEntry.fallback = path;
			} else {
				const path = makeQueryRecordPath(definition.name, key);
				files[path] = {
					serialization,
					fnName: definition.name,
					data: record
				};
				queryEntry.records[key] = path;
			}
		}
		if (!Object.keys(queryEntry.records).length && !queryEntry.fallback) continue;
		manifest[definition.name] = queryEntry;
	}
	return {
		manifest,
		files
	};
}
//#endregion
export { dumpFunctions as a, serializeDumpError as c, hash as d, createClientFromDump as i, validateDefinition as l, getRpcHandler as n, getDefinitionsWithDumps as o, getRpcResolvedSetupResult as r, reviveDumpError as s, collectStaticRpcDump as t, validateDefinitions as u };
