// devframe@0.5.4/dist/serialization-BD_qXGjd.mjs
import { t as diagnostics } from "./diagnostics-GitRGKbr.mjs";
//#region src/rpc/serialization.ts
/**
* Wire format used by the WS RPC transport.
*
* - **JSON (default, unprefixed):** payload is plain JSON text. Used when
*   the dispatched method is declared `jsonSerializable: true`. Encoded
*   via {@link strictJsonStringify} (rejects non-JSON values), decoded
*   via `JSON.parse`.
* - **Structured-clone (`s:` prefix):** payload is `s:` followed by
*   `structured-clone-es` text. Used when the method is declared
*   `jsonSerializable: false` (or omitted, the default). Round-trips
*   `Map`, `Set`, `Date`, `BigInt`, cycles, and class instances.
*
* birpc envelopes always start with `{`, so a leading byte that is not
* `s` is unambiguously JSON. Each direction independently chooses its
* encoding from local definitions — request and response are not
* coupled by a mirror rule.
*/
const STRUCTURED_CLONE_PREFIX = "s:";
/**
* `JSON.stringify` with a single-pass strict replacer.
*
* Throws `DF0020` synchronously when the value contains a type JSON
* cannot round-trip losslessly: `Map`, `Set`, `Date`, `BigInt`, class
* instances, or `undefined` inside an array (silently becomes `null`).
*
* Native pass-throughs (no extra work needed):
*   - circular references — `JSON.stringify` raises `TypeError`.
*   - `BigInt` at top level — caught here for a friendlier error path.
*
* Lenient cases (allowed without throwing):
*   - `undefined` as an object property — legitimate optional field;
*     JSON.stringify just omits it.
*   - `undefined` at the root — legitimate "action returned nothing".
*   - `Symbol` / `Function` values — semantically "drop me" in JSON.
*
* `fnName` is used only for the diagnostic message — pass the RPC
* function name when calling from a wire serializer / dump writer so
* the error points at the offending function.
*/
function strictJsonStringify(value, fnName = "") {
	return JSON.stringify(value, function strictReplacer(key, val) {
		const holder = this;
		const original = holder != null ? holder[key] : val;
		if (original === void 0) {
			if (Array.isArray(holder)) throw nonJsonAt(fnName, "undefined", holder, key);
			return val;
		}
		if (original === null) return val;
		if (typeof original === "bigint") throw nonJsonAt(fnName, "BigInt", holder, key);
		if (typeof original === "object") {
			if (original instanceof Map) throw nonJsonAt(fnName, "Map", holder, key);
			if (original instanceof Set) throw nonJsonAt(fnName, "Set", holder, key);
			if (original instanceof Date) throw nonJsonAt(fnName, "Date", holder, key);
			if (Array.isArray(original)) return val;
			const proto = Object.getPrototypeOf(original);
			if (proto !== null && proto !== Object.prototype) throw nonJsonAt(fnName, original.constructor?.name ?? "class instance", holder, key);
		}
		return val;
	});
}
function nonJsonAt(fnName, type, parent, key) {
	const path = formatPath(parent, key);
	return diagnostics.DF0020({
		name: fnName || "<anonymous>",
		type,
		path
	});
}
function formatPath(parent, key) {
	if (Array.isArray(parent)) return `[${key}]`;
	if (key === "") return "<root>";
	return key;
}
//#endregion
export { strictJsonStringify as n, STRUCTURED_CLONE_PREFIX as t };
