// devframe@0.5.4/dist/server-BBaBJaUL.mjs
import { createRpcServer } from "./rpc/server.mjs";
import { attachWsRpcTransport } from "./rpc/transports/ws-server.mjs";
import { t as getInternalContext } from "./context-DaKmhhHY.mjs";
import { WebSocketServer } from "ws";
import { AsyncLocalStorage } from "node:async_hooks";
import { createServer } from "node:http";
import { H3, toNodeHandler } from "h3";
//#region src/node/server.ts
/**
* Compose an h3 + WebSocket server for a devframe context. The RPC
* group is bound to `context.rpc.functions`; the WS endpoint lives on
* the same port as the HTTP server.
*/
async function startHttpAndWs(options) {
	const { context, port } = options;
	const bindHost = options.host ?? "localhost";
	const app = options.app ?? new H3();
	const httpServer = createServer(toNodeHandler(app));
	const wss = new WebSocketServer({ server: httpServer });
	const rpcHost = context.rpc;
	const asyncStorage = new AsyncLocalStorage();
	const rpcGroup = createRpcServer(rpcHost.functions, { rpcOptions: { resolver(name, fn) {
		const rpc = this;
		if (!fn) return void 0;
		return async function(...args) {
			return await asyncStorage.run({
				rpc,
				meta: rpc.$meta
			}, async () => {
				return (await fn).apply(this, args);
			});
		};
	} } });
	attachWsRpcTransport(rpcGroup, {
		wss,
		onDisconnected: (_ws, meta) => {
			rpcHost._emitSessionDisconnected(meta);
		}
	});
	rpcHost._rpcGroup = rpcGroup;
	rpcHost._asyncStorage = asyncStorage;
	rpcHost._authDisabled = options.auth === false;
	if (options.auth === false && !rpcHost.definitions.has("devframe:anonymous:auth")) rpcHost.register({
		name: "devframe:anonymous:auth",
		type: "action",
		handler: () => {
			const session = rpcHost.getCurrentRpcSession();
			if (session) session.meta.isTrusted = true;
			return { isTrusted: true };
		}
	});
	await new Promise((resolveListen) => {
		httpServer.listen(port, bindHost, () => resolveListen());
	});
	const address = httpServer.address();
	const resolvedPort = typeof address === "object" && address ? address.port : port;
	const origin = `http://${bindHost}:${resolvedPort}`;
	const internal = getInternalContext(context);
	const wsUrl = origin.replace(/^http/, "ws");
	internal.wsEndpoint = { url: wsUrl };
	if (options.onReady) await options.onReady({
		origin,
		port: resolvedPort,
		app
	});
	return {
		origin,
		port: resolvedPort,
		app,
		wss,
		rpcGroup,
		async close() {
			for (const ws of wss.clients) ws.terminate();
			await new Promise((r) => wss.close(() => r()));
			await new Promise((r) => httpServer.close(() => r()));
			if (getInternalContext(context).wsEndpoint?.url === wsUrl) getInternalContext(context).wsEndpoint = void 0;
		}
	};
}
//#endregion
export { startHttpAndWs as t };
