// unhead@3.1.4/dist/legacy.mjs
import { P as PromisesPlugin, T as TemplateParamsPlugin, A as AliasSortingPlugin } from './shared/unhead.BZYKjYpn.mjs';
import { d as defineHeadPlugin } from './shared/unhead.CUXLLRtV.mjs';
import { c as createUnhead } from './shared/unhead.CfgPMHXt.mjs';
import { c as createHead$1 } from './shared/unhead.DqXjmljN.mjs';
import { a as createHead$2 } from './shared/unhead.yYqx02R-.mjs';
import './shared/unhead.BGFxPGPQ.mjs';
import './shared/unhead.fg-0ge_u.mjs';
import './shared/unhead.mB5lMBMV.mjs';
import 'hookable';
import './shared/unhead.BBvzuk-m.mjs';
import './shared/unhead.Cdnk2khL.mjs';

const DeprecationsPlugin = /* @__PURE__ */ defineHeadPlugin({
  key: "deprecations",
  hooks: {
    "entries:normalize": ({ tags }) => {
      for (const tag of tags) {
        if (tag.props.children) {
          tag.innerHTML = tag.props.children;
          delete tag.props.children;
        }
        if (tag.props.hid) {
          tag.key = tag.props.hid;
          delete tag.props.hid;
        }
        if (tag.props.vmid) {
          tag.key = tag.props.vmid;
          delete tag.props.vmid;
        }
        if ("body" in tag.props) {
          if (tag.props.body) {
            tag.tagPosition = "bodyClose";
          }
          delete tag.props.body;
        }
        if (tag.props.renderPriority != null) {
          tag.tagPriority = tag.props.renderPriority;
          delete tag.props.renderPriority;
        }
      }
    }
  }
});
const legacyPlugins = [DeprecationsPlugin, PromisesPlugin, TemplateParamsPlugin, AliasSortingPlugin];
const activeHead = { value: null };
function getActiveHead() {
  return activeHead.value;
}
function createHead(options = {}) {
  return activeHead.value = createHead$1({
    ...options,
    plugins: [...legacyPlugins, ...options.plugins || []]
  });
}
function createServerHead(options = {}) {
  return activeHead.value = createHead$2({
    ...options,
    plugins: [...legacyPlugins, ...options.plugins || []]
  });
}
const createHeadCore = createUnhead;

export { DeprecationsPlugin, activeHead, createHead, createHeadCore, createServerHead, getActiveHead, legacyPlugins };
