// @slidev/types@52.16.0/dist/index.mjs
//#region src/setups.ts
function defineSetup$1(fn) {
	return fn;
}
const defineShikiSetup = defineSetup$1;
const defineUnoSetup = defineSetup$1;
const defineMonacoSetup = defineSetup$1;
const defineAppSetup = defineSetup$1;
const defineRootSetup = defineSetup$1;
const defineRoutesSetup = defineSetup$1;
const defineMermaidSetup = defineSetup$1;
const defineMermaidRendererSetup = defineSetup$1;
const defineKatexSetup = defineSetup$1;
const defineShortcutsSetup = defineSetup$1;
const defineTransformersSetup = defineSetup$1;
const definePreparserSetup = defineSetup$1;
const defineVitePluginsSetup = defineSetup$1;
const defineCodeRunnersSetup = defineSetup$1;
const defineContextMenuSetup = defineSetup$1;
//#endregion
//#region src/transform.ts
function defineSetup(fn) {
	return fn;
}
const defineMarkdownTransformer = defineSetup;
const defineCodeblockTransformer = defineSetup;
//#endregion
export { defineAppSetup, defineCodeRunnersSetup, defineCodeblockTransformer, defineContextMenuSetup, defineKatexSetup, defineMarkdownTransformer, defineMermaidRendererSetup, defineMermaidSetup, defineMonacoSetup, definePreparserSetup, defineRootSetup, defineRoutesSetup, defineShikiSetup, defineShortcutsSetup, defineTransformersSetup, defineUnoSetup, defineVitePluginsSetup };
