// @fix-webm-duration/parser@1.0.1/src/lib/WebmUint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebmUint = void 0;
const WebmBase_1 = require("./WebmBase");
function padHex(hex) {
    return hex.length % 2 === 1 ? "0" + hex : hex;
}
class WebmUint extends WebmBase_1.WebmBase {
    constructor(name, start = 0) {
        super(name, start);
    }
    getType() {
        return "Uint";
    }
    updateBySource() {
        // use hex representation of a number instead of number value
        this.data = "";
        for (let i = 0; i < this.source.length; i++) {
            const hex = this.source[i].toString(16);
            this.data += padHex(hex);
        }
    }
    updateByData() {
        const length = this.data.length / 2;
        this.source = new Uint8Array(length);
        for (let i = 0; i < length; i++) {
            const hex = this.data.substring(i * 2, i * 2 + 2);
            this.source[i] = parseInt(hex, 16);
        }
    }
    getValue() {
        return parseInt(this.data, 16);
    }
    setValue(value) {
        this.setData(padHex(value.toString(16)));
    }
}
exports.WebmUint = WebmUint;
