// @unhead/bundler@3.1.4/dist/shared/bundler.D1qf9ShC.mjs
import { existsSync, readFileSync } from 'node:fs';
import { createRequire } from 'node:module';
import { relative, resolve, dirname } from 'node:path';
import { pathToFileURL, fileURLToPath } from 'node:url';
import MagicString from 'magic-string';
import { parseAndWalk, ScopeTracker, walk, ScopeTrackerImport } from 'oxc-walker';
import { defineRpcFunction } from '@vitejs/devtools-kit';
import { parseSync } from 'oxc-parser';
import { parseURL, parseQuery } from 'ufo';
import { createUnplugin } from 'unplugin';
import { createContext, runInContext } from 'node:vm';
import { resolveMetaKeyType, resolveMetaKeyValue, resolvePackedMetaObjectValue } from 'unhead/utils';

const getConfigRpc = defineRpcFunction({
  name: "unhead:get-config",
  type: "static",
  setup: (ctx) => ({
    handler: () => ({
      cwd: ctx.cwd,
      mode: ctx.mode
    })
  })
});

async function tryRequire() {
  try {
    const mod = await import('@unhead/cli');
    return { runLint: mod.runLint };
  } catch {
    return null;
  }
}
const runLintRpc = defineRpcFunction({
  name: "unhead:run-lint",
  type: "static",
  setup: (ctx) => ({
    handler: async (args = {}) => {
      const lib = await tryRequire();
      if (!lib) {
        return {
          available: false,
          message: "Install @unhead/cli (and eslint as a peer) to enable in-devtools auditing."
        };
      }
      const start = Date.now();
      const mode = args.mode === "migrate" ? "migrate" : "audit";
      const cwd = ctx.cwd;
      const { results, errorCount, warningCount, fixableErrorCount, fixableWarningCount } = await lib.runLint({
        patterns: ["**/*.{js,mjs,cjs,ts,mts,cts,jsx,tsx,vue,svelte}"],
        mode,
        cwd,
        ignore: ["**/node_modules/**", "**/dist/**", "**/.output/**", "**/.nuxt/**"]
      });
      const files = results.map((r) => ({
        filePath: r.filePath,
        relativePath: relative(cwd, r.filePath),
        errorCount: r.errorCount,
        warningCount: r.warningCount,
        fixableErrorCount: r.fixableErrorCount,
        fixableWarningCount: r.fixableWarningCount,
        fixed: typeof r.output === "string",
        messages: (r.messages || []).map((m) => ({
          ruleId: m.ruleId ?? null,
          message: m.message,
          severity: m.severity === 2 ? "error" : "warn",
          line: m.line,
          column: m.column,
          endLine: m.endLine,
          endColumn: m.endColumn,
          // Only `fix` is auto-applied by `--fix`; suggestions are editor-only.
          fixable: Boolean(m.fix)
        }))
      })).filter((f) => f.errorCount > 0 || f.warningCount > 0 || f.fixed);
      const filesFixed = files.filter((f) => f.fixed).length;
      return {
        available: true,
        mode,
        files,
        errorCount,
        warningCount,
        fixableErrorCount,
        fixableWarningCount,
        filesFixed,
        durationMs: Date.now() - start
      };
    }
  })
});

const HEAD_COMPOSABLES = ["useHead", "useSeoMeta", "useHeadSafe", "useScript"];
const FILE_RE$1 = /\.(vue|tsx?|jsx?|svelte)$/;
const LEADING_SLASH_RE = /^\//;
const UNHEAD_VERSION_RE = /__UNHEAD_VERSION__ = ['"]'?["']/;
function findPkgRoot(fromUrl) {
  let dir = dirname(fileURLToPath(fromUrl));
  while (dir !== dirname(dir)) {
    if (existsSync(resolve(dir, "package.json")))
      return dir;
    dir = dirname(dir);
  }
  return dir;
}
const UNHEAD_ICON = `data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='32' height='32' viewBox='0 0 24 24'%3E%3Cdefs%3E%3ClinearGradient id='g' x1='0%25' y1='0%25' x2='100%25' y2='100%25'%3E%3Cstop offset='0%25' stop-color='%23FBBF24'/%3E%3Cstop offset='100%25' stop-color='%23f0db4f'/%3E%3C/linearGradient%3E%3Cmask id='m'%3E%3Crect width='100%25' height='100%25' fill='white'/%3E%3Cpath d='M12 32 L1 32 L15 15 Z' fill='black'/%3E%3C/mask%3E%3C/defs%3E%3Cpath fill='none' stroke='url(%23g)' stroke-linecap='round' stroke-linejoin='round' stroke-width='3' d='M6 4v14a2 2 0 0 0 2 2h8a2 2 0 0 0 2-2V4' mask='url(%23m)'/%3E%3C/svg%3E`;
const DEVTOOLS_UI_ROUTE = "/__unhead/";
function transformSourceLocations(code, id, root) {
  if (!HEAD_COMPOSABLES.some((c) => code.includes(c)))
    return;
  const s = new MagicString(code);
  let transformed = false;
  const relativePath = id.startsWith(root) ? id.slice(root.length).replace(LEADING_SLASH_RE, "") : id;
  parseAndWalk(code, id, {
    parseOptions: { lang: "ts" },
    enter(node) {
      if (node.type !== "CallExpression")
        return;
      const callee = node.callee;
      if (!callee)
        return;
      const name = callee.type === "Identifier" ? callee.name : callee.type === "MemberExpression" && callee.property?.type === "Identifier" ? callee.property.name : null;
      if (!name || !HEAD_COMPOSABLES.includes(name))
        return;
      const args = node.arguments;
      if (!args || args.length === 0)
        return;
      const lineNumber = code.slice(0, node.start).split("\n").length;
      const sourceValue = `${relativePath}:${lineNumber}`;
      if (args.length === 1) {
        const argEnd = args[0].end;
        s.appendRight(argEnd, `, { _source: ${JSON.stringify(sourceValue)} }`);
        transformed = true;
      } else if (args.length >= 2 && args[1].type === "ObjectExpression") {
        const objStart = args[1].start + 1;
        s.appendRight(objStart, ` _source: ${JSON.stringify(sourceValue)},`);
        transformed = true;
      }
    }
  });
  if (!transformed)
    return;
  return {
    code: s.toString(),
    map: s.generateMap({ includeContent: true, source: id })
  };
}
function unheadDevtools(options) {
  let root = "";
  let enabled = false;
  let bridgeCode;
  let unheadVersion = "";
  const pkgDir = findPkgRoot(import.meta.url);
  const devtoolsUiDir = resolve(pkgDir, "dist/devtools-ui");
  return {
    name: "@unhead/devtools",
    apply: "serve",
    configResolved(config) {
      root = config.root;
      enabled = config.plugins.some((p) => p.name?.startsWith("vite:devtools"));
      if (!enabled)
        return;
      if (options?._ctx) {
        options._ctx.addRuntimePlugin({
          import: { name: "devtoolsPlugin", source: "@unhead/bundler", as: "__unhead_devtoolsPlugin" },
          client: "window.__unhead_devtools__=_h",
          server: "_h.use(__unhead_devtoolsPlugin())"
        });
      }
      try {
        const unheadEntry = createRequire(import.meta.url).resolve("unhead");
        const unheadPkg = resolve(findPkgRoot(pathToFileURL(unheadEntry).href), "package.json");
        unheadVersion = JSON.parse(readFileSync(unheadPkg, "utf-8")).version || "";
      } catch {
      }
      const bridgePath = resolve(pkgDir, "dist/devtools/bridge.mjs");
      if (existsSync(bridgePath))
        bridgeCode = readFileSync(bridgePath, "utf-8");
    },
    configureServer(server) {
      if (!enabled)
        return;
      server.middlewares.use("/@unhead/bridge.mjs", async (_req, res) => {
        const result = await server.transformRequest("/@unhead/bridge.mjs");
        res.setHeader("Content-Type", "application/javascript");
        res.end(result?.code || 'console.warn("[unhead devtools] bridge not built")');
      });
    },
    resolveId(id) {
      if (!enabled)
        return;
      if (id === "/@unhead/bridge.mjs")
        return id;
    },
    load(id) {
      if (!enabled || id !== "/@unhead/bridge.mjs")
        return;
      if (!bridgeCode)
        return 'console.warn("[unhead devtools] bridge not built")';
      let code = bridgeCode;
      if (unheadVersion)
        code = code.replace(UNHEAD_VERSION_RE, `__UNHEAD_VERSION__ = '${unheadVersion}'`);
      const kitClientPath = resolve(pkgDir, "node_modules/@vitejs/devtools-kit/dist/client.js");
      if (existsSync(kitClientPath))
        return code.replace(`'@vitejs/devtools-kit/client'`, `'${kitClientPath}'`);
      return code;
    },
    transform: {
      filter: { id: FILE_RE$1 },
      handler(code, id) {
        if (!enabled)
          return;
        return transformSourceLocations(code, id, root);
      }
    },
    transformIndexHtml: {
      // Run before non-pre HTML transforms so the injected module import
      // goes through the full Vite plugin pipeline.
      order: "pre",
      handler() {
        if (!enabled)
          return [];
        return [{
          tag: "script",
          attrs: { type: "module" },
          children: `import("/@unhead/bridge.mjs")`,
          injectTo: "head"
        }];
      }
    },
    devtools: {
      setup(ctx) {
        if (existsSync(devtoolsUiDir)) {
          ctx.views.hostStatic(DEVTOOLS_UI_ROUTE, devtoolsUiDir);
        }
        ctx.docks.register({
          id: "unhead",
          title: "Unhead",
          icon: UNHEAD_ICON,
          type: "iframe",
          url: DEVTOOLS_UI_ROUTE
        });
        ctx.rpc.register(getConfigRpc);
        ctx.rpc.register(runLintRpc);
      }
    }
  };
}

const FILE_RE = /\.(vue|tsx?|jsx?|svelte)$/;
const UNHEAD_SOURCE_RE = /^(?:@unhead\/[^/]+|unhead)(?:\/[^?]*)?$/;
function createHeadTransformContext() {
  const registrations = [];
  return {
    addRuntimePlugin(reg) {
      registrations.push(reg);
    },
    getRegistrations() {
      return registrations;
    }
  };
}
function CreateHeadTransform(ctx) {
  let root = "";
  return {
    name: "@unhead/create-head-transform",
    apply: "serve",
    configResolved(config) {
      root = config.root;
    },
    transform: {
      filter: { id: FILE_RE },
      handler(code, id) {
        const registrations = ctx.getRegistrations();
        if (!registrations.length)
          return;
        if (!code.includes("createHead"))
          return;
        const isServer = this.environment?.config?.consumer === "server";
        const envRegistrations = registrations.filter((r) => isServer ? r.server : r.client);
        if (!envRegistrations.length)
          return;
        const s = new MagicString(code);
        let transformed = false;
        const directCreateHeadNames = /* @__PURE__ */ new Set();
        const namespaceNames = /* @__PURE__ */ new Set();
        parseAndWalk(code, id, {
          parseOptions: { lang: "ts" },
          enter(node) {
            if (node.type === "ImportDeclaration") {
              const source = node.source?.value;
              if (typeof source !== "string" || !UNHEAD_SOURCE_RE.test(source))
                return;
              for (const spec of node.specifiers || []) {
                if (spec.type === "ImportSpecifier" && spec.imported?.name === "createHead")
                  directCreateHeadNames.add(spec.local.name);
                else if (spec.type === "ImportNamespaceSpecifier")
                  namespaceNames.add(spec.local.name);
              }
              return;
            }
            if (node.type !== "CallExpression")
              return;
            const callee = node.callee;
            if (!callee)
              return;
            const isDirect = callee.type === "Identifier" && directCreateHeadNames.has(callee.name);
            const isNamespaced = callee.type === "MemberExpression" && callee.object?.type === "Identifier" && namespaceNames.has(callee.object.name) && callee.property?.type === "Identifier" && callee.property.name === "createHead";
            if (!isDirect && !isNamespaced)
              return;
            const statements = envRegistrations.map((r) => (isServer ? r.server : r.client).replace(/__ROOT__/g, JSON.stringify(root))).join(",");
            s.prependLeft(node.start, `((_h)=>(${statements},_h))(`);
            s.appendRight(node.end, `)`);
            transformed = true;
          }
        });
        if (!transformed)
          return;
        for (const reg of envRegistrations) {
          s.prepend(`import { ${reg.import.name} as ${reg.import.as} } from '${reg.import.source}';
`);
        }
        return {
          code: s.toString(),
          map: s.generateMap({ includeContent: true, source: id })
        };
      }
    }
  };
}

const NODE_MODULES_RE$2 = /[\\/]node_modules[\\/]/;
const TRANSFORM_RE$2 = /\.(?:(?:c|m)?j|t)sx?$/;
const SKIP_JS_TYPES = /* @__PURE__ */ new Set(["application/json", "application/ld+json", "speculationrules", "importmap"]);
const MinifyTransform = createUnplugin((options = {}) => {
  const jsMinifier = options.js !== false ? options.js : void 0;
  const cssMinifier = options.css !== false ? options.css : void 0;
  const doJS = !!jsMinifier;
  const doCSS = !!cssMinifier;
  const HEAD_FN_NAMES = /* @__PURE__ */ new Set(["useHead", "useServerHead"]);
  const CONTENT_PROPS = /* @__PURE__ */ new Set(["innerHTML", "textContent"]);
  return {
    name: "unhead:minify-transform",
    enforce: "post",
    transformInclude(id) {
      const { pathname, search } = parseURL(decodeURIComponent(pathToFileURL(id).href));
      const { type } = parseQuery(search);
      if (NODE_MODULES_RE$2.test(pathname))
        return false;
      if (options.filter?.include?.some((pattern) => id.match(pattern)))
        return true;
      if (options.filter?.exclude?.some((pattern) => id.match(pattern)))
        return false;
      if (pathname.endsWith(".vue") && (type === "script" || !search))
        return true;
      if (TRANSFORM_RE$2.test(pathname))
        return true;
      return false;
    },
    async transform(code, id) {
      if (!code.includes("useHead") && !code.includes("useServerHead"))
        return;
      let ast;
      try {
        ast = parseSync(id, code);
      } catch {
        return;
      }
      const scopeTracker = new ScopeTracker();
      const s = new MagicString(code);
      const pendingMinifications = [];
      walk(ast.program, {
        scopeTracker,
        enter(node, _parent) {
          if (node.type !== "CallExpression" || node.callee.type !== "Identifier")
            return;
          const decl = scopeTracker.getDeclaration(node.callee.name);
          let originalName;
          if (decl instanceof ScopeTrackerImport) {
            if (decl.node.type !== "ImportSpecifier" || decl.node.imported.type !== "Identifier")
              return;
            originalName = decl.node.imported.name;
          } else if (!decl && HEAD_FN_NAMES.has(node.callee.name)) {
            originalName = node.callee.name;
          } else {
            return;
          }
          if (!HEAD_FN_NAMES.has(originalName))
            return;
          const arg = node.arguments[0];
          if (!arg || arg.type !== "ObjectExpression")
            return;
          for (const prop of arg.properties) {
            if (prop.type !== "Property" || prop.key?.type !== "Identifier")
              continue;
            const tagType = prop.key.name;
            if (tagType !== "script" && tagType !== "style")
              continue;
            if (tagType === "script" && !doJS)
              continue;
            if (tagType === "style" && !doCSS)
              continue;
            const elements = prop.value?.type === "ArrayExpression" ? prop.value.elements : [prop.value];
            for (const element of elements) {
              if (!element || element.type !== "ObjectExpression")
                continue;
              processScriptOrStyleObject(element, tagType, code, s, pendingMinifications);
            }
          }
        }
      });
      await Promise.all(pendingMinifications);
      if (s.hasChanged()) {
        return {
          code: s.toString(),
          map: s.generateMap({ includeContent: true, source: id })
        };
      }
    }
  };
  function processScriptOrStyleObject(objectNode, tagType, code, s, pendingMinifications) {
    if (tagType === "script") {
      const typeProp = objectNode.properties.find(
        (p) => p.type === "Property" && p.key?.type === "Identifier" && p.key.name === "type"
      );
      if (typeProp?.value?.type === "Literal" && SKIP_JS_TYPES.has(typeProp.value.value))
        return;
    }
    for (const prop of objectNode.properties) {
      if (prop.type !== "Property" || prop.key?.type !== "Identifier")
        continue;
      if (!CONTENT_PROPS.has(prop.key.name))
        continue;
      if (prop.value?.type === "Literal") {
        const raw = prop.value.value;
        if (raw.length < 20)
          continue;
        pendingMinifications.push(
          minifyStringContent(raw, tagType).then((minified) => {
            if (minified && minified.length < raw.length) {
              s.overwrite(prop.value.start, prop.value.end, JSON.stringify(minified));
            }
          })
        );
      } else if (prop.value?.type === "TemplateLiteral" && prop.value.expressions.length === 0) {
        const raw = prop.value.quasis[0]?.value?.cooked;
        if (!raw || raw.length < 20)
          continue;
        pendingMinifications.push(
          minifyStringContent(raw, tagType).then((minified) => {
            if (minified && minified.length < raw.length) {
              s.overwrite(prop.value.start, prop.value.end, JSON.stringify(minified));
            }
          })
        );
      }
    }
  }
  async function minifyStringContent(content, tagType) {
    if (tagType === "script" && jsMinifier)
      return jsMinifier(content);
    if (tagType === "style" && cssMinifier)
      return cssMinifier(content);
    return null;
  }
});

const UNHEAD_MODULE_RE = /[\\/]node_modules[\\/](?:@unhead[\\/][^\\/]+|unhead)[\\/]/;
const HEAD_SSR_RE = /\bhead\.ssr\b/g;
const JS_RE = /\.(?:c|m)?js$/;
const SSRStaticReplace = createUnplugin(() => {
  let ssr = false;
  let enabled = true;
  return {
    name: "unhead:ssr-static-replace",
    enforce: "pre",
    transformInclude(id) {
      if (!enabled)
        return false;
      if (!UNHEAD_MODULE_RE.test(id))
        return false;
      return JS_RE.test(id);
    },
    transform(code) {
      if (!code.includes("head.ssr"))
        return;
      const s = new MagicString(code);
      for (const match of code.matchAll(HEAD_SSR_RE)) {
        s.overwrite(match.index, match.index + match[0].length, String(ssr));
      }
      if (s.hasChanged()) {
        return {
          code: s.toString(),
          map: s.generateMap({ includeContent: true })
        };
      }
    },
    webpack(ctx) {
      if (ctx.name === "server")
        ssr = true;
    },
    vite: {
      apply(_config, env) {
        if (env.command === "serve") {
          enabled = false;
          return true;
        }
        if (env.isSsrBuild)
          ssr = true;
        return true;
      }
    }
  };
});

const NODE_MODULES_RE$1 = /[\\/]node_modules[\\/]/;
const TRANSFORM_RE$1 = /\.(?:(?:c|m)?j|t)sx?$/;
const functionNames = [
  "useServerHead",
  "useServerHeadSafe",
  "useServerSeoMeta",
  // plugins
  "useSchemaOrg"
];
const TreeshakeServerComposables = createUnplugin((options = {}) => {
  options.enabled = options.enabled !== void 0 ? options.enabled : true;
  return {
    name: "unhead:remove-server-composables",
    enforce: "post",
    transformInclude(id) {
      if (!options.enabled)
        return false;
      const { pathname, search } = parseURL(decodeURIComponent(pathToFileURL(id).href));
      const { type } = parseQuery(search);
      if (NODE_MODULES_RE$1.test(pathname))
        return false;
      if (options.filter?.include?.some((pattern) => id.match(pattern)))
        return true;
      if (options.filter?.exclude?.some((pattern) => id.match(pattern)))
        return false;
      if (pathname.endsWith(".vue") && (type === "script" || !search))
        return true;
      if (TRANSFORM_RE$1.test(pathname))
        return true;
      return false;
    },
    transform(code, id) {
      if (!code.includes("useServerHead") && !code.includes("useServerHeadSafe") && !code.includes("useServerSeoMeta") && !code.includes("useSchemaOrg")) {
        return;
      }
      const ast = parseSync(id, code);
      const s = new MagicString(code);
      walk(ast.program, {
        enter(node) {
          if (node.type === "ExpressionStatement" && node.expression.type === "CallExpression" && node.expression.callee.type === "Identifier" && functionNames.includes(node.expression.callee.name)) {
            s.remove(node.start, node.end);
          }
        }
      });
      if (s.hasChanged()) {
        return {
          code: s.toString(),
          map: s.generateMap({ includeContent: true, source: id })
        };
      }
    },
    webpack(ctx) {
      if (ctx.name === "server")
        options.enabled = false;
    },
    vite: {
      apply(config, env) {
        if (env.isSsrBuild) {
          options.enabled = false;
          return true;
        }
        return false;
      }
    }
  };
});

const NODE_MODULES_RE = /[\\/]node_modules[\\/]/;
const TRANSFORM_RE = /\.(?:(?:c|m)?j|t)sx?$/;
const SEO_META_NAMES = /* @__PURE__ */ new Set(["useSeoMeta", "useServerSeoMeta"]);
const MEDIA_KEYS = /* @__PURE__ */ new Set(["ogImage", "ogVideo", "ogAudio", "twitterImage"]);
const UseSeoMetaTransform = createUnplugin((options = {}) => {
  options.imports = options.imports || true;
  function isValidPackage(s) {
    if (s === "unhead" || s.startsWith("@unhead")) {
      return true;
    }
    return [...options.importPaths || []].includes(s);
  }
  return {
    name: "unhead:use-seo-meta-transform",
    enforce: "post",
    transformInclude(id) {
      const { pathname, search } = parseURL(decodeURIComponent(pathToFileURL(id).href));
      const { type } = parseQuery(search);
      if (NODE_MODULES_RE.test(pathname))
        return false;
      if (options.filter?.include?.some((pattern) => id.match(pattern)))
        return true;
      if (options.filter?.exclude?.some((pattern) => id.match(pattern)))
        return false;
      if (pathname.endsWith(".vue") && (type === "script" || !search))
        return true;
      if (TRANSFORM_RE.test(pathname))
        return true;
      return false;
    },
    async transform(code, id) {
      if (!code.includes("useSeoMeta") && !code.includes("useServerSeoMeta"))
        return;
      const scopeTracker = new ScopeTracker();
      const ast = parseSync(id, code);
      const s = new MagicString(code);
      const importRewrites = /* @__PURE__ */ new Map();
      const valueReferenced = /* @__PURE__ */ new Set();
      const untransformedCallees = /* @__PURE__ */ new Set();
      walk(ast.program, {
        scopeTracker,
        enter(node, parent) {
          if (node.type === "Identifier" && !(parent?.type === "CallExpression" && parent.callee === node) && parent?.type !== "ImportSpecifier") {
            const decl2 = scopeTracker.getDeclaration(node.name);
            if (decl2 instanceof ScopeTrackerImport && isValidPackage(decl2.importNode.source.value) && decl2.node.type === "ImportSpecifier" && decl2.node.imported.type === "Identifier" && SEO_META_NAMES.has(decl2.node.imported.name)) {
              valueReferenced.add(decl2.node.imported.name);
            }
          }
          if (node.type !== "CallExpression" || node.callee.type !== "Identifier")
            return;
          const decl = scopeTracker.getDeclaration(node.callee.name);
          let originalName;
          let importDecl = null;
          if (decl instanceof ScopeTrackerImport) {
            if (!isValidPackage(decl.importNode.source.value) || decl.node.type !== "ImportSpecifier" || decl.node.imported.type !== "Identifier")
              return;
            originalName = decl.node.imported.name;
            importDecl = decl.importNode;
          } else if (!decl && SEO_META_NAMES.has(node.callee.name)) {
            originalName = node.callee.name;
          } else {
            return;
          }
          if (!SEO_META_NAMES.has(originalName))
            return;
          const properties = node.arguments[0]?.properties;
          if (!properties) {
            if (importDecl)
              untransformedCallees.add(originalName);
            return;
          }
          let output = [];
          const title = properties.find((property) => property.key?.name === "title");
          const titleTemplate = properties.find((property) => property.key?.name === "titleTemplate");
          const meta = properties.filter((property) => property.key?.name !== "title" && property.key?.name !== "titleTemplate");
          if (title || titleTemplate || originalName === "useSeoMeta") {
            output.push("useHead({");
            if (title) {
              output.push(`  title: ${code.substring(title.value.start, title.value.end)},`);
            }
            if (titleTemplate) {
              output.push(`  titleTemplate: ${code.substring(titleTemplate.value.start, titleTemplate.value.end)},`);
            }
          }
          if (originalName === "useServerSeoMeta") {
            if (output.length) {
              const secondArg = node.arguments[1];
              if (secondArg)
                output.push(`}, ${code.substring(secondArg.start, secondArg.end)});`);
              else
                output.push("});");
            }
            output.push("useServerHead({");
          }
          if (meta.length)
            output.push("  meta: [");
          meta.forEach((property) => {
            if (property.type === "SpreadElement") {
              output = false;
              return;
            }
            if (property.key.type !== "Identifier" || !property.value) {
              output = false;
              return;
            }
            if (output === false)
              return;
            const propertyKey = property.key;
            let key = resolveMetaKeyType(propertyKey.name);
            const keyValue = resolveMetaKeyValue(propertyKey.name);
            let valueKey = "content";
            if (keyValue === "charset") {
              valueKey = "charset";
              key = "charset";
            }
            let value = code.substring(property.value.start, property.value.end);
            if (MEDIA_KEYS.has(propertyKey.name)) {
              const expandObject = (objNode) => {
                const tags = [];
                for (const p of objNode.properties) {
                  if (p.type === "SpreadElement" || p.computed || p.method || p.kind !== "init" || p.key?.type !== "Identifier")
                    return false;
                  const name = p.key.name;
                  const suffix = name === "url" ? "" : `:${name === "secureUrl" ? "secure_url" : name}`;
                  tags.push(`    { ${key}: '${keyValue}${suffix}', ${valueKey}: ${code.substring(p.value.start, p.value.end)} },`);
                }
                return tags.join("\n");
              };
              if (property.value.type === "ObjectExpression") {
                const expanded = expandObject(property.value);
                if (expanded === false) {
                  output = false;
                  return;
                }
                output.push(expanded);
                return;
              }
              if (property.value.type === "ArrayExpression") {
                if (!property.value.elements.length)
                  return;
                const parts = [];
                for (const element of property.value.elements) {
                  if (!element || element.type !== "ObjectExpression") {
                    output = false;
                    return;
                  }
                  const expanded = expandObject(element);
                  if (expanded === false) {
                    output = false;
                    return;
                  }
                  parts.push(expanded);
                }
                output.push(parts.join("\n"));
                return;
              }
              const v = property.value;
              const primitive = typeof v.value === "string" || typeof v.value === "number" || typeof v.value === "boolean";
              const isScalar = v.type === "TemplateLiteral" || (v.type === "Literal" || v.type === "StringLiteral" || v.type === "NumericLiteral") && primitive;
              if (!isScalar) {
                output = false;
                return;
              }
            }
            if (property.value.type === "ArrayExpression") {
              const elements = property.value.elements;
              if (!elements.length)
                return;
              const metaTags = elements.map((element) => {
                if (element.type !== "ObjectExpression")
                  return `    { ${key}: '${keyValue}', ${valueKey}: ${code.substring(element.start, element.end)} },`;
                return element.properties.map((p) => {
                  const propKey = p.key.name;
                  const propValue = code.substring(p.value.start, p.value.end);
                  return `    { ${key}: '${keyValue}:${propKey}', ${valueKey}: ${propValue} },`;
                }).join("\n");
              });
              output.push(metaTags.join("\n"));
              return;
            } else if (property.value.type === "ObjectExpression") {
              const isStatic = property.value.properties.every((p) => p.value.type === "StringLiteral" && typeof p.value.value === "string");
              if (!isStatic) {
                output = false;
                return;
              }
              const context = createContext({
                resolvePackedMetaObjectValue
              });
              const start = property.value.start;
              const end = property.value.end;
              try {
                value = JSON.stringify(runInContext(`resolvePackedMetaObjectValue(${code.slice(start, end)})`, context));
              } catch {
                output = false;
                return;
              }
            }
            if (valueKey === "charset")
              output.push(`    { ${key}: ${value} },`);
            else
              output.push(`    { ${key}: '${keyValue}', ${valueKey}: ${value} },`);
          });
          if (output) {
            if (meta.length)
              output.push("  ]");
            if (node.arguments.length >= 2) {
              const optionsArg = code.substring(node.arguments[1].start, node.arguments[1].end);
              output.push(`}, ${optionsArg})`);
            } else {
              output.push("})");
            }
            s.overwrite(node.start, node.end, output.join("\n"));
            if (importDecl) {
              if (!importRewrites.has(importDecl))
                importRewrites.set(importDecl, /* @__PURE__ */ new Set());
              importRewrites.get(importDecl).add(originalName);
            }
          } else if (importDecl) {
            untransformedCallees.add(originalName);
          }
        }
      });
      if (options.imports && importRewrites.size > 0) {
        for (const [importNode, transformedNames] of importRewrites) {
          const newSpecifiers = /* @__PURE__ */ new Set();
          for (const spec of importNode.specifiers) {
            if (spec.type !== "ImportSpecifier")
              continue;
            const importedName = spec.imported.name;
            const keepOriginal = importedName === spec.local.name ? importedName : `${importedName} as ${spec.local.name}`;
            if (transformedNames.has(importedName)) {
              newSpecifiers.add(importedName.includes("Server") ? "useServerHead" : "useHead");
              if (valueReferenced.has(importedName) || untransformedCallees.has(importedName))
                newSpecifiers.add(keepOriginal);
            } else {
              newSpecifiers.add(keepOriginal);
            }
          }
          s.overwrite(
            importNode.specifiers[0].start,
            importNode.specifiers.at(-1).end,
            [...newSpecifiers].join(", ")
          );
        }
      }
      if (s.hasChanged()) {
        return {
          code: s.toString(),
          map: s.generateMap({ includeContent: true, source: id })
        };
      }
    }
  };
});

export { CreateHeadTransform as C, MinifyTransform as M, SSRStaticReplace as S, TreeshakeServerComposables as T, UseSeoMetaTransform as U, createHeadTransformContext as c, unheadDevtools as u };
